# Step-by-step scRNA-seq Pipeline

## Step 1. Load the R packages and the input data
- Load the R packages.

```{R, eval=FALSE}
# sc libraries
library(Seurat)
library(phateR)
library(DoubletFinder)
library(monocle)
library(slingshot)
library(URD)
library(GSVA)
library(limma)
library(plyr)
library(dplyr)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(CellChat)
library(velocyto.R)
library(SeuratWrappers)
library(stringr)
library(scran)
library(ggpubr)
library(viridis)
library(pheatmap)
library(parallel)
library(reticulate)
library(SCENIC)
library(feather)
library(AUCell)
library(RcisTarget)
library(Matrix)
library(foreach)
library(doParallel)
library(clusterProfiler)
library(OpenXGR)
# st libraries
library(RColorBrewer)
library(Rfast2)
library(SeuratDisk)
library(abcCellmap)
library(biomaRt)
library(copykat)
library(gelnet)
library(ggplot2)
library(parallelDist)
library(patchwork)
library(markdown)
# getpot
library(getopt)
library(tools)
# HemaScopeR
library(HemaScopeR)
```

- Set the paths for the input data, the output results, and the Python installation.
```{R, eval=FALSE}
input.data.dirs = c('./SRR7881399/outs/filtered_feature_bc_matrix',
                    './SRR7881400/outs/filtered_feature_bc_matrix',
                    './SRR7881401/outs/filtered_feature_bc_matrix',
                    './SRR7881402/outs/filtered_feature_bc_matrix',
                    './SRR7881403/outs/filtered_feature_bc_matrix',
                    './SRR7881404/outs/filtered_feature_bc_matrix',
                    './SRR7881405/outs/filtered_feature_bc_matrix',
                    './SRR7881406/outs/filtered_feature_bc_matrix',
                    './SRR7881407/outs/filtered_feature_bc_matrix',
                    './SRR7881408/outs/filtered_feature_bc_matrix',
                    './SRR7881409/outs/filtered_feature_bc_matrix',
                    './SRR7881410/outs/filtered_feature_bc_matrix',
                    './SRR7881411/outs/filtered_feature_bc_matrix',
                    './SRR7881412/outs/filtered_feature_bc_matrix',
                    './SRR7881413/outs/filtered_feature_bc_matrix',
                    './SRR7881414/outs/filtered_feature_bc_matrix',
                    './SRR7881415/outs/filtered_feature_bc_matrix',
                    './SRR7881416/outs/filtered_feature_bc_matrix',
                    './SRR7881417/outs/filtered_feature_bc_matrix',
                    './SRR7881418/outs/filtered_feature_bc_matrix',
                    './SRR7881419/outs/filtered_feature_bc_matrix',
                    './SRR7881420/outs/filtered_feature_bc_matrix',
                    './SRR7881421/outs/filtered_feature_bc_matrix',
                    './SRR7881422/outs/filtered_feature_bc_matrix',
                    './SRR7881423/outs/filtered_feature_bc_matrix')
output.dir = './output/'
pythonPath = '/home/anaconda3/envs/HemaScopeR/bin/python'
```
- Set the parameters for loading the data sets.
```{R, eval=FALSE}
project.names = c('SRR7881399',
                  'SRR7881400',
                  'SRR7881401',
                  'SRR7881402',
                  'SRR7881403',
                  'SRR7881404',
                  'SRR7881405',
                  'SRR7881406',
                  'SRR7881407',
                  'SRR7881408',
                  'SRR7881409',
                  'SRR7881410',
                  'SRR7881411',
                  'SRR7881412',
                  'SRR7881413',
                  'SRR7881414',
                  'SRR7881415',
                  'SRR7881416',
                  'SRR7881417',
                  'SRR7881418',
                  'SRR7881419',
                  'SRR7881420',
                  'SRR7881421',
                  'SRR7881422',
                  'SRR7881423')
gene.column = 2
min.cells = 10
min.feature = 200
mt.pattern = '^MT-'
Step1_Input_Data.type = 'cellranger-count'
```
- Create folders for saving the results of HemaScopeR analysis.
```{R, eval=FALSE}
wdir <- getwd()

if(is.null(pythonPath)==FALSE){ reticulate::use_python(pythonPath) }else{print('Please set the path of Python.')}

if (!file.exists(paste0(output.dir, '/HemaScopeR_results/'))) {
 dir.create(paste0(output.dir, '/HemaScopeR_results/'))
}

output.dir <- paste0(output.dir,'/HemaScopeR_results/')

if (!file.exists(paste0(output.dir, '/RDSfiles/'))) {
 dir.create(paste0(output.dir, '/RDSfiles/'))
}

previous_results_path <- paste0(output.dir, '/RDSfiles/')  
# if (Whether_load_previous_results) {
#    print('Loading the previous results...') 
#    Load_previous_results(previous_results_path = previous_results_path)
# }

# Step1. Input data-----------------------------------------------------------------------------

print('Step1. Input data.')
if (!file.exists(paste0(output.dir, '/Step1.Input_data/'))) {
  dir.create(paste0(output.dir, '/Step1.Input_data/'))
}  
```
- Load the data sets.
```{R, eval=FALSE}
file.copy(from = input.data.dirs, to = paste0(output.dir,'/Step1.Input_data/'), recursive = TRUE)

if(Step1_Input_Data.type == 'cellranger-count'){
  if(length(input.data.dirs) > 1){
    input.data.list <- c()
    for (i in 1:length(input.data.dirs)) {
        
          sc_data.temp <- Read10X(data.dir = input.data.dirs[i],
                                  gene.column = gene.column)
          sc_object.temp <- CreateSeuratObject(counts = sc_data.temp,
                                               project = project.names[i],
                                               min.cells = min.cells,
                                               min.feature = min.feature)
          sc_object.temp[["percent.mt"]] <- PercentageFeatureSet(sc_object.temp, pattern = mt.pattern)
          input.data.list <- c(input.data.list, sc_object.temp)}
      
  }else{
      
          sc_data <- Read10X(data.dir = input.data.dirs,
                             gene.column = gene.column)
          sc_object <- CreateSeuratObject(counts = sc_data,
                                          project = project.names,
                                          min.cells = min.cells,
                                          min.feature = min.feature)
          sc_object[["percent.mt"]] <- PercentageFeatureSet(sc_object, pattern = mt.pattern) 
      
  }
}else if(Step1_Input_Data.type == 'Seurat'){
  if(length(input.data.dirs) > 1){
    input.data.list <- c()
    for (i in 1:length(input.data.dirs)) {
          sc_object.temp <- readRDS(input.data.dirs[i])
          sc_object.temp[["percent.mt"]] <- PercentageFeatureSet(sc_object.temp, pattern = mt.pattern)
          input.data.list <- c(input.data.list, sc_object.temp)
    }
  }else{
      sc_object <- readRDS(input.data.dirs)
      sc_object[["percent.mt"]] <- PercentageFeatureSet(sc_object, pattern = mt.pattern) 
  }
}else if(Step1_Input_Data.type == 'Matrix'){
    if(length(input.data.dirs) > 1){
    input.data.list <- c()
    for (i in 1:length(input.data.dirs)) {
          sc_data.temp <- readRDS(input.data.dirs[i])
          sc_object.temp <- CreateSeuratObject(counts = sc_data.temp,
                                               project = project.names[i],
                                               min.cells = min.cells,
                                               min.feature = min.feature)
          sc_object.temp[["percent.mt"]] <- PercentageFeatureSet(sc_object.temp, pattern = mt.pattern)
          input.data.list <- c(input.data.list, sc_object.temp)}
      
  }else{
      
          sc_data <- readRDS(input.data.dirs)
          sc_object <- CreateSeuratObject(counts = sc_data,
                                          project = project.names,
                                          min.cells = min.cells,
                                          min.feature = min.feature)
          sc_object[["percent.mt"]] <- PercentageFeatureSet(sc_object, pattern = mt.pattern) 
      
  }
}else{
  stop('Please input data generated by the cellranger-count software, or a Seurat object, or a gene expression matrix. HemaScopeR does not support other formats of input data.')
}
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}

```

## Step 2. Quality Control
- Set the parameters for quality control.
```{R, eval=FALSE}
# quality control and preprocessing
nFeature_RNA.limit = 200
percent.mt.limit = 20
scale.factor = 10000
nfeatures = 3000
ndims = 50
vars.to.regress = NULL
PCs = 1:35
resolution = 0.4
n.neighbors = 50
# remove doublets
doublet.percentage = 0.04
doublerFinderwraper.PCs = 1:20
doublerFinderwraper.pN = 0.25
doublerFinderwraper.pK = 0.1
Step2_Quality_Control.RemoveBatches = TRUE
Step2_Quality_Control.RemoveDoublets = TRUE
```
- Create a folder for saving the results of quality control.
```{R, eval=FALSE}
print('Step2. Quality control.')
if (!file.exists(paste0(output.dir, '/Step2.Quality_control/'))) {
  dir.create(paste0(output.dir, '/Step2.Quality_control/'))
}
```
- Run the quality control process.
```{R, eval=FALSE}
if(length(input.data.dirs) > 1){
# preprocess and quality control for multiple scRNA-Seq data sets
sc_object <- QC_multiple_scRNASeq(seuratObjects = input.data.list,
                                  datasetID = project.names,
                                  output.dir = paste0(output.dir,'/Step2.Quality_control/'),
                                  Step2_Quality_Control.RemoveBatches = Step2_Quality_Control.RemoveBatches,
                                  Step2_Quality_Control.RemoveDoublets = Step2_Quality_Control.RemoveDoublets,
                                  nFeature_RNA.limit = nFeature_RNA.limit,
                                  percent.mt.limit = percent.mt.limit,
                                  scale.factor = scale.factor,
                                  nfeatures = nfeatures,
                                  ndims = ndims,
                                  vars.to.regress = vars.to.regress,
                                  PCs = PCs,
                                  resolution = resolution,
                                  n.neighbors = n.neighbors,
                                  percentage = doublet.percentage,
                                  doublerFinderwraper.PCs = doublerFinderwraper.PCs,
                                  doublerFinderwraper.pN = doublerFinderwraper.pN,
                                  doublerFinderwraper.pK = doublerFinderwraper.pK
                                  )

}else{
# preprocess and quality control for single scRNA-Seq data set
sc_object <- QC_single_scRNASeq(sc_object = sc_object,
                                datasetID = project.names,
                                output.dir = paste0(output.dir,'/Step2.Quality_control/'),
                                Step2_Quality_Control.RemoveDoublets = Step2_Quality_Control.RemoveDoublets,
                                nFeature_RNA.limit = nFeature_RNA.limit,
                                percent.mt.limit = percent.mt.limit,
                                scale.factor = scale.factor,
                                nfeatures = nfeatures,
                                vars.to.regress = vars.to.regress,
                                ndims = ndims,
                                PCs = PCs,
                                resolution = resolution,
                                n.neighbors = n.neighbors,
                                percentage = doublet.percentage,
                                doublerFinderwraper.PCs = doublerFinderwraper.PCs,
                                doublerFinderwraper.pN = doublerFinderwraper.pN,
                                doublerFinderwraper.pK = doublerFinderwraper.pK)
}
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
var <- get(var_name)  # Get the variable by its name
saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}
```

## Step 3. Clustering
- Set the parameters for clustering.
```{R, eval=FALSE}
PCs = 1:35
resolution = 0.4
n.neighbors = 50
```
- Create a folder for saving the results of Louvain clustering.
```{R, eval=FALSE}
print('Step3. Clustering.')
if (!file.exists(paste0(output.dir, '/Step3.Clustering/'))) {
  dir.create(paste0(output.dir, '/Step3.Clustering/'))
}   
```
- Run Louvian clustering.
```{R, eval=FALSE}
if( (length(input.data.dirs) > 1) & Step2_Quality_Control.RemoveBatches ){graph.name <- 'integrated_snn'}else{graph.name <- 'RNA_snn'}
sc_object <- FindNeighbors(sc_object, dims = PCs, k.param = n.neighbors, force.recalc = TRUE)
sc_object <- FindClusters(sc_object, resolution = resolution, graph.name = graph.name)
sc_object@meta.data$seurat_clusters <- as.character(as.numeric(sc_object@meta.data$seurat_clusters))

# plot clustering
pdf(paste0(paste0(output.dir,'/Step3.Clustering/'), '/sc_object ','tsne_cluster.pdf'), width = 6, height = 6)
 print(DimPlot(sc_object, reduction = "tsne", group.by = "seurat_clusters", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()

pdf(paste0(paste0(output.dir,'/Step3.Clustering/'), '/sc_object ','umap_cluster.pdf'), width = 6, height = 6)
 print(DimPlot(sc_object, reduction = "umap", group.by = "seurat_clusters", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()

png(paste0(paste0(output.dir,'/Step3.Clustering/'), '/sc_object ','tsne_cluster.png'), width = 600, height = 600)
 print(DimPlot(sc_object, reduction = "tsne", group.by = "seurat_clusters", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()

png(paste0(paste0(output.dir,'/Step3.Clustering/'), '/sc_object ','umap_cluster.png'), width = 600, height = 600)
 print(DimPlot(sc_object, reduction = "umap", group.by = "seurat_clusters", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()    
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```

## Step 4. Identify Cell Types
- Set the path for the database.
```{R, eval=FALSE}
databasePath = "~/HemaScopeR/database/"
```
- Set the parameters for cell type identification.
```{R, eval=FALSE}
Step4_Use_Which_Labels = 'clustering'
Step4_Cluster_Labels = NULL
Step4_Changed_Labels = NULL
Org = 'hsa'
ncores = 10
```
- Create a folder for saving the results of cell type identification.
```{R, eval=FALSE}
print('Step4. Identify cell types automatically.')
if (!file.exists(paste0(output.dir, '/Step4.Identify_Cell_Types/'))) {
dir.create(paste0(output.dir, '/Step4.Identify_Cell_Types/'))
} 
```
- Run the cell type identification process and the copy number variation (CNV) analysis.
```{R, eval=FALSE}
sc_object <- run_cell_annotation(object = sc_object, 
                               assay = 'RNA', 
                               species = Org,
                               output.dir = paste0(output.dir,'/Step4.Identify_Cell_Types/'))

if(Org == 'hsa'){
load(paste0(databasePath,"/HematoMap.reference.rdata"))
if(length(intersect(rownames(HematoMap.reference), rownames(sc_object))) < 1000){
      HematoMap.reference <- RenameGenesSeurat(obj = HematoMap.reference,
                                               newnames = toupper(rownames(HematoMap.reference)),
                                               gene.use = rownames(HematoMap.reference), 
                                               de.assay = "RNA",
                                               lassays = "RNA")
  }
  
if(sc_object@active.assay == 'integrated'){
    DefaultAssay(sc_object) <- 'RNA'
    sc_object <- mapDataToRef(ref_object = HematoMap.reference,
                              ref_labels = HematoMap.reference@meta.data$CellType,
                              query_object = sc_object,
                              PCs = PCs,
                              output.dir = paste0(output.dir, '/Step4.Identify_Cell_Types/'))
    DefaultAssay(sc_object) <- 'integrated'
}else{
    sc_object <- mapDataToRef(ref_object = HematoMap.reference,
                              ref_labels = HematoMap.reference@meta.data$CellType,
                              query_object = sc_object,
                              PCs = PCs,
                              output.dir = paste0(output.dir, '/Step4.Identify_Cell_Types/'))
}

}
```
- Set the cell labels.
```{R, eval=FALSE}
# set the cell labels
if(Step4_Use_Which_Labels == 'clustering'){
  sc_object@meta.data$selectLabels <- sc_object@meta.data$seurat_clusters
  Idents(sc_object) <- sc_object@meta.data$selectLabels
}else if(Step4_Use_Which_Labels == 'abcCellmap.1'){
  sc_object@meta.data$selectLabels <- sc_object@meta.data$Seurat.RNACluster
  Idents(sc_object) <- sc_object@meta.data$selectLabels
}else if(Step4_Use_Which_Labels == 'abcCellmap.2'){
  sc_object@meta.data$selectLabels <- sc_object@meta.data$scmap.RNACluster
  Idents(sc_object) <- sc_object@meta.data$selectLabels
}else if(Step4_Use_Which_Labels == 'abcCellmap.3'){
  sc_object@meta.data$selectLabels <- sc_object@meta.data$Seurat.Immunophenotype
  Idents(sc_object) <- sc_object@meta.data$selectLabels          
}else if(Step4_Use_Which_Labels == 'abcCellmap.4'){
  sc_object@meta.data$selectLabels <- sc_object@meta.data$scmap.Immunophenotype
  Idents(sc_object) <- sc_object@meta.data$selectLabels          
}else if(Step4_Use_Which_Labels == 'HematoMap'){
  if(Org == 'hsa'){
    sc_object@meta.data$selectLabels <- sc_object@meta.data$predicted.id
    Idents(sc_object) <- sc_object@meta.data$selectLabels
  }else{print("'HematoMap' is only applicable to human data ('Org' = 'hsa').")}    
}else if(Step4_Use_Which_Labels == 'changeLabels'){
  if (!is.null(Step4_Cluster_Labels) && !is.null(Step4_Changed_Labels) && length(Step4_Cluster_Labels) == length(Step4_Changed_Labels)){
   sc_object@meta.data$selectLabels <- plyr::mapvalues(sc_object@meta.data$seurat_clusters,
                                                       from = as.character(Step4_Cluster_Labels),
                                                       to = as.character(Step4_Changed_Labels),
                                                       warn_missing = FALSE)
   Idents(sc_object) <- sc_object@meta.data$selectLabels
  }else{
   print("Please input the 'Step4_Cluster_Labels' parameter as Seurat clustering labels, and the 'Step4_Changed_Labels' parameter as new labels. Please note that these two parameters should be of equal length.")
  }
}else{
  print('Please set the "Step4_Use_Which_Labels" parameter as "clustering", "abcCellmap.1", "abcCellmap.2", "HematoMap" or "changeLabels".')
}
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
var <- get(var_name)  # Get the variable by its name
saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```
- Run the CNV analysis.
```{R, eval=FALSE}
sc_CNV(sc_object=sc_object,
       save_path=paste0(output.dir,'/Step4.Identify_Cell_Types/'),
       assay = 'RNA',
       LOW.DR = 0.05,
       UP.DR = 0.1,
       win.size = 25,
       distance = "euclidean",
       genome = NULL,
       n.cores = ncores,
       species = Org)
```

## Step 5. Visualization
- Create a folder for saving the visualization results.
```{R, eval=FALSE}
print('Step5. Visualization.')
if (!file.exists(paste0(output.dir, '/Step5.Visualization/'))) {
  dir.create(paste0(output.dir, '/Step5.Visualization/'))
}
```
- The statistical results for the numbers and proportions of cell groups.
```{R, eval=FALSE}
# statistical results
cells_labels <- as.data.frame(cbind(rownames(sc_object@meta.data), as.character(sc_object@meta.data$selectLabels)))
colnames(cells_labels) <- c('cell_id', 'cluster_id')
cluster_counts <- cells_labels %>%
  group_by(cluster_id) %>%
  summarise(count = n())
total_cells <- nrow(cells_labels)
cluster_counts <- cluster_counts %>%
  mutate(proportion = count / total_cells)

cluster_counts <- as.data.frame(cluster_counts)

cluster_counts$percentages <- scales::percent(cluster_counts$proportion, accuracy = 0.1)
cluster_counts <- cluster_counts[,-which(colnames(cluster_counts)=='proportion')]

cluster_counts$cluster_id_count_percentages <- paste(cluster_counts$cluster_id, " (", cluster_counts$count, ' cells; ', cluster_counts$percentages, ")", sep='')

cluster_counts <- cluster_counts[order(cluster_counts$count, decreasing = TRUE),]

cluster_counts <- rbind(cluster_counts, c('Total', sum(cluster_counts$count), '100%', 'all cells'))

sc_object@meta.data$cluster_id_count_percentages <- mapvalues(sc_object@meta.data$selectLabels, 
                                                              from=cluster_counts$cluster_id, 
                                                              to=cluster_counts$cluster_id_count_percentages,
                                                              warn_missing=FALSE)

colnames(sc_object@meta.data)[which(colnames(sc_object@meta.data) == 'cluster_id_count_percentages')] <- paste('Total ', nrow(sc_object@meta.data), ' cells', sep='')

cluster_counts <- cluster_counts[,-which(colnames(cluster_counts)=='cluster_id_count_percentages')]

colnames(cluster_counts) <- c('Cell types', 'Cell counts', 'Percentages')

# names(colorvector) <- mapvalues(names(colorvector), 
#                                 from=cluster_counts$cluster_id, 
#                                 to=cluster_counts$cluster_id_count_percentages,
#                                 warn_missing=FALSE)

write.csv(cluster_counts, file=paste(paste0(output.dir, '/Step5.Visualization/'), '/cell types_cell counts_percentages.csv', sep=''), quote=FALSE, row.names=FALSE)
```
- The UMAP visualization.
```{R, eval=FALSE}
pdf(paste(paste0(output.dir, '/Step5.Visualization/'), '/cell types_cell counts_percentages_umap.pdf', sep=''), width = 14, height = 6)
  print(DimPlot(sc_object, 
                reduction = "umap", 
                group.by = paste('Total ', nrow(sc_object@meta.data), ' cells', sep=''), 
                label = FALSE, 
                pt.size = 0.1, 
                raster = FALSE))
dev.off()
```
- Set the parameters for phateR.
```{R, eval=FALSE}
phate.knn = 50
phate.npca = 20
phate.t = 10
phate.ndim = 2
```
- Run phateR for dimensional reduction and visualization.
```{R, eval=FALSE}
# run phateR
if( (length(input.data.dirs) > 1) & Step2_Quality_Control.RemoveBatches ){
    DefaultAssay(sc_object) <- 'integrated'
}else{
    DefaultAssay(sc_object) <- 'RNA'}

if(!is.null(pythonPath)){
  run_phateR(sc_object = sc_object,
             output.dir = paste0(output.dir,'/Step5.Visualization/'),
             pythonPath = pythonPath,
             phate.knn = phate.knn,
             phate.npca = phate.npca,
             phate.t = phate.t,
             phate.ndim = phate.ndim)     
}
```
- Perform visualization using UMAP and TSNE.
```{R, eval=FALSE}
# plot cell types
pdf(paste0(paste0(output.dir,'/Step5.Visualization/'), '/sc_object ','tsne cell types.pdf'), width = 6, height = 6)
 print(DimPlot(sc_object, reduction = "tsne", group.by = "ident", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()

pdf(paste0(paste0(output.dir,'/Step5.Visualization/'), '/sc_object ','umap cell types.pdf'), width = 6, height = 6)
 print(DimPlot(sc_object, reduction = "umap", group.by = "ident", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()

png(paste0(paste0(output.dir,'/Step5.Visualization/'), '/sc_object ','tsne cell types.png'), width = 600, height = 600)
 print(DimPlot(sc_object, reduction = "tsne", group.by = "ident", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()

png(paste0(paste0(output.dir,'/Step5.Visualization/'), '/sc_object ','umap cell types.png'), width = 600, height = 600)
 print(DimPlot(sc_object, reduction = "umap", group.by = "ident", label = FALSE, pt.size = 0.1, raster = FALSE))
dev.off()    
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```

## Step 6. Find DEGs
- Set the parameters for identifying differentially expressed genes.
```{R, eval=FALSE}
min.pct = 0.25
logfc.threshold = 0.25
```
- Create a folder for the DEGs analysis.
```{R, eval=FALSE}
print('Step6. Find DEGs.')
if (!file.exists(paste0(output.dir, '/Step6.Find_DEGs/'))) {
  dir.create(paste0(output.dir, '/Step6.Find_DEGs/'))
}   
```
- Identify DEGs using Wilcoxon Rank-Sum Test.
```{R, eval=FALSE}
sc_object.markers <- FindAllMarkers(sc_object, only.pos = TRUE, min.pct = min.pct, logfc.threshold = logfc.threshold)
write.csv(sc_object.markers,
          file = paste0(paste0(output.dir, '/Step6.Find_DEGs/'),'sc_object.markerGenes.csv'),
          quote=FALSE)
```
- Set the parameters for GPTCelltype.
```{R, eval=FALSE}
your_openai_API_key = ''
tissuename = 'human bone marrow'
gptmodel = 'gpt-3.5'
```
- Use GPTCelltype to assist cell type annotation.
```{R, eval=FALSE}
GPT_annotation( marker.genes = sc_object.markers,
                your_openai_API_key = your_openai_API_key,
                tissuename = tissuename,
                gptmodel = gptmodel,
                output.dir = paste0(output.dir, '/Step6.Find_DEGs/'))
```
- Perform GO and KEGG enrichment.
```{R, eval=FALSE}
# GO enrichment
if(Org=='mmu'){
  OrgDb <- 'org.Mm.eg.db'
}else if(Org=='hsa'){
  OrgDb <- 'org.Hs.eg.db'
}else{
  stop("Org should be 'mmu' or 'hsa'.")
}

HemaScopeREnrichment(DEGs=sc_object.markers,
                     OrgDb=OrgDb,
                     output.dir=paste0(output.dir, '/Step6.Find_DEGs/'))

sc_object.markers.top5 <- sc_object.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)

pdf(paste0(paste0(output.dir, '/Step6.Find_DEGs/'), 'sc_object_markerGenesTop5.pdf'), 
    width = 0.5*length(unique(sc_object.markers.top5$gene)), 
    height = 0.5*length(unique(Idents(sc_object))))
    print(DotPlot(sc_object,
            features = unique(sc_object.markers.top5$gene),
            cols=c("lightgrey",'red'))+theme(axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1)))
dev.off()

png(paste0(paste0(output.dir, '/Step6.Find_DEGs/'), 'sc_object_markerGenesTop5.png'), 
    width = 20*length(unique(sc_object.markers.top5$gene)), 
    height = 30*length(unique(Idents(sc_object))))
    print(DotPlot(sc_object,
            features = unique(sc_object.markers.top5$gene),
            cols=c("lightgrey",'red'))+theme(axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1)))
dev.off()  
```
- Create a folder for saving the results of gene network analysis.
```{R, eval=FALSE}
if (!file.exists(paste0(output.dir, '/Step6.Find_DEGs/OpenXGR/'))) {
  dir.create(paste0(output.dir, '/Step6.Find_DEGs/OpenXGR/'))
}
```
- Perform gene network analysis.
```{R, eval=FALSE}
OpenXGR_SAG(sc_object.markers = sc_object.markers,
            output.dir = paste0(output.dir, '/Step6.Find_DEGs/OpenXGR/'),
            subnet.size = 10)
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```

## Step 7. Assign Cell Cycles
- Create a folder for saving the results of cell cycle analysis.
```{R, eval=FALSE}
print('Step7. Assign cell cycles.')
if (!file.exists(paste0(output.dir, '/Step7.Assign_cell_cycles/'))) {
  dir.create(paste0(output.dir, '/Step7.Assign_cell_cycles/'))
}
```
- Set the parameters for the cell cycle analysis.
```{R, eval=FALSE}
cellcycleCutoff = NULL
```
- Run the cell cycle analysis.
```{R, eval=FALSE}
datasets.before.batch.removal <- readRDS(paste0(paste0(output.dir, '/RDSfiles/'),'datasets.before.batch.removal.rds'))  
sc_object <- cellCycle(sc_object=sc_object,
                       counts_matrix = GetAssayData(object = datasets.before.batch.removal, slot = "counts")%>%as.matrix(),
                       data_matrix = GetAssayData(object = datasets.before.batch.removal, slot = "data")%>%as.matrix(),
                       cellcycleCutoff = cellcycleCutoff,
                       cellTypeOrders = unique(sc_object@meta.data$selectLabels),
                       output.dir=paste0(output.dir, '/Step7.Assign_cell_cycles/'),
                       databasePath = databasePath,
                       Org = Org)
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```
## Step 8. Calculate Heterogeneity
- Create a folder for saving the results of heterogeneity calculation.
```{R, eval=FALSE}
print('Step8. Calculate heterogeneity.')
if (!file.exists(paste0(output.dir, '/Step8.Calculate_heterogeneity/'))) {
  dir.create(paste0(output.dir, '/Step8.Calculate_heterogeneity/'))
}  
```
- Run heterogeneity calculation process.
```{R, eval=FALSE}
expression_matrix <- GetAssayData(object = datasets.before.batch.removal, slot = "data")%>%as.matrix()
expression_matrix <- expression_matrix[,rownames(sc_object@meta.data)]
cell_types_groups <- as.data.frame(cbind(sc_object@meta.data$selectLabels,
                                         sc_object@meta.data$datasetID))
colnames(cell_types_groups) <- c('clusters', 'datasetID')

if(is.null(ViolinPlot.cellTypeOrders)){
  cellTypes_orders <- unique(sc_object@meta.data$selectLabels)
}else{
  cellTypes_orders <- ViolinPlot.cellTypeOrders  
} 

heterogeneity(expression_matrix = expression_matrix,
              cell_types_groups = cell_types_groups,
              cellTypeOrders = cellTypes_orders,
              output.dir = paste0(output.dir, '/Step8.Calculate_heterogeneity/'))
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
} 
```
## Step 9. Violin Plot for Marker Genes
- Create a folder for saving the violin plots of marker genes.
```{R, eval=FALSE}
print('Step9. Violin plot for marker genes.')
if (!file.exists(paste0(output.dir, '/Step9.Violin_plot_for_marker_genes/'))) {
  dir.create(paste0(output.dir, '/Step9.Violin_plot_for_marker_genes/'))
}
```
- Run violin plot visualization.
```{R, eval=FALSE}
if( (length(input.data.dirs) > 1) & Step2_Quality_Control.RemoveBatches ){
    DefaultAssay(sc_object) <- 'integrated'
}else{
    DefaultAssay(sc_object) <- 'RNA'}

dataMatrix <- GetAssayData(object = sc_object, slot = "scale.data")

if(is.null(marker.genes)&(Org == 'mmu')){
    # mpp genes are from 'The bone marrow microenvironment at single cell resolution'
    # the other genes are from 'single cell characterization of haematopoietic progenitors and their trajectories in homeostasis and perturbed haematopoiesis'
    # the aliases of these genes were changed in gecodeM16：Gpr64 -> Adgrg2, Sdpr -> Cavin2, Hbb-b1 -> Hbb-bs, Sfpi1 -> Spi1
    HSC_lineage_signatures <- c('Slamf1', 'Itga2b', 'Kit', 'Ly6a', 'Bmi1', 'Gata2', 'Hlf', 'Meis1', 'Mpl', 'Mcl1', 'Gfi1', 'Gfi1b', 'Hoxb5')
    Mpp_genes <- c('Mki67', 'Mpo', 'Elane', 'Ctsg', 'Calr')
    Erythroid_lineage_signatures <- c('Klf1', 'Gata1', 'Mpl', 'Epor', 'Vwf', 'Zfpm1', 'Fhl1', 'Adgrg2', 'Cavin2','Gypa', 'Tfrc', 'Hbb-bs', 'Hbb-y')
    Lymphoid_lineage_signatures <- c('Tcf3', 'Ikzf1', 'Notch1', 'Flt3', 'Dntt', 'Btg2', 'Tcf7', 'Rag1', 'Ptprc', 'Ly6a', 'Blnk')
    Myeloid_lineage_signatures <- c('Gfi1', 'Spi1', 'Mpo', 'Csf2rb', 'Csf1r', 'Gfi1b', 'Hk3', 'Csf2ra', 'Csf3r', 'Sp1', 'Fcgr3')
    marker.genes <- c(HSC_lineage_signatures, Mpp_genes, Erythroid_lineage_signatures, Lymphoid_lineage_signatures, Myeloid_lineage_signatures)
}else if(is.null(marker.genes)&(Org == 'hsa')){
    HSPCs_lineage_signatures <- c('CD34','KIT','AVP','FLT3','MME','CD7','CD38','CSF1R','FCGR1A','MPO','ELANE','IL3RA')
    Myeloids_lineage_signatures <- c('LYZ','CD36','MPO','FCGR1A','CD4','CD14','CD300E','ITGAX','FCGR3A','FLT3','AXL',
                  'SIGLEC6','CLEC4C','IRF4','LILRA4','IL3RA','IRF8','IRF7','XCR1','CD1C','THBD',
                  'MRC1','CD34','KIT','ITGA2B','PF4','CD9','ENG','KLF','TFRC')
    B_cells_lineage_signatures <- c('CD79A','IGLL1','RAG1','RAG2','VPREB1','MME','IL7R','DNTT','MKI67','PCNA','TCL1A','MS4A1','IGHD','CD27','IGHG3')
    T_NK_cells_lineage_signatures <- c('CD3D','CD3E','CD8A','CCR7','IL7R','SELL','KLRG1','CD27','GNLY',
                    'NKG7','PDCD1','TNFRSF9','LAG3','CD160','CD4','CD40LG','IL2RA',
                    'FOXP3','DUSP4','IL2RB','KLRF1','FCGR3A','NCAM1','XCL1','MKI67','PCNA','KLRF')
    marker.genes <- c(HSPCs_lineage_signatures, Myeloids_lineage_signatures, B_cells_lineage_signatures, T_NK_cells_lineage_signatures)
}

if(is.null(ViolinPlot.cellTypeOrders)){
  ViolinPlot.cellTypeOrders <- unique(sc_object@meta.data$selectLabels)
}

if(is.null(ViolinPlot.cellTypeColors)){
  ViolinPlot.cellTypeColors <- viridis::viridis(length(unique(sc_object@meta.data$selectLabels)))
}

combinedViolinPlot(dataMatrix = dataMatrix,
                   features = marker.genes,
                   CellTypes = sc_object@meta.data$selectLabels,
                   cellTypeOrders = ViolinPlot.cellTypeOrders,
                   cellTypeColors = ViolinPlot.cellTypeColors,
                   Org = Org,
                   output.dir = paste0(output.dir, '/Step9.Violin_plot_for_marker_genes/'),
                   databasePath = databasePath)
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```
## Step 10. Calculate Lineage Scores
- Create a folder for saving the results of lineage score calculation.
```{R, eval=FALSE}
print('Step10. Calculate lineage scores.')
# we use normalized data here
if (!file.exists(paste0(output.dir, '/Step10.Calculate_lineage_scores/'))) {
  dir.create(paste0(output.dir, '/Step10.Calculate_lineage_scores/'))
}
```
- Run lineage score calculation.
```{R, eval=FALSE}
if(is.null(lineage.genelist)&is.null(lineage.names)&(Org == 'mmu')){
    lineage.genelist <- c(list(HSC_lineage_signatures),
                          list(Mpp_genes),
                          list(Erythroid_lineage_signatures),
                          list(Lymphoid_lineage_signatures),
                          list(Myeloid_lineage_signatures))
    lineage.names <- c('HSC_lineage_signatures',
                       'Mpp_genes',
                       'Erythroid_lineage_signatures',
                       'Lymphoid_lineage_signatures',
                       'Myeloid_lineage_signatures')
}else if(is.null(lineage.genelist)&is.null(lineage.names)&(Org == 'hsa')){
    lineage.genelist <- c(list(HSPCs_lineage_signatures),
                          list(Myeloids_lineage_signatures),
                          list(B_cells_lineage_signatures),
                          list(T_NK_cells_lineage_signatures))
    lineage.names <- c('HSPCs_lineage_signatures',
                       'Myeloids_lineage_signatures',
                       'B_cells_lineage_signatures',
                       'T_NK_cells_lineage_signatures') 
}

if(is.null(ViolinPlot.cellTypeOrders)){
  cellTypes_orders <- unique(sc_object@meta.data$selectLabels)
}else{
  cellTypes_orders <- ViolinPlot.cellTypeOrders  
}
    
lineageScores(expression_matrix = expression_matrix,
              cellTypes = sc_object@meta.data$selectLabels,
              cellTypes_orders = cellTypes_orders,
              cellTypes_colors = ViolinPlot.cellTypeColors,
              groups = sc_object@meta.data$datasetID,
              groups_orders = unique(sc_object@meta.data$datasetID),
              groups_colors = groups_colors,
              lineage.genelist = lineage.genelist,
              lineage.names = lineage.names,
              Org = Org,
              output.dir = paste0(output.dir, '/Step10.Calculate_lineage_scores/'),
              databasePath = databasePath)
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
} 
```
## Step 11. GSVA
- Create a folder for saving the results of GSVA.
```{R, eval=FALSE}
print('Step11. GSVA.')
if (!file.exists(paste0(output.dir, '/Step11.GSVA/'))) {
  dir.create(paste0(output.dir, '/Step11.GSVA/'))
} 
```
- Run GSVA.
```{R, eval=FALSE}
setwd(wdir)

if(Org=='mmu'){
  load(paste0(databasePath,"/mouse_c2_v5p2.rdata"))
  GSVA.genelist <- Mm.c2
  assign('OrgDB', org.Mm.eg.db)
}else if(Org=='hsa'){
  load(paste0(databasePath,"/human_c2_v5p2.rdata"))
  GSVA.genelist <- Hs.c2
  assign('OrgDB', org.Hs.eg.db)
}else{
  stop("Org should be 'mmu' or 'hsa'.")
}

if(is.null(ViolinPlot.cellTypeOrders)){
  cellTypes_orders <- unique(sc_object@meta.data$selectLabels)
}else{
  cellTypes_orders <- ViolinPlot.cellTypeOrders  
}    
run_GSVA(sc_object = sc_object,
         GSVA.genelist = GSVA.genelist,
         GSVA.cellTypes = sc_object@meta.data$selectLabels,
         GSVA.cellTypes.orders = cellTypes_orders,
         GSVA.cellGroups = sc_object@meta.data$datasetID,
         GSVA.identify.cellType.features = Step11_GSVA.identify.cellType.features,
         GSVA.identify.diff.features = Step11_GSVA.identify.diff.features,
         GSVA.comparison.design = Step11_GSVA.comparison.design,
         OrgDB = OrgDB,
         output.dir = paste0(output.dir, '/Step11.GSVA/'))
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```
## Step 12. Construct Trajectories
- Load gene symbols and ensemble IDs.
```{R, eval=FALSE}
DefaultAssay(sc_object) <- 'RNA'
countsSlot <- GetAssayData(object = sc_object, slot = "counts")
gene_metadata <- as.data.frame(rownames(countsSlot))
rownames(gene_metadata) <- gene_metadata[,1]        
if(Org == 'mmu'){
   load(paste0(databasePath,"/mouseGeneSymbolandEnsembleID.rdata"))
   gene_metadata $ ensembleID <- mapvalues(x = gene_metadata[,1],
                                           from = mouseGeneSymbolandEnsembleID$geneName,
                                           to = mouseGeneSymbolandEnsembleID$ensemblIDNoDot,
                                           warn_missing = FALSE)
}else if(Org == 'hsa'){
   load(paste0(databasePath,"/humanGeneSymbolandEnsembleID.rdata"))
   gene_metadata $ ensembleID <- mapvalues(x = gene_metadata[,1],
                                           from = humanGeneSymbolandEnsembleID$geneName,
                                           to = humanGeneSymbolandEnsembleID$ensemblIDNoDot,
                                           warn_missing = FALSE) 
}
colnames(gene_metadata) <- c('gene_short_name','ensembleID')
```
- Create folders for saving the results of trajectory construction.
```{R, eval=FALSE}
print('Step12. Construct trajectories.')
if (!file.exists(paste0(output.dir, '/Step12.Construct_trajectories/'))) {
  dir.create(paste0(output.dir, '/Step12.Construct_trajectories/'))
}

if (!file.exists(paste0(output.dir, '/Step12.Construct_trajectories/monocle2/'))) {
  dir.create(paste0(output.dir, '/Step12.Construct_trajectories/monocle2/'))
} 

if (!file.exists(paste0(output.dir, '/Step12.Construct_trajectories/slingshot/'))) {
  dir.create(paste0(output.dir, '/Step12.Construct_trajectories/slingshot/'))
} 

if (!file.exists(paste0(output.dir, '/Step12.Construct_trajectories/scVelo/'))) {
  dir.create(paste0(output.dir, '/Step12.Construct_trajectories/scVelo/'))
} 
```
- Prepare the input data.
```{R, eval=FALSE}
if(is.null(Step12_Construct_Trajectories.clusters)){
    sc_object.subset <- sc_object
    countsSlot.subset <- GetAssayData(object = sc_object.subset, slot = "counts")
}else{
    sc_object.subset <- subset(sc_object, subset = selectLabels %in% Step12_Construct_Trajectories.clusters)
    countsSlot.subset <- GetAssayData(object = sc_object.subset, slot = "counts")
}
```
- Run monocle2.
```{R, eval=FALSE}
# monocle2

phenoData <- sc_object.subset@meta.data
featureData <- gene_metadata
run_monocle(cellData = countsSlot.subset,
            phenoData = phenoData,
            featureData = featureData,
            lowerDetectionLimit = 0.5,
            expressionFamily = VGAM::negbinomial.size(),
            cellTypes='selectLabels',
            monocle.orders=Step12_Construct_Trajectories.clusters,
            monocle.colors = ViolinPlot.cellTypeColors,
            output.dir = paste0(output.dir, '/Step12.Construct_trajectories/monocle2/'))
```
- Run slingshot.
```{R, eval=FALSE}
# slingshot

if( (length(input.data.dirs) > 1) & Step2_Quality_Control.RemoveBatches ){
    DefaultAssay(sc_object.subset) <- 'integrated'
}else{
    DefaultAssay(sc_object.subset) <- 'RNA'}
run_slingshot(slingshot.PCAembeddings = Embeddings(sc_object.subset, reduction = "pca")[, PCs],
              slingshot.cellTypes = sc_object.subset@meta.data$selectLabels,
              slingshot.start.clus = slingshot.start.clus,
              slingshot.end.clus = slingshot.end.clus,
              slingshot.colors = slingshot.colors,
              output.dir = paste0(output.dir, '/Step12.Construct_trajectories/slingshot/'))
```
- Run scVelo.
```{R, eval=FALSE}
# scVelo
if((!is.null(loom.files.path))&(!is.null(pythonPath))){
    
    prepareDataForScvelo(sc_object = sc_object.subset,
                         loom.files.path = loom.files.path,
                         scvelo.reduction = 'pca',
                         scvelo.column = 'selectLabels',
                         output.dir = paste0(output.dir, '/Step12.Construct_trajectories/scVelo/'))

    reticulate::py_run_string(paste0("import os\noutputDir = '", output.dir, "'"))
    reticulate::py_run_file(file.path(system.file(package = "HemaScopeR"), "python/sc_run_scvelo.py"), convert = FALSE)
}
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}
```
## Step 13. TF Analysis
- Create folders for saving the results of TF analysis.
```{R, eval=FALSE}
print('Step13. TF analysis.')
if (!file.exists(paste0(output.dir, '/Step13.TF_analysis/'))) {
  dir.create(paste0(output.dir, '/Step13.TF_analysis/'))
}
```
- Run SCENIC to perform TF analysis.
```{R, eval=FALSE}
run_SCENIC(countMatrix = countsSlot,
           cellTypes = sc_object@meta.data$selectLabels,
           datasetID = sc_object@meta.data$datasetID,
           cellTypes_colors = Step13_TF_Analysis.cellTypes_colors,
           cellTypes_orders = unique(sc_object@meta.data$selectLabels),
           groups_colors = Step13_TF_Analysis.groups_colors,
           groups_orders = unique(sc_object@meta.data$datasetID),
           Org = Org,
           output.dir = paste0(output.dir, '/Step13.TF_analysis/'),
           pythonPath = pythonPath,
           databasePath = databasePath)
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}
```
## Step 14. Cell-Cell Interaction
- Create folders for saving the results of cell-cell interaction analysis.
```{R, eval=FALSE}
print('Step14. Cell-cell interaction.')
if (!file.exists(paste0(output.dir, '/Step14.Cell_cell_interection/'))) {
  dir.create(paste0(output.dir, '/Step14.Cell_cell_interection/'))
}
```
- Run CellChat to perform cell-cell interaction analysis.
```{R, eval=FALSE}
tempwd <- getwd()
run_CellChat(data.input=countsSlot,
             labels = sc_object@meta.data$selectLabels,
             cell.orders = ViolinPlot.cellTypeOrders,
             cell.colors = ViolinPlot.cellTypeColors,
             sample.names = rownames(sc_object@meta.data),
             Org = Org,
             sorting = sorting,
             output.dir = paste0(output.dir, '/Step14.Cell_cell_interection/'))
setwd(tempwd)
```
- Save the variables.
```{R, eval=FALSE}
# Get the names of all variables in the current environment
variable_names <- ls()
# Loop through the variable names and save them as RDS files
for (var_name in variable_names) {
  var <- get(var_name)  # Get the variable by its name
  saveRDS(var, file = paste0(output.dir, '/RDSfiles/', var_name, ".rds"))  # Save as RDS with the variable's name
}  
```

