<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4 Step-by-step scRNA-seq Pipeline | HemaScope Tutorial</title>
  <meta name="description" content="This is the tutorial for HemaScope." />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="4 Step-by-step scRNA-seq Pipeline | HemaScope Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the tutorial for HemaScope." />
  <meta name="github-repo" content="https://github.com/ZhenyiWangTHU/HemaScopeR/" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4 Step-by-step scRNA-seq Pipeline | HemaScope Tutorial" />
  
  <meta name="twitter:description" content="This is the tutorial for HemaScope." />
  

<meta name="author" content="HemaScope team" />


<meta name="date" content="2024-11-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="integrated-scrna-seq-pipeline.html"/>
<link rel="next" href="integrated-st-pipeline.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">HemaScope</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>2</b> Installation</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installation.html"><a href="installation.html#create-a-new-conda-environment-and-activate-it"><i class="fa fa-check"></i><b>2.1</b> Create a new conda environment and activate it</a></li>
<li class="chapter" data-level="2.2" data-path="installation.html"><a href="installation.html#set-the-channels-in-conda"><i class="fa fa-check"></i><b>2.2</b> Set the channels in conda</a></li>
<li class="chapter" data-level="2.3" data-path="installation.html"><a href="installation.html#install-r"><i class="fa fa-check"></i><b>2.3</b> Install R</a></li>
<li class="chapter" data-level="2.4" data-path="installation.html"><a href="installation.html#install-required-r-packages"><i class="fa fa-check"></i><b>2.4</b> Install required R-packages</a></li>
<li class="chapter" data-level="2.5" data-path="installation.html"><a href="installation.html#create-the-required-python-virtual-environment"><i class="fa fa-check"></i><b>2.5</b> Create the required python virtual environment</a></li>
<li class="chapter" data-level="2.6" data-path="installation.html"><a href="installation.html#download-the-database"><i class="fa fa-check"></i><b>2.6</b> Download the database</a></li>
<li class="chapter" data-level="2.7" data-path="installation.html"><a href="installation.html#download-demo-datasets"><i class="fa fa-check"></i><b>2.7</b> Download demo datasets</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="integrated-scrna-seq-pipeline.html"><a href="integrated-scrna-seq-pipeline.html"><i class="fa fa-check"></i><b>3</b> Integrated scRNA-seq pipeline</a></li>
<li class="chapter" data-level="4" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html"><i class="fa fa-check"></i><b>4</b> Step-by-step scRNA-seq Pipeline</a>
<ul>
<li class="chapter" data-level="4.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-1.-load-the-input-data"><i class="fa fa-check"></i><b>4.2</b> Step 1. Load the input data</a></li>
<li class="chapter" data-level="4.3" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-2.-quality-control"><i class="fa fa-check"></i><b>4.3</b> Step 2. Quality Control</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments"><i class="fa fa-check"></i><b>4.3.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.3.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step2"><i class="fa fa-check"></i><b>4.3.2</b> codes for running step2</a></li>
<li class="chapter" data-level="4.3.3" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#outputs"><i class="fa fa-check"></i><b>4.3.3</b> Outputs</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-3.-clustering"><i class="fa fa-check"></i><b>4.4</b> Step 3. Clustering</a></li>
<li class="chapter" data-level="4.5" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-4.-identify-cell-types"><i class="fa fa-check"></i><b>4.5</b> Step 4. Identify Cell Types</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-abccellmap"><i class="fa fa-check"></i><b>4.5.1</b> codes for running abcCellmap</a></li>
<li class="chapter" data-level="4.5.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-the-cnv-analysis"><i class="fa fa-check"></i><b>4.5.2</b> codes for running the CNV analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-5.-visualization"><i class="fa fa-check"></i><b>4.6</b> Step 5. Visualization</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-peforming-three-dimensional-reduction-methods"><i class="fa fa-check"></i><b>4.6.1</b> codes for peforming three dimensional reduction methods</a></li>
<li class="chapter" data-level="4.6.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-calculating-the-proportions"><i class="fa fa-check"></i><b>4.6.2</b> codes for calculating the proportions</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-6.-find-degs"><i class="fa fa-check"></i><b>4.7</b> Step 6. Find DEGs</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-finding-degs"><i class="fa fa-check"></i><b>4.7.1</b> codes for finding DEGs</a></li>
<li class="chapter" data-level="4.7.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-using-gptcelltype"><i class="fa fa-check"></i><b>4.7.2</b> codes for using GPTCelltype</a></li>
<li class="chapter" data-level="4.7.3" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#perform-go-and-kegg-enrichment."><i class="fa fa-check"></i><b>4.7.3</b> Perform GO and KEGG enrichment.</a></li>
<li class="chapter" data-level="4.7.4" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#perform-subnetwork-analysis"><i class="fa fa-check"></i><b>4.7.4</b> Perform subnetwork analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-7.-assign-cell-cycles"><i class="fa fa-check"></i><b>4.8</b> Step 7. Assign Cell Cycles</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments-1"><i class="fa fa-check"></i><b>4.8.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.8.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-step7"><i class="fa fa-check"></i><b>4.8.2</b> codes for step7</a></li>
<li class="chapter" data-level="4.8.3" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#outputs-1"><i class="fa fa-check"></i><b>4.8.3</b> Outputs</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-8.-calculate-heterogeneity"><i class="fa fa-check"></i><b>4.9</b> Step 8. Calculate Heterogeneity</a>
<ul>
<li class="chapter" data-level="4.9.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments-2"><i class="fa fa-check"></i><b>4.9.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.9.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-step8"><i class="fa fa-check"></i><b>4.9.2</b> codes for step8</a></li>
</ul></li>
<li class="chapter" data-level="4.10" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-9.-violin-plot-for-marker-genes"><i class="fa fa-check"></i><b>4.10</b> Step 9. Violin Plot for Marker Genes</a>
<ul>
<li class="chapter" data-level="4.10.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments-3"><i class="fa fa-check"></i><b>4.10.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.10.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-step9"><i class="fa fa-check"></i><b>4.10.2</b> codes for step9</a></li>
</ul></li>
<li class="chapter" data-level="4.11" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-10.-calculate-lineage-scores"><i class="fa fa-check"></i><b>4.11</b> Step 10. Calculate Lineage Scores</a>
<ul>
<li class="chapter" data-level="4.11.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments-4"><i class="fa fa-check"></i><b>4.11.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.11.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-step10"><i class="fa fa-check"></i><b>4.11.2</b> codes for step10</a></li>
</ul></li>
<li class="chapter" data-level="4.12" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-11.-gsva"><i class="fa fa-check"></i><b>4.12</b> Step 11. GSVA</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments-5"><i class="fa fa-check"></i><b>4.12.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.12.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step11"><i class="fa fa-check"></i><b>4.12.2</b> codes for running step11</a></li>
</ul></li>
<li class="chapter" data-level="4.13" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-12.-construct-trajectories"><i class="fa fa-check"></i><b>4.13</b> Step 12. Construct Trajectories</a>
<ul>
<li class="chapter" data-level="4.13.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#data-preparation"><i class="fa fa-check"></i><b>4.13.1</b> data preparation</a></li>
<li class="chapter" data-level="4.13.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#monocle2"><i class="fa fa-check"></i><b>4.13.2</b> monocle2</a></li>
<li class="chapter" data-level="4.13.3" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#slingshot"><i class="fa fa-check"></i><b>4.13.3</b> Slingshot</a></li>
<li class="chapter" data-level="4.13.4" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#scvelo"><i class="fa fa-check"></i><b>4.13.4</b> scVelo</a></li>
</ul></li>
<li class="chapter" data-level="4.14" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-13.-tf-analysis"><i class="fa fa-check"></i><b>4.14</b> Step 13. TF Analysis</a>
<ul>
<li class="chapter" data-level="4.14.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments-9"><i class="fa fa-check"></i><b>4.14.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.14.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step13"><i class="fa fa-check"></i><b>4.14.2</b> codes for running step13</a></li>
</ul></li>
<li class="chapter" data-level="4.15" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#step-14.-cell-cell-interaction"><i class="fa fa-check"></i><b>4.15</b> Step 14. Cell-Cell Interaction</a>
<ul>
<li class="chapter" data-level="4.15.1" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#function-arguments-10"><i class="fa fa-check"></i><b>4.15.1</b> Function arguments:</a></li>
<li class="chapter" data-level="4.15.2" data-path="step-by-step-scrna-seq-pipeline.html"><a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step14"><i class="fa fa-check"></i><b>4.15.2</b> codes for running step14</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="integrated-st-pipeline.html"><a href="integrated-st-pipeline.html"><i class="fa fa-check"></i><b>5</b> Integrated ST pipeline</a>
<ul>
<li class="chapter" data-level="5.1" data-path="integrated-st-pipeline.html"><a href="integrated-st-pipeline.html#for-10x-visium-data"><i class="fa fa-check"></i><b>5.1</b> For 10X Visium data</a></li>
<li class="chapter" data-level="5.2" data-path="integrated-st-pipeline.html"><a href="integrated-st-pipeline.html#for-merfish-data"><i class="fa fa-check"></i><b>5.2</b> For MERFISH data</a></li>
<li class="chapter" data-level="5.3" data-path="integrated-st-pipeline.html"><a href="integrated-st-pipeline.html#for-stereo-seq-data"><i class="fa fa-check"></i><b>5.3</b> For stereo-seq data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html"><i class="fa fa-check"></i><b>6</b> Stey-by-step st-seq pipeline</a>
<ul>
<li class="chapter" data-level="6.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-1.-data-loading"><i class="fa fa-check"></i><b>6.1</b> Step 1. Data loading</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-11"><i class="fa fa-check"></i><b>6.1.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.1.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#funciton-behavior"><i class="fa fa-check"></i><b>6.1.2</b> Funciton behavior:</a></li>
<li class="chapter" data-level="6.1.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example"><i class="fa fa-check"></i><b>6.1.3</b> An example:</a></li>
<li class="chapter" data-level="6.1.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-2"><i class="fa fa-check"></i><b>6.1.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-2.-quality-control-1"><i class="fa fa-check"></i><b>6.2</b> Step 2. Quality Control</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-12"><i class="fa fa-check"></i><b>6.2.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.2.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior"><i class="fa fa-check"></i><b>6.2.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.2.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-1"><i class="fa fa-check"></i><b>6.2.3</b> An example:</a></li>
<li class="chapter" data-level="6.2.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-3"><i class="fa fa-check"></i><b>6.2.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-3.-clustering-1"><i class="fa fa-check"></i><b>6.3</b> Step 3. Clustering</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-13"><i class="fa fa-check"></i><b>6.3.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.3.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-1"><i class="fa fa-check"></i><b>6.3.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.3.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-2"><i class="fa fa-check"></i><b>6.3.3</b> An example:</a></li>
<li class="chapter" data-level="6.3.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-4"><i class="fa fa-check"></i><b>6.3.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-4.-degs"><i class="fa fa-check"></i><b>6.4</b> Step 4. DEGs</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-14"><i class="fa fa-check"></i><b>6.4.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.4.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-2"><i class="fa fa-check"></i><b>6.4.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.4.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-3"><i class="fa fa-check"></i><b>6.4.3</b> An example:</a></li>
<li class="chapter" data-level="6.4.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-5"><i class="fa fa-check"></i><b>6.4.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-5.-spatially-variable-features"><i class="fa fa-check"></i><b>6.5</b> Step 5. Spatially variable features</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-15"><i class="fa fa-check"></i><b>6.5.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.5.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-3"><i class="fa fa-check"></i><b>6.5.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.5.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-4"><i class="fa fa-check"></i><b>6.5.3</b> An example:</a></li>
<li class="chapter" data-level="6.5.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-6"><i class="fa fa-check"></i><b>6.5.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-6.-spatial-interaction"><i class="fa fa-check"></i><b>6.6</b> Step 6. Spatial interaction</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-16"><i class="fa fa-check"></i><b>6.6.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.6.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-4"><i class="fa fa-check"></i><b>6.6.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.6.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-5"><i class="fa fa-check"></i><b>6.6.3</b> An example:</a></li>
<li class="chapter" data-level="6.6.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-7"><i class="fa fa-check"></i><b>6.6.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-7.-cnv-analysis"><i class="fa fa-check"></i><b>6.7</b> Step 7. CNV analysis</a>
<ul>
<li class="chapter" data-level="6.7.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-17"><i class="fa fa-check"></i><b>6.7.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.7.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-5"><i class="fa fa-check"></i><b>6.7.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.7.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-6"><i class="fa fa-check"></i><b>6.7.3</b> An example:</a></li>
<li class="chapter" data-level="6.7.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-8"><i class="fa fa-check"></i><b>6.7.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.8" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-8.-deconvolution"><i class="fa fa-check"></i><b>6.8</b> Step 8. Deconvolution</a>
<ul>
<li class="chapter" data-level="6.8.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-18"><i class="fa fa-check"></i><b>6.8.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.8.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-6"><i class="fa fa-check"></i><b>6.8.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.8.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-7"><i class="fa fa-check"></i><b>6.8.3</b> An example:</a></li>
<li class="chapter" data-level="6.8.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-9"><i class="fa fa-check"></i><b>6.8.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.9" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-9.-cell-cycle"><i class="fa fa-check"></i><b>6.9</b> Step 9. Cell cycle</a>
<ul>
<li class="chapter" data-level="6.9.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-19"><i class="fa fa-check"></i><b>6.9.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.9.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-7"><i class="fa fa-check"></i><b>6.9.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.9.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-8"><i class="fa fa-check"></i><b>6.9.3</b> An example:</a></li>
<li class="chapter" data-level="6.9.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-10"><i class="fa fa-check"></i><b>6.9.4</b> Outputs:</a></li>
</ul></li>
<li class="chapter" data-level="6.10" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#step-10.-niche-analysis"><i class="fa fa-check"></i><b>6.10</b> Step 10. Niche analysis</a>
<ul>
<li class="chapter" data-level="6.10.1" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-arguments-20"><i class="fa fa-check"></i><b>6.10.1</b> Function arguments:</a></li>
<li class="chapter" data-level="6.10.2" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#function-behavior-8"><i class="fa fa-check"></i><b>6.10.2</b> Function behavior:</a></li>
<li class="chapter" data-level="6.10.3" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#an-example-9"><i class="fa fa-check"></i><b>6.10.3</b> An example:</a></li>
<li class="chapter" data-level="6.10.4" data-path="stey-by-step-st-seq-pipeline.html"><a href="stey-by-step-st-seq-pipeline.html#outputs-11"><i class="fa fa-check"></i><b>6.10.4</b> Outputs:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="step-by-step-shiny.html"><a href="step-by-step-shiny.html"><i class="fa fa-check"></i><b>7</b> Step-by-step shiny</a>
<ul>
<li class="chapter" data-level="7.1" data-path="step-by-step-shiny.html"><a href="step-by-step-shiny.html#option-1-run-shiny-on-linux"><i class="fa fa-check"></i><b>7.1</b> Option 1: Run shiny on Linux</a></li>
<li class="chapter" data-level="7.2" data-path="step-by-step-shiny.html"><a href="step-by-step-shiny.html#option-2-run-shiny-on-rstudio-web-page"><i class="fa fa-check"></i><b>7.2</b> Option 2: Run shiny on Rstudio web page</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="step-by-step-shiny.html"><a href="step-by-step-shiny.html#step-1.-enter-r-and-get-the-path-of-the-installed-r-packages"><i class="fa fa-check"></i><b>7.2.1</b> Step 1. Enter R and get the path of the installed R packages</a></li>
<li class="chapter" data-level="7.2.2" data-path="step-by-step-shiny.html"><a href="step-by-step-shiny.html#step-2.-run-shiny-code"><i class="fa fa-check"></i><b>7.2.2</b> Step 2. Run shiny code</a></li>
<li class="chapter" data-level="7.2.3" data-path="step-by-step-shiny.html"><a href="step-by-step-shiny.html#step-3.-use-hemascopeshiny-via-the-gui"><i class="fa fa-check"></i><b>7.2.3</b> Step 3. Use HemaScopeShiny via the GUI</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html"><i class="fa fa-check"></i><b>8</b> Operation Manual for the HemaScopeCloud</a>
<ul>
<li class="chapter" data-level="8.1" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#user-login"><i class="fa fa-check"></i><b>8.1</b> User Login</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#enter-the-url-in-a-web-browser"><i class="fa fa-check"></i><b>8.1.1</b> Enter the URL in a web browser:</a></li>
<li class="chapter" data-level="8.1.2" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#to-obtain-free-computational-resources"><i class="fa fa-check"></i><b>8.1.2</b> To obtain free computational resources:</a></li>
<li class="chapter" data-level="8.1.3" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#to-browse-hemascopecloud-without-needing-computational-resources"><i class="fa fa-check"></i><b>8.1.3</b> To browse HemaScopeCloud without needing computational resources:</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#homepage"><i class="fa fa-check"></i><b>8.2</b> Homepage</a></li>
<li class="chapter" data-level="8.3" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#data-page"><i class="fa fa-check"></i><b>8.3</b> Data Page</a></li>
<li class="chapter" data-level="8.4" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#analysis-page"><i class="fa fa-check"></i><b>8.4</b> Analysis Page</a></li>
<li class="chapter" data-level="8.5" data-path="operation-manual-for-the-hemascopecloud.html"><a href="operation-manual-for-the-hemascopecloud.html#projects-page"><i class="fa fa-check"></i><b>8.5</b> Projects page</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="hemascopedocker.html"><a href="hemascopedocker.html"><i class="fa fa-check"></i><b>9</b> HemaScopeDocker</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">HemaScope Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="step-by-step-scrna-seq-pipeline" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">4</span> Step-by-step scRNA-seq Pipeline<a href="step-by-step-scrna-seq-pipeline.html#step-by-step-scrna-seq-pipeline" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="before-you-begin" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Before you begin<a href="step-by-step-scrna-seq-pipeline.html#before-you-begin" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li>Load the R packages.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="step-by-step-scrna-seq-pipeline.html#cb14-1" tabindex="-1"></a><span class="co"># sc libraries</span></span>
<span id="cb14-2"><a href="step-by-step-scrna-seq-pipeline.html#cb14-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb14-3"><a href="step-by-step-scrna-seq-pipeline.html#cb14-3" tabindex="-1"></a><span class="fu">library</span>(phateR)</span>
<span id="cb14-4"><a href="step-by-step-scrna-seq-pipeline.html#cb14-4" tabindex="-1"></a><span class="fu">library</span>(DoubletFinder)</span>
<span id="cb14-5"><a href="step-by-step-scrna-seq-pipeline.html#cb14-5" tabindex="-1"></a><span class="fu">library</span>(monocle)</span>
<span id="cb14-6"><a href="step-by-step-scrna-seq-pipeline.html#cb14-6" tabindex="-1"></a><span class="fu">library</span>(slingshot)</span>
<span id="cb14-7"><a href="step-by-step-scrna-seq-pipeline.html#cb14-7" tabindex="-1"></a><span class="fu">library</span>(GSVA)</span>
<span id="cb14-8"><a href="step-by-step-scrna-seq-pipeline.html#cb14-8" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb14-9"><a href="step-by-step-scrna-seq-pipeline.html#cb14-9" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb14-10"><a href="step-by-step-scrna-seq-pipeline.html#cb14-10" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb14-11"><a href="step-by-step-scrna-seq-pipeline.html#cb14-11" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span>
<span id="cb14-12"><a href="step-by-step-scrna-seq-pipeline.html#cb14-12" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb14-13"><a href="step-by-step-scrna-seq-pipeline.html#cb14-13" tabindex="-1"></a><span class="fu">library</span>(CellChat)</span>
<span id="cb14-14"><a href="step-by-step-scrna-seq-pipeline.html#cb14-14" tabindex="-1"></a><span class="fu">library</span>(velocyto.R)</span>
<span id="cb14-15"><a href="step-by-step-scrna-seq-pipeline.html#cb14-15" tabindex="-1"></a><span class="fu">library</span>(SeuratWrappers)</span>
<span id="cb14-16"><a href="step-by-step-scrna-seq-pipeline.html#cb14-16" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb14-17"><a href="step-by-step-scrna-seq-pipeline.html#cb14-17" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb14-18"><a href="step-by-step-scrna-seq-pipeline.html#cb14-18" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb14-19"><a href="step-by-step-scrna-seq-pipeline.html#cb14-19" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb14-20"><a href="step-by-step-scrna-seq-pipeline.html#cb14-20" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb14-21"><a href="step-by-step-scrna-seq-pipeline.html#cb14-21" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb14-22"><a href="step-by-step-scrna-seq-pipeline.html#cb14-22" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb14-23"><a href="step-by-step-scrna-seq-pipeline.html#cb14-23" tabindex="-1"></a><span class="fu">library</span>(SCENIC)</span>
<span id="cb14-24"><a href="step-by-step-scrna-seq-pipeline.html#cb14-24" tabindex="-1"></a><span class="fu">library</span>(feather)</span>
<span id="cb14-25"><a href="step-by-step-scrna-seq-pipeline.html#cb14-25" tabindex="-1"></a><span class="fu">library</span>(AUCell)</span>
<span id="cb14-26"><a href="step-by-step-scrna-seq-pipeline.html#cb14-26" tabindex="-1"></a><span class="fu">library</span>(RcisTarget)</span>
<span id="cb14-27"><a href="step-by-step-scrna-seq-pipeline.html#cb14-27" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb14-28"><a href="step-by-step-scrna-seq-pipeline.html#cb14-28" tabindex="-1"></a><span class="fu">library</span>(foreach)</span>
<span id="cb14-29"><a href="step-by-step-scrna-seq-pipeline.html#cb14-29" tabindex="-1"></a><span class="fu">library</span>(doParallel)</span>
<span id="cb14-30"><a href="step-by-step-scrna-seq-pipeline.html#cb14-30" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb14-31"><a href="step-by-step-scrna-seq-pipeline.html#cb14-31" tabindex="-1"></a><span class="fu">library</span>(GPTCelltype)</span>
<span id="cb14-32"><a href="step-by-step-scrna-seq-pipeline.html#cb14-32" tabindex="-1"></a><span class="fu">library</span>(openai)</span>
<span id="cb14-33"><a href="step-by-step-scrna-seq-pipeline.html#cb14-33" tabindex="-1"></a><span class="fu">library</span>(HematoMap)</span>
<span id="cb14-34"><a href="step-by-step-scrna-seq-pipeline.html#cb14-34" tabindex="-1"></a><span class="co"># st libraries</span></span>
<span id="cb14-35"><a href="step-by-step-scrna-seq-pipeline.html#cb14-35" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb14-36"><a href="step-by-step-scrna-seq-pipeline.html#cb14-36" tabindex="-1"></a><span class="fu">library</span>(Rfast2)</span>
<span id="cb14-37"><a href="step-by-step-scrna-seq-pipeline.html#cb14-37" tabindex="-1"></a><span class="fu">library</span>(SeuratDisk)</span>
<span id="cb14-38"><a href="step-by-step-scrna-seq-pipeline.html#cb14-38" tabindex="-1"></a><span class="fu">library</span>(abcCellmap)</span>
<span id="cb14-39"><a href="step-by-step-scrna-seq-pipeline.html#cb14-39" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb14-40"><a href="step-by-step-scrna-seq-pipeline.html#cb14-40" tabindex="-1"></a><span class="fu">library</span>(copykat)</span>
<span id="cb14-41"><a href="step-by-step-scrna-seq-pipeline.html#cb14-41" tabindex="-1"></a><span class="fu">library</span>(gelnet)</span>
<span id="cb14-42"><a href="step-by-step-scrna-seq-pipeline.html#cb14-42" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb14-43"><a href="step-by-step-scrna-seq-pipeline.html#cb14-43" tabindex="-1"></a><span class="fu">library</span>(parallelDist)</span>
<span id="cb14-44"><a href="step-by-step-scrna-seq-pipeline.html#cb14-44" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb14-45"><a href="step-by-step-scrna-seq-pipeline.html#cb14-45" tabindex="-1"></a><span class="fu">library</span>(markdown)</span>
<span id="cb14-46"><a href="step-by-step-scrna-seq-pipeline.html#cb14-46" tabindex="-1"></a><span class="co"># getpot</span></span>
<span id="cb14-47"><a href="step-by-step-scrna-seq-pipeline.html#cb14-47" tabindex="-1"></a><span class="fu">library</span>(getopt)</span>
<span id="cb14-48"><a href="step-by-step-scrna-seq-pipeline.html#cb14-48" tabindex="-1"></a><span class="fu">library</span>(tools)</span>
<span id="cb14-49"><a href="step-by-step-scrna-seq-pipeline.html#cb14-49" tabindex="-1"></a><span class="co"># HemaScopeR</span></span>
<span id="cb14-50"><a href="step-by-step-scrna-seq-pipeline.html#cb14-50" tabindex="-1"></a><span class="fu">library</span>(HemaScopeR)</span></code></pre></div>
<ul>
<li>Set the paths for the output results, and the Python installation.</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="step-by-step-scrna-seq-pipeline.html#cb15-1" tabindex="-1"></a>output.dir <span class="ot">=</span> <span class="st">&#39;./output&#39;</span></span>
<span id="cb15-2"><a href="step-by-step-scrna-seq-pipeline.html#cb15-2" tabindex="-1"></a>pythonPath <span class="ot">=</span> <span class="st">&#39;/home/anaconda3/envs/HemaScopeR/bin/python&#39;</span></span></code></pre></div>
<ul>
<li>Create folders for saving the results of HemaScopeR analysis.</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="step-by-step-scrna-seq-pipeline.html#cb16-1" tabindex="-1"></a>wdir <span class="ot">&lt;-</span> <span class="fu">getwd</span>()</span>
<span id="cb16-2"><a href="step-by-step-scrna-seq-pipeline.html#cb16-2" tabindex="-1"></a></span>
<span id="cb16-3"><a href="step-by-step-scrna-seq-pipeline.html#cb16-3" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(pythonPath)<span class="sc">==</span><span class="cn">FALSE</span>){ reticulate<span class="sc">::</span><span class="fu">use_python</span>(pythonPath) }<span class="cf">else</span>{<span class="fu">print</span>(<span class="st">&#39;Please set the path of Python.&#39;</span>)}</span>
<span id="cb16-4"><a href="step-by-step-scrna-seq-pipeline.html#cb16-4" tabindex="-1"></a></span>
<span id="cb16-5"><a href="step-by-step-scrna-seq-pipeline.html#cb16-5" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/HemaScopeR_results&#39;</span>))) {</span>
<span id="cb16-6"><a href="step-by-step-scrna-seq-pipeline.html#cb16-6" tabindex="-1"></a> <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/HemaScopeR_results&#39;</span>),<span class="at">recursive =</span>T)</span>
<span id="cb16-7"><a href="step-by-step-scrna-seq-pipeline.html#cb16-7" tabindex="-1"></a>}</span>
<span id="cb16-8"><a href="step-by-step-scrna-seq-pipeline.html#cb16-8" tabindex="-1"></a></span>
<span id="cb16-9"><a href="step-by-step-scrna-seq-pipeline.html#cb16-9" tabindex="-1"></a>output.dir <span class="ot">&lt;-</span> <span class="fu">paste0</span>(output.dir,<span class="st">&#39;/HemaScopeR_results&#39;</span>)</span>
<span id="cb16-10"><a href="step-by-step-scrna-seq-pipeline.html#cb16-10" tabindex="-1"></a></span>
<span id="cb16-11"><a href="step-by-step-scrna-seq-pipeline.html#cb16-11" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>))) {</span>
<span id="cb16-12"><a href="step-by-step-scrna-seq-pipeline.html#cb16-12" tabindex="-1"></a> <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>))</span>
<span id="cb16-13"><a href="step-by-step-scrna-seq-pipeline.html#cb16-13" tabindex="-1"></a>}</span>
<span id="cb16-14"><a href="step-by-step-scrna-seq-pipeline.html#cb16-14" tabindex="-1"></a></span>
<span id="cb16-15"><a href="step-by-step-scrna-seq-pipeline.html#cb16-15" tabindex="-1"></a><span class="co">#set the path for loading previous results, if necessary</span></span>
<span id="cb16-16"><a href="step-by-step-scrna-seq-pipeline.html#cb16-16" tabindex="-1"></a>previous_results_path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>)  </span>
<span id="cb16-17"><a href="step-by-step-scrna-seq-pipeline.html#cb16-17" tabindex="-1"></a><span class="co"># if (Whether_load_previous_results) {</span></span>
<span id="cb16-18"><a href="step-by-step-scrna-seq-pipeline.html#cb16-18" tabindex="-1"></a><span class="co">#    print(&#39;Loading the previous results...&#39;) </span></span>
<span id="cb16-19"><a href="step-by-step-scrna-seq-pipeline.html#cb16-19" tabindex="-1"></a><span class="co">#    Load_previous_results(previous_results_path = previous_results_path)</span></span>
<span id="cb16-20"><a href="step-by-step-scrna-seq-pipeline.html#cb16-20" tabindex="-1"></a><span class="co"># }</span></span></code></pre></div>
</div>
<div id="step-1.-load-the-input-data" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> Step 1. Load the input data<a href="step-by-step-scrna-seq-pipeline.html#step-1.-load-the-input-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li>Create a folder for step1</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="step-by-step-scrna-seq-pipeline.html#cb17-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step1. Input data.&#39;</span>)</span>
<span id="cb17-2"><a href="step-by-step-scrna-seq-pipeline.html#cb17-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step1.Input_data/&#39;</span>))) {</span>
<span id="cb17-3"><a href="step-by-step-scrna-seq-pipeline.html#cb17-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step1.Input_data/&#39;</span>))</span>
<span id="cb17-4"><a href="step-by-step-scrna-seq-pipeline.html#cb17-4" tabindex="-1"></a>}  </span></code></pre></div>
<ul>
<li>Set the parameters for loading the data sets.</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="step-by-step-scrna-seq-pipeline.html#cb18-1" tabindex="-1"></a>input.data.dirs <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&#39;./SRR7881399/outs/filtered_feature_bc_matrix&#39;</span>)<span class="co">#,</span></span>
<span id="cb18-2"><a href="step-by-step-scrna-seq-pipeline.html#cb18-2" tabindex="-1"></a>                    <span class="co">#&#39;./SRR7881400/outs/filtered_feature_bc_matrix&#39;,</span></span>
<span id="cb18-3"><a href="step-by-step-scrna-seq-pipeline.html#cb18-3" tabindex="-1"></a>                    <span class="co">#&#39;./SRR7881401/outs/filtered_feature_bc_matrix&#39;,</span></span>
<span id="cb18-4"><a href="step-by-step-scrna-seq-pipeline.html#cb18-4" tabindex="-1"></a>                    <span class="co">#&#39;./SRR7881402/outs/filtered_feature_bc_matrix&#39;,</span></span>
<span id="cb18-5"><a href="step-by-step-scrna-seq-pipeline.html#cb18-5" tabindex="-1"></a>                    <span class="co">#&#39;./SRR7881403/outs/filtered_feature_bc_matrix&#39;</span></span>
<span id="cb18-6"><a href="step-by-step-scrna-seq-pipeline.html#cb18-6" tabindex="-1"></a></span>
<span id="cb18-7"><a href="step-by-step-scrna-seq-pipeline.html#cb18-7" tabindex="-1"></a>project.names <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&#39;SRR7881399&#39;</span>)<span class="co">#,</span></span>
<span id="cb18-8"><a href="step-by-step-scrna-seq-pipeline.html#cb18-8" tabindex="-1"></a>                  <span class="co">#&#39;SRR7881400&#39;,</span></span>
<span id="cb18-9"><a href="step-by-step-scrna-seq-pipeline.html#cb18-9" tabindex="-1"></a>                  <span class="co">#&#39;SRR7881401&#39;,</span></span>
<span id="cb18-10"><a href="step-by-step-scrna-seq-pipeline.html#cb18-10" tabindex="-1"></a>                  <span class="co">#&#39;SRR7881402&#39;,</span></span>
<span id="cb18-11"><a href="step-by-step-scrna-seq-pipeline.html#cb18-11" tabindex="-1"></a>                  <span class="co">#&#39;SRR7881403&#39;</span></span>
<span id="cb18-12"><a href="step-by-step-scrna-seq-pipeline.html#cb18-12" tabindex="-1"></a>gene.column <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb18-13"><a href="step-by-step-scrna-seq-pipeline.html#cb18-13" tabindex="-1"></a>min.cells <span class="ot">=</span> <span class="dv">10</span></span>
<span id="cb18-14"><a href="step-by-step-scrna-seq-pipeline.html#cb18-14" tabindex="-1"></a>min.feature <span class="ot">=</span> <span class="dv">200</span></span>
<span id="cb18-15"><a href="step-by-step-scrna-seq-pipeline.html#cb18-15" tabindex="-1"></a>mt.pattern <span class="ot">=</span> <span class="st">&#39;^MT-&#39;</span> <span class="co"># set &#39;^mt-&#39; for mouse data</span></span>
<span id="cb18-16"><a href="step-by-step-scrna-seq-pipeline.html#cb18-16" tabindex="-1"></a>Step1_Input_Data.type <span class="ot">=</span> <span class="st">&#39;cellranger-count&#39;</span></span>
<span id="cb18-17"><a href="step-by-step-scrna-seq-pipeline.html#cb18-17" tabindex="-1"></a>loom.files.path <span class="ot">=</span><span class="st">&quot;./SRR7881399/loom&quot;</span></span></code></pre></div>
<ul>
<li>Load the data sets</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="step-by-step-scrna-seq-pipeline.html#cb19-1" tabindex="-1"></a><span class="fu">file.copy</span>(<span class="at">from =</span> input.data.dirs, <span class="at">to =</span> <span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step1.Input_data/&#39;</span>), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb19-2"><a href="step-by-step-scrna-seq-pipeline.html#cb19-2" tabindex="-1"></a></span>
<span id="cb19-3"><a href="step-by-step-scrna-seq-pipeline.html#cb19-3" tabindex="-1"></a><span class="cf">if</span>(Step1_Input_Data.type <span class="sc">==</span> <span class="st">&#39;cellranger-count&#39;</span>){</span>
<span id="cb19-4"><a href="step-by-step-scrna-seq-pipeline.html#cb19-4" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb19-5"><a href="step-by-step-scrna-seq-pipeline.html#cb19-5" tabindex="-1"></a>    input.data.list <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb19-6"><a href="step-by-step-scrna-seq-pipeline.html#cb19-6" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(input.data.dirs)) {</span>
<span id="cb19-7"><a href="step-by-step-scrna-seq-pipeline.html#cb19-7" tabindex="-1"></a>        </span>
<span id="cb19-8"><a href="step-by-step-scrna-seq-pipeline.html#cb19-8" tabindex="-1"></a>          sc_data.temp <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> input.data.dirs[i],</span>
<span id="cb19-9"><a href="step-by-step-scrna-seq-pipeline.html#cb19-9" tabindex="-1"></a>                                  <span class="at">gene.column =</span> gene.column)</span>
<span id="cb19-10"><a href="step-by-step-scrna-seq-pipeline.html#cb19-10" tabindex="-1"></a>          sc_object.temp <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> sc_data.temp,</span>
<span id="cb19-11"><a href="step-by-step-scrna-seq-pipeline.html#cb19-11" tabindex="-1"></a>                                               <span class="at">project =</span> project.names[i],</span>
<span id="cb19-12"><a href="step-by-step-scrna-seq-pipeline.html#cb19-12" tabindex="-1"></a>                                               <span class="at">min.cells =</span> min.cells,</span>
<span id="cb19-13"><a href="step-by-step-scrna-seq-pipeline.html#cb19-13" tabindex="-1"></a>                                               <span class="at">min.feature =</span> min.feature)</span>
<span id="cb19-14"><a href="step-by-step-scrna-seq-pipeline.html#cb19-14" tabindex="-1"></a>          sc_object.temp[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_object.temp, <span class="at">pattern =</span> mt.pattern)</span>
<span id="cb19-15"><a href="step-by-step-scrna-seq-pipeline.html#cb19-15" tabindex="-1"></a>          input.data.list <span class="ot">&lt;-</span> <span class="fu">c</span>(input.data.list, sc_object.temp)}</span>
<span id="cb19-16"><a href="step-by-step-scrna-seq-pipeline.html#cb19-16" tabindex="-1"></a>      </span>
<span id="cb19-17"><a href="step-by-step-scrna-seq-pipeline.html#cb19-17" tabindex="-1"></a>  }<span class="cf">else</span>{</span>
<span id="cb19-18"><a href="step-by-step-scrna-seq-pipeline.html#cb19-18" tabindex="-1"></a>      </span>
<span id="cb19-19"><a href="step-by-step-scrna-seq-pipeline.html#cb19-19" tabindex="-1"></a>          sc_data <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> input.data.dirs,</span>
<span id="cb19-20"><a href="step-by-step-scrna-seq-pipeline.html#cb19-20" tabindex="-1"></a>                             <span class="at">gene.column =</span> gene.column)</span>
<span id="cb19-21"><a href="step-by-step-scrna-seq-pipeline.html#cb19-21" tabindex="-1"></a>          sc_object <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> sc_data,</span>
<span id="cb19-22"><a href="step-by-step-scrna-seq-pipeline.html#cb19-22" tabindex="-1"></a>                                          <span class="at">project =</span> project.names,</span>
<span id="cb19-23"><a href="step-by-step-scrna-seq-pipeline.html#cb19-23" tabindex="-1"></a>                                          <span class="at">min.cells =</span> min.cells,</span>
<span id="cb19-24"><a href="step-by-step-scrna-seq-pipeline.html#cb19-24" tabindex="-1"></a>                                          <span class="at">min.feature =</span> min.feature)</span>
<span id="cb19-25"><a href="step-by-step-scrna-seq-pipeline.html#cb19-25" tabindex="-1"></a>          sc_object[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_object, <span class="at">pattern =</span> mt.pattern) </span>
<span id="cb19-26"><a href="step-by-step-scrna-seq-pipeline.html#cb19-26" tabindex="-1"></a>      </span>
<span id="cb19-27"><a href="step-by-step-scrna-seq-pipeline.html#cb19-27" tabindex="-1"></a>  }</span>
<span id="cb19-28"><a href="step-by-step-scrna-seq-pipeline.html#cb19-28" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step1_Input_Data.type <span class="sc">==</span> <span class="st">&#39;Seurat&#39;</span>){</span>
<span id="cb19-29"><a href="step-by-step-scrna-seq-pipeline.html#cb19-29" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb19-30"><a href="step-by-step-scrna-seq-pipeline.html#cb19-30" tabindex="-1"></a>    input.data.list <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb19-31"><a href="step-by-step-scrna-seq-pipeline.html#cb19-31" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(input.data.dirs)) {</span>
<span id="cb19-32"><a href="step-by-step-scrna-seq-pipeline.html#cb19-32" tabindex="-1"></a>          sc_object.temp <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(input.data.dirs[i])</span>
<span id="cb19-33"><a href="step-by-step-scrna-seq-pipeline.html#cb19-33" tabindex="-1"></a>          sc_object.temp[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_object.temp, <span class="at">pattern =</span> mt.pattern)</span>
<span id="cb19-34"><a href="step-by-step-scrna-seq-pipeline.html#cb19-34" tabindex="-1"></a>          input.data.list <span class="ot">&lt;-</span> <span class="fu">c</span>(input.data.list, sc_object.temp)</span>
<span id="cb19-35"><a href="step-by-step-scrna-seq-pipeline.html#cb19-35" tabindex="-1"></a>    }</span>
<span id="cb19-36"><a href="step-by-step-scrna-seq-pipeline.html#cb19-36" tabindex="-1"></a>  }<span class="cf">else</span>{</span>
<span id="cb19-37"><a href="step-by-step-scrna-seq-pipeline.html#cb19-37" tabindex="-1"></a>      sc_object <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(input.data.dirs)</span>
<span id="cb19-38"><a href="step-by-step-scrna-seq-pipeline.html#cb19-38" tabindex="-1"></a>      sc_object[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_object, <span class="at">pattern =</span> mt.pattern) </span>
<span id="cb19-39"><a href="step-by-step-scrna-seq-pipeline.html#cb19-39" tabindex="-1"></a>  }</span>
<span id="cb19-40"><a href="step-by-step-scrna-seq-pipeline.html#cb19-40" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step1_Input_Data.type <span class="sc">==</span> <span class="st">&#39;Matrix&#39;</span>){</span>
<span id="cb19-41"><a href="step-by-step-scrna-seq-pipeline.html#cb19-41" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb19-42"><a href="step-by-step-scrna-seq-pipeline.html#cb19-42" tabindex="-1"></a>    input.data.list <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb19-43"><a href="step-by-step-scrna-seq-pipeline.html#cb19-43" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(input.data.dirs)) {</span>
<span id="cb19-44"><a href="step-by-step-scrna-seq-pipeline.html#cb19-44" tabindex="-1"></a>          sc_data.temp <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(input.data.dirs[i])</span>
<span id="cb19-45"><a href="step-by-step-scrna-seq-pipeline.html#cb19-45" tabindex="-1"></a>          sc_object.temp <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> sc_data.temp,</span>
<span id="cb19-46"><a href="step-by-step-scrna-seq-pipeline.html#cb19-46" tabindex="-1"></a>                                               <span class="at">project =</span> project.names[i],</span>
<span id="cb19-47"><a href="step-by-step-scrna-seq-pipeline.html#cb19-47" tabindex="-1"></a>                                               <span class="at">min.cells =</span> min.cells,</span>
<span id="cb19-48"><a href="step-by-step-scrna-seq-pipeline.html#cb19-48" tabindex="-1"></a>                                               <span class="at">min.feature =</span> min.feature)</span>
<span id="cb19-49"><a href="step-by-step-scrna-seq-pipeline.html#cb19-49" tabindex="-1"></a>          sc_object.temp[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_object.temp, <span class="at">pattern =</span> mt.pattern)</span>
<span id="cb19-50"><a href="step-by-step-scrna-seq-pipeline.html#cb19-50" tabindex="-1"></a>          input.data.list <span class="ot">&lt;-</span> <span class="fu">c</span>(input.data.list, sc_object.temp)}</span>
<span id="cb19-51"><a href="step-by-step-scrna-seq-pipeline.html#cb19-51" tabindex="-1"></a>      </span>
<span id="cb19-52"><a href="step-by-step-scrna-seq-pipeline.html#cb19-52" tabindex="-1"></a>  }<span class="cf">else</span>{</span>
<span id="cb19-53"><a href="step-by-step-scrna-seq-pipeline.html#cb19-53" tabindex="-1"></a>      </span>
<span id="cb19-54"><a href="step-by-step-scrna-seq-pipeline.html#cb19-54" tabindex="-1"></a>          sc_data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(input.data.dirs)</span>
<span id="cb19-55"><a href="step-by-step-scrna-seq-pipeline.html#cb19-55" tabindex="-1"></a>          sc_object <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> sc_data,</span>
<span id="cb19-56"><a href="step-by-step-scrna-seq-pipeline.html#cb19-56" tabindex="-1"></a>                                          <span class="at">project =</span> project.names,</span>
<span id="cb19-57"><a href="step-by-step-scrna-seq-pipeline.html#cb19-57" tabindex="-1"></a>                                          <span class="at">min.cells =</span> min.cells,</span>
<span id="cb19-58"><a href="step-by-step-scrna-seq-pipeline.html#cb19-58" tabindex="-1"></a>                                          <span class="at">min.feature =</span> min.feature)</span>
<span id="cb19-59"><a href="step-by-step-scrna-seq-pipeline.html#cb19-59" tabindex="-1"></a>          sc_object[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_object, <span class="at">pattern =</span> mt.pattern) </span>
<span id="cb19-60"><a href="step-by-step-scrna-seq-pipeline.html#cb19-60" tabindex="-1"></a>      </span>
<span id="cb19-61"><a href="step-by-step-scrna-seq-pipeline.html#cb19-61" tabindex="-1"></a>  }</span>
<span id="cb19-62"><a href="step-by-step-scrna-seq-pipeline.html#cb19-62" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb19-63"><a href="step-by-step-scrna-seq-pipeline.html#cb19-63" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">&#39;Please input data generated by the cellranger-count software, or a Seurat object, or a gene expression matrix. HemaScopeR does not support other formats of input data.&#39;</span>)</span>
<span id="cb19-64"><a href="step-by-step-scrna-seq-pipeline.html#cb19-64" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Save the variables after executing each step, if necessary.</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="step-by-step-scrna-seq-pipeline.html#cb20-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb20-2"><a href="step-by-step-scrna-seq-pipeline.html#cb20-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb20-3"><a href="step-by-step-scrna-seq-pipeline.html#cb20-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb20-4"><a href="step-by-step-scrna-seq-pipeline.html#cb20-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb20-5"><a href="step-by-step-scrna-seq-pipeline.html#cb20-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb20-6"><a href="step-by-step-scrna-seq-pipeline.html#cb20-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb20-7"><a href="step-by-step-scrna-seq-pipeline.html#cb20-7" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="step-2.-quality-control" class="section level2 hasAnchor" number="4.3">
<h2><span class="header-section-number">4.3</span> Step 2. Quality Control<a href="step-by-step-scrna-seq-pipeline.html#step-2.-quality-control" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this step, the following quality control steps will be performed:</p>
<ol style="list-style-type: decimal">
<li><p>Normalize data using the LogNormalize method.</p></li>
<li><p>Find variable features using the vst method.</p></li>
<li><p>Scale data using the identified variable features and specified variables to regress out.</p></li>
<li><p>Perform principal component analysis (PCA) on the scaled data.</p></li>
<li><p>Find K nearest neighbors based on PCA dimensions.</p></li>
<li><p>Perform clustering analysis based on the found neighbors.</p></li>
<li><p>Optionally, remove doublets using doubletFinder.</p></li>
<li><p>Optionally, integrate multiple datasets by removing batch effects.</p></li>
</ol>
<div id="function-arguments" class="section level3 hasAnchor" number="4.3.1">
<h3><span class="header-section-number">4.3.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>nFeature_RNA.limit</code>: The cutoff of the minimum number of detected genes in each cell.</li>
<li><code>percent.mt.limit</code>: The cutoff of the maximum percentage of mitochondria genes in each cell.</li>
<li><code>scale.factor</code>: The scale factor for the ‘data’ slot in the seurat object.</li>
<li><code>nfeatures</code>: The number of selected highly variable features for down stream analysis.</li>
<li><code>ndims</code>: The number of principle components in PCA.</li>
<li><code>vars.to.regress</code>: Variables to regress out (previously latent.vars in RegressOut). For example, nUMI, or percent.mito. (ScaleData in Seurat)</li>
<li><code>PCs</code>: Which dimensions to use as input features.(RunTSNE and RunUMAP in Seurat)</li>
<li><code>resolution</code>: Value of the resolution parameter, use a value above (below) 1.0 if you want to obtain a larger (smaller) number of communities. (FindClusters in Seurat)</li>
<li><code>n.neighbors</code>: Defines k for the k-nearest neighbor algorithm. (FindNeighbors in Seurat)</li>
<li><code>percentage</code>: Assuming ‘percentage’ doublet formation rate - tailor for your dataset. The default value is 0.05.</li>
<li><code>doublerFinderwraper.PCs</code> Which dimensions to use as input features for doubletFinder.</li>
<li><code>doublerFinderwraper.pN</code>: The percentage of real-artifical data for doubletFinder.</li>
<li><code>doublerFinderwraper.pK</code>: The pK parameter controls the doublet cell detection by determining the number of nearest neighbors and influencing the calculation of pANN scores and the final cell classification results. Adjusting the pK value allows optimization of the doublet cell detection process based on specific data and analysis requirements.</li>
</ul>
</div>
<div id="codes-for-running-step2" class="section level3 hasAnchor" number="4.3.2">
<h3><span class="header-section-number">4.3.2</span> codes for running step2<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the results of quality control.</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="step-by-step-scrna-seq-pipeline.html#cb21-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step2. Quality control.&#39;</span>)</span>
<span id="cb21-2"><a href="step-by-step-scrna-seq-pipeline.html#cb21-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step2.Quality_control/&#39;</span>))) {</span>
<span id="cb21-3"><a href="step-by-step-scrna-seq-pipeline.html#cb21-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step2.Quality_control/&#39;</span>))</span>
<span id="cb21-4"><a href="step-by-step-scrna-seq-pipeline.html#cb21-4" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Set the parameters for quality control.</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="step-by-step-scrna-seq-pipeline.html#cb22-1" tabindex="-1"></a><span class="co"># quality control</span></span>
<span id="cb22-2"><a href="step-by-step-scrna-seq-pipeline.html#cb22-2" tabindex="-1"></a>nFeature_RNA.limit <span class="ot">=</span> <span class="dv">200</span></span>
<span id="cb22-3"><a href="step-by-step-scrna-seq-pipeline.html#cb22-3" tabindex="-1"></a>percent.mt.limit <span class="ot">=</span> <span class="dv">20</span></span>
<span id="cb22-4"><a href="step-by-step-scrna-seq-pipeline.html#cb22-4" tabindex="-1"></a></span>
<span id="cb22-5"><a href="step-by-step-scrna-seq-pipeline.html#cb22-5" tabindex="-1"></a><span class="co"># preprocessing</span></span>
<span id="cb22-6"><a href="step-by-step-scrna-seq-pipeline.html#cb22-6" tabindex="-1"></a>nfeatures <span class="ot">=</span> <span class="dv">3000</span></span>
<span id="cb22-7"><a href="step-by-step-scrna-seq-pipeline.html#cb22-7" tabindex="-1"></a>scale.factor <span class="ot">=</span> <span class="dv">10000</span></span>
<span id="cb22-8"><a href="step-by-step-scrna-seq-pipeline.html#cb22-8" tabindex="-1"></a>ndims <span class="ot">=</span> <span class="dv">50</span></span>
<span id="cb22-9"><a href="step-by-step-scrna-seq-pipeline.html#cb22-9" tabindex="-1"></a>vars.to.regress <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb22-10"><a href="step-by-step-scrna-seq-pipeline.html#cb22-10" tabindex="-1"></a>PCs <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">35</span></span>
<span id="cb22-11"><a href="step-by-step-scrna-seq-pipeline.html#cb22-11" tabindex="-1"></a>resolution <span class="ot">=</span> <span class="fl">0.4</span></span>
<span id="cb22-12"><a href="step-by-step-scrna-seq-pipeline.html#cb22-12" tabindex="-1"></a>n.neighbors <span class="ot">=</span> <span class="dv">50</span></span>
<span id="cb22-13"><a href="step-by-step-scrna-seq-pipeline.html#cb22-13" tabindex="-1"></a></span>
<span id="cb22-14"><a href="step-by-step-scrna-seq-pipeline.html#cb22-14" tabindex="-1"></a><span class="co"># removing doublets</span></span>
<span id="cb22-15"><a href="step-by-step-scrna-seq-pipeline.html#cb22-15" tabindex="-1"></a>Step2_Quality_Control.RemoveDoublets <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb22-16"><a href="step-by-step-scrna-seq-pipeline.html#cb22-16" tabindex="-1"></a>doublet.percentage <span class="ot">=</span> <span class="fl">0.04</span></span>
<span id="cb22-17"><a href="step-by-step-scrna-seq-pipeline.html#cb22-17" tabindex="-1"></a>doublerFinderwraper.PCs <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span></span>
<span id="cb22-18"><a href="step-by-step-scrna-seq-pipeline.html#cb22-18" tabindex="-1"></a>doublerFinderwraper.pN <span class="ot">=</span> <span class="fl">0.25</span></span>
<span id="cb22-19"><a href="step-by-step-scrna-seq-pipeline.html#cb22-19" tabindex="-1"></a>doublerFinderwraper.pK <span class="ot">=</span> <span class="fl">0.1</span></span>
<span id="cb22-20"><a href="step-by-step-scrna-seq-pipeline.html#cb22-20" tabindex="-1"></a></span>
<span id="cb22-21"><a href="step-by-step-scrna-seq-pipeline.html#cb22-21" tabindex="-1"></a><span class="co"># removing batch effect</span></span>
<span id="cb22-22"><a href="step-by-step-scrna-seq-pipeline.html#cb22-22" tabindex="-1"></a>Step2_Quality_Control.RemoveBatches <span class="ot">=</span> <span class="cn">TRUE</span></span></code></pre></div>
<ul>
<li>Run the quality control process.</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="step-by-step-scrna-seq-pipeline.html#cb23-1" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb23-2"><a href="step-by-step-scrna-seq-pipeline.html#cb23-2" tabindex="-1"></a><span class="co"># preprocess and quality control for multiple scRNA-Seq data sets</span></span>
<span id="cb23-3"><a href="step-by-step-scrna-seq-pipeline.html#cb23-3" tabindex="-1"></a>sc_object <span class="ot">&lt;-</span> <span class="fu">QC_multiple_scRNASeq</span>(<span class="at">seuratObjects =</span> input.data.list,</span>
<span id="cb23-4"><a href="step-by-step-scrna-seq-pipeline.html#cb23-4" tabindex="-1"></a>                                  <span class="at">datasetID =</span> project.names,</span>
<span id="cb23-5"><a href="step-by-step-scrna-seq-pipeline.html#cb23-5" tabindex="-1"></a>                                  <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step2.Quality_control/&#39;</span>),</span>
<span id="cb23-6"><a href="step-by-step-scrna-seq-pipeline.html#cb23-6" tabindex="-1"></a>                                  <span class="at">Step2_Quality_Control.RemoveBatches =</span> Step2_Quality_Control.RemoveBatches,</span>
<span id="cb23-7"><a href="step-by-step-scrna-seq-pipeline.html#cb23-7" tabindex="-1"></a>                                  <span class="at">Step2_Quality_Control.RemoveDoublets =</span> Step2_Quality_Control.RemoveDoublets,</span>
<span id="cb23-8"><a href="step-by-step-scrna-seq-pipeline.html#cb23-8" tabindex="-1"></a>                                  <span class="at">nFeature_RNA.limit =</span> nFeature_RNA.limit,</span>
<span id="cb23-9"><a href="step-by-step-scrna-seq-pipeline.html#cb23-9" tabindex="-1"></a>                                  <span class="at">percent.mt.limit =</span> percent.mt.limit,</span>
<span id="cb23-10"><a href="step-by-step-scrna-seq-pipeline.html#cb23-10" tabindex="-1"></a>                                  <span class="at">scale.factor =</span> scale.factor,</span>
<span id="cb23-11"><a href="step-by-step-scrna-seq-pipeline.html#cb23-11" tabindex="-1"></a>                                  <span class="at">nfeatures =</span> nfeatures,</span>
<span id="cb23-12"><a href="step-by-step-scrna-seq-pipeline.html#cb23-12" tabindex="-1"></a>                                  <span class="at">ndims =</span> ndims,</span>
<span id="cb23-13"><a href="step-by-step-scrna-seq-pipeline.html#cb23-13" tabindex="-1"></a>                                  <span class="at">vars.to.regress =</span> vars.to.regress,</span>
<span id="cb23-14"><a href="step-by-step-scrna-seq-pipeline.html#cb23-14" tabindex="-1"></a>                                  <span class="at">PCs =</span> PCs,</span>
<span id="cb23-15"><a href="step-by-step-scrna-seq-pipeline.html#cb23-15" tabindex="-1"></a>                                  <span class="at">resolution =</span> resolution,</span>
<span id="cb23-16"><a href="step-by-step-scrna-seq-pipeline.html#cb23-16" tabindex="-1"></a>                                  <span class="at">n.neighbors =</span> n.neighbors,</span>
<span id="cb23-17"><a href="step-by-step-scrna-seq-pipeline.html#cb23-17" tabindex="-1"></a>                                  <span class="at">percentage =</span> doublet.percentage,</span>
<span id="cb23-18"><a href="step-by-step-scrna-seq-pipeline.html#cb23-18" tabindex="-1"></a>                                  <span class="at">doublerFinderwraper.PCs =</span> doublerFinderwraper.PCs,</span>
<span id="cb23-19"><a href="step-by-step-scrna-seq-pipeline.html#cb23-19" tabindex="-1"></a>                                  <span class="at">doublerFinderwraper.pN =</span> doublerFinderwraper.pN,</span>
<span id="cb23-20"><a href="step-by-step-scrna-seq-pipeline.html#cb23-20" tabindex="-1"></a>                                  <span class="at">doublerFinderwraper.pK =</span> doublerFinderwraper.pK</span>
<span id="cb23-21"><a href="step-by-step-scrna-seq-pipeline.html#cb23-21" tabindex="-1"></a>                                  )</span>
<span id="cb23-22"><a href="step-by-step-scrna-seq-pipeline.html#cb23-22" tabindex="-1"></a></span>
<span id="cb23-23"><a href="step-by-step-scrna-seq-pipeline.html#cb23-23" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb23-24"><a href="step-by-step-scrna-seq-pipeline.html#cb23-24" tabindex="-1"></a><span class="co"># preprocess and quality control for single scRNA-Seq data set</span></span>
<span id="cb23-25"><a href="step-by-step-scrna-seq-pipeline.html#cb23-25" tabindex="-1"></a>sc_object <span class="ot">&lt;-</span> <span class="fu">QC_single_scRNASeq</span>(<span class="at">sc_object =</span> sc_object,</span>
<span id="cb23-26"><a href="step-by-step-scrna-seq-pipeline.html#cb23-26" tabindex="-1"></a>                                <span class="at">datasetID =</span> project.names,</span>
<span id="cb23-27"><a href="step-by-step-scrna-seq-pipeline.html#cb23-27" tabindex="-1"></a>                                <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step2.Quality_control/&#39;</span>),</span>
<span id="cb23-28"><a href="step-by-step-scrna-seq-pipeline.html#cb23-28" tabindex="-1"></a>                                <span class="at">Step2_Quality_Control.RemoveDoublets =</span> Step2_Quality_Control.RemoveDoublets,</span>
<span id="cb23-29"><a href="step-by-step-scrna-seq-pipeline.html#cb23-29" tabindex="-1"></a>                                <span class="at">nFeature_RNA.limit =</span> nFeature_RNA.limit,</span>
<span id="cb23-30"><a href="step-by-step-scrna-seq-pipeline.html#cb23-30" tabindex="-1"></a>                                <span class="at">percent.mt.limit =</span> percent.mt.limit,</span>
<span id="cb23-31"><a href="step-by-step-scrna-seq-pipeline.html#cb23-31" tabindex="-1"></a>                                <span class="at">scale.factor =</span> scale.factor,</span>
<span id="cb23-32"><a href="step-by-step-scrna-seq-pipeline.html#cb23-32" tabindex="-1"></a>                                <span class="at">nfeatures =</span> nfeatures,</span>
<span id="cb23-33"><a href="step-by-step-scrna-seq-pipeline.html#cb23-33" tabindex="-1"></a>                                <span class="at">vars.to.regress =</span> vars.to.regress,</span>
<span id="cb23-34"><a href="step-by-step-scrna-seq-pipeline.html#cb23-34" tabindex="-1"></a>                                <span class="at">ndims =</span> ndims,</span>
<span id="cb23-35"><a href="step-by-step-scrna-seq-pipeline.html#cb23-35" tabindex="-1"></a>                                <span class="at">PCs =</span> PCs,</span>
<span id="cb23-36"><a href="step-by-step-scrna-seq-pipeline.html#cb23-36" tabindex="-1"></a>                                <span class="at">resolution =</span> resolution,</span>
<span id="cb23-37"><a href="step-by-step-scrna-seq-pipeline.html#cb23-37" tabindex="-1"></a>                                <span class="at">n.neighbors =</span> n.neighbors,</span>
<span id="cb23-38"><a href="step-by-step-scrna-seq-pipeline.html#cb23-38" tabindex="-1"></a>                                <span class="at">percentage =</span> doublet.percentage,</span>
<span id="cb23-39"><a href="step-by-step-scrna-seq-pipeline.html#cb23-39" tabindex="-1"></a>                                <span class="at">doublerFinderwraper.PCs =</span> doublerFinderwraper.PCs,</span>
<span id="cb23-40"><a href="step-by-step-scrna-seq-pipeline.html#cb23-40" tabindex="-1"></a>                                <span class="at">doublerFinderwraper.pN =</span> doublerFinderwraper.pN,</span>
<span id="cb23-41"><a href="step-by-step-scrna-seq-pipeline.html#cb23-41" tabindex="-1"></a>                                <span class="at">doublerFinderwraper.pK =</span> doublerFinderwraper.pK)</span>
<span id="cb23-42"><a href="step-by-step-scrna-seq-pipeline.html#cb23-42" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="step-by-step-scrna-seq-pipeline.html#cb24-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb24-2"><a href="step-by-step-scrna-seq-pipeline.html#cb24-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb24-3"><a href="step-by-step-scrna-seq-pipeline.html#cb24-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb24-4"><a href="step-by-step-scrna-seq-pipeline.html#cb24-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb24-5"><a href="step-by-step-scrna-seq-pipeline.html#cb24-5" tabindex="-1"></a>var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb24-6"><a href="step-by-step-scrna-seq-pipeline.html#cb24-6" tabindex="-1"></a><span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb24-7"><a href="step-by-step-scrna-seq-pipeline.html#cb24-7" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="outputs" class="section level3 hasAnchor" number="4.3.3">
<h3><span class="header-section-number">4.3.3</span> Outputs<a href="step-by-step-scrna-seq-pipeline.html#outputs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-15"></span>
<img src="sc_img/step2/sc_object_nFeature_nCount_percentMt.png" alt="Violin plots showing the nFeature, nCount and percent.mt for each sample" width="800px" />
<p class="caption">
Figure 4.1: Violin plots showing the nFeature, nCount and percent.mt for each sample
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-16"></span>
<img src="sc_img/step2/sc_object_FeatureScatters.png" alt="Figures showing the correlation between nFeature and nCount, as well as between nCount and percent.mt" width="600px" />
<p class="caption">
Figure 4.2: Figures showing the correlation between nFeature and nCount, as well as between nCount and percent.mt
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-17"></span>
<img src="sc_img/step2/sc_object_VariableFeaturePlot.png" alt="Figures showing the variable features used for downstream analysis" width="400px" />
<p class="caption">
Figure 4.3: Figures showing the variable features used for downstream analysis
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-18"></span>
<img src="sc_img/step2/sc_object_ElbowPlot.png" alt="ElbowPlot showing suitable number of PCs used for further analysis" width="400px" />
<p class="caption">
Figure 4.4: ElbowPlot showing suitable number of PCs used for further analysis
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-20"></span>
<img src="sc_img/step2/sc_object_findDoublets.png" alt="UMAP plot showing doublets found by DoubletFinder" width="400px" />
<p class="caption">
Figure 4.5: UMAP plot showing doublets found by DoubletFinder
</p>
</div>
</div>
</div>
<div id="step-3.-clustering" class="section level2 hasAnchor" number="4.4">
<h2><span class="header-section-number">4.4</span> Step 3. Clustering<a href="step-by-step-scrna-seq-pipeline.html#step-3.-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li>Create a folder for saving the results of Louvain clustering.</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="step-by-step-scrna-seq-pipeline.html#cb25-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step3. Clustering.&#39;</span>)</span>
<span id="cb25-2"><a href="step-by-step-scrna-seq-pipeline.html#cb25-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step3.Clustering/&#39;</span>))) {</span>
<span id="cb25-3"><a href="step-by-step-scrna-seq-pipeline.html#cb25-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step3.Clustering/&#39;</span>))</span>
<span id="cb25-4"><a href="step-by-step-scrna-seq-pipeline.html#cb25-4" tabindex="-1"></a>}   </span></code></pre></div>
<ul>
<li>Set the parameters for clustering.</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="step-by-step-scrna-seq-pipeline.html#cb26-1" tabindex="-1"></a>PCs <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">35</span></span>
<span id="cb26-2"><a href="step-by-step-scrna-seq-pipeline.html#cb26-2" tabindex="-1"></a>resolution <span class="ot">=</span> <span class="fl">0.4</span></span>
<span id="cb26-3"><a href="step-by-step-scrna-seq-pipeline.html#cb26-3" tabindex="-1"></a>n.neighbors <span class="ot">=</span> <span class="dv">50</span></span></code></pre></div>
<ul>
<li>Run Louvian clustering.</li>
</ul>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="step-by-step-scrna-seq-pipeline.html#cb27-1" tabindex="-1"></a><span class="cf">if</span>( (<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">&amp;</span> Step2_Quality_Control.RemoveBatches ){graph.name <span class="ot">&lt;-</span> <span class="st">&#39;integrated_snn&#39;</span>}<span class="cf">else</span>{graph.name <span class="ot">&lt;-</span> <span class="st">&#39;RNA_snn&#39;</span>}</span>
<span id="cb27-2"><a href="step-by-step-scrna-seq-pipeline.html#cb27-2" tabindex="-1"></a>sc_object <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(sc_object, <span class="at">dims =</span> PCs, <span class="at">k.param =</span> n.neighbors, <span class="at">force.recalc =</span> <span class="cn">TRUE</span>)</span>
<span id="cb27-3"><a href="step-by-step-scrna-seq-pipeline.html#cb27-3" tabindex="-1"></a>sc_object <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(sc_object, <span class="at">resolution =</span> resolution, <span class="at">graph.name =</span> graph.name)</span>
<span id="cb27-4"><a href="step-by-step-scrna-seq-pipeline.html#cb27-4" tabindex="-1"></a>sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>seurat_clusters <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">as.numeric</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>seurat_clusters))</span>
<span id="cb27-5"><a href="step-by-step-scrna-seq-pipeline.html#cb27-5" tabindex="-1"></a></span>
<span id="cb27-6"><a href="step-by-step-scrna-seq-pipeline.html#cb27-6" tabindex="-1"></a><span class="co"># plot clustering</span></span>
<span id="cb27-7"><a href="step-by-step-scrna-seq-pipeline.html#cb27-7" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step3.Clustering/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;tsne_cluster.pdf&#39;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb27-8"><a href="step-by-step-scrna-seq-pipeline.html#cb27-8" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;tsne&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb27-9"><a href="step-by-step-scrna-seq-pipeline.html#cb27-9" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb27-10"><a href="step-by-step-scrna-seq-pipeline.html#cb27-10" tabindex="-1"></a></span>
<span id="cb27-11"><a href="step-by-step-scrna-seq-pipeline.html#cb27-11" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step3.Clustering/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;umap_cluster.pdf&#39;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb27-12"><a href="step-by-step-scrna-seq-pipeline.html#cb27-12" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb27-13"><a href="step-by-step-scrna-seq-pipeline.html#cb27-13" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb27-14"><a href="step-by-step-scrna-seq-pipeline.html#cb27-14" tabindex="-1"></a></span>
<span id="cb27-15"><a href="step-by-step-scrna-seq-pipeline.html#cb27-15" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step3.Clustering/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;tsne_cluster.png&#39;</span>), <span class="at">width =</span> <span class="dv">600</span>, <span class="at">height =</span> <span class="dv">600</span>)</span>
<span id="cb27-16"><a href="step-by-step-scrna-seq-pipeline.html#cb27-16" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;tsne&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb27-17"><a href="step-by-step-scrna-seq-pipeline.html#cb27-17" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb27-18"><a href="step-by-step-scrna-seq-pipeline.html#cb27-18" tabindex="-1"></a></span>
<span id="cb27-19"><a href="step-by-step-scrna-seq-pipeline.html#cb27-19" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step3.Clustering/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;umap_cluster.png&#39;</span>), <span class="at">width =</span> <span class="dv">600</span>, <span class="at">height =</span> <span class="dv">600</span>)</span>
<span id="cb27-20"><a href="step-by-step-scrna-seq-pipeline.html#cb27-20" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb27-21"><a href="step-by-step-scrna-seq-pipeline.html#cb27-21" tabindex="-1"></a><span class="fu">dev.off</span>()    </span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="step-by-step-scrna-seq-pipeline.html#cb28-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb28-2"><a href="step-by-step-scrna-seq-pipeline.html#cb28-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb28-3"><a href="step-by-step-scrna-seq-pipeline.html#cb28-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb28-4"><a href="step-by-step-scrna-seq-pipeline.html#cb28-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb28-5"><a href="step-by-step-scrna-seq-pipeline.html#cb28-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb28-6"><a href="step-by-step-scrna-seq-pipeline.html#cb28-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb28-7"><a href="step-by-step-scrna-seq-pipeline.html#cb28-7" tabindex="-1"></a>}  </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-25"></span>
<img src="sc_img/step3/sc_object%20umap_cluster.png" alt="UMAP plot showing clustering results" width="400px" />
<p class="caption">
Figure 4.6: UMAP plot showing clustering results
</p>
</div>
</div>
<div id="step-4.-identify-cell-types" class="section level2 hasAnchor" number="4.5">
<h2><span class="header-section-number">4.5</span> Step 4. Identify Cell Types<a href="step-by-step-scrna-seq-pipeline.html#step-4.-identify-cell-types" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this step, users can predict the cell types of hematopoietic cells by implementing two approaches (Scmap and Seurat) through abcCellmap packages. Cells are labeled by 43 different RNA clusters according to unsupervised clustering of single-cell transcriptional profiles, and also labeled by 32 immunophenotypic cell types. In addition, users can use Copykat to measure copy number variation (CNV) and determine the ploidy of each cell.</p>
<div id="codes-for-running-abccellmap" class="section level3 hasAnchor" number="4.5.1">
<h3><span class="header-section-number">4.5.1</span> codes for running abcCellmap<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-abccellmap" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the results of cell type identification.</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="step-by-step-scrna-seq-pipeline.html#cb29-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step4. Identify cell types automatically.&#39;</span>)</span>
<span id="cb29-2"><a href="step-by-step-scrna-seq-pipeline.html#cb29-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step4.Identify_Cell_Types/&#39;</span>))) {</span>
<span id="cb29-3"><a href="step-by-step-scrna-seq-pipeline.html#cb29-3" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step4.Identify_Cell_Types/&#39;</span>))</span>
<span id="cb29-4"><a href="step-by-step-scrna-seq-pipeline.html#cb29-4" tabindex="-1"></a>} </span></code></pre></div>
<ul>
<li>Set the path for the database.</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="step-by-step-scrna-seq-pipeline.html#cb30-1" tabindex="-1"></a>databasePath <span class="ot">=</span> <span class="st">&quot;~/HemaScopeR/database/&quot;</span></span></code></pre></div>
<ul>
<li>Set the parameters for cell type identification.</li>
</ul>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="step-by-step-scrna-seq-pipeline.html#cb31-1" tabindex="-1"></a>Step4_Use_Which_Labels <span class="ot">=</span> <span class="st">&#39;clustering&#39;</span></span>
<span id="cb31-2"><a href="step-by-step-scrna-seq-pipeline.html#cb31-2" tabindex="-1"></a>Step4_Cluster_Labels <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb31-3"><a href="step-by-step-scrna-seq-pipeline.html#cb31-3" tabindex="-1"></a>Step4_Changed_Labels <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb31-4"><a href="step-by-step-scrna-seq-pipeline.html#cb31-4" tabindex="-1"></a>Org <span class="ot">=</span> <span class="st">&#39;hsa&#39;</span></span>
<span id="cb31-5"><a href="step-by-step-scrna-seq-pipeline.html#cb31-5" tabindex="-1"></a>ncores <span class="ot">=</span> <span class="dv">10</span></span></code></pre></div>
<ul>
<li>Run the cell type identification process.</li>
</ul>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="step-by-step-scrna-seq-pipeline.html#cb32-1" tabindex="-1"></a>sc_object <span class="ot">&lt;-</span> <span class="fu">run_cell_annotation</span>(<span class="at">object =</span> sc_object, </span>
<span id="cb32-2"><a href="step-by-step-scrna-seq-pipeline.html#cb32-2" tabindex="-1"></a>                               <span class="at">assay =</span> <span class="st">&#39;RNA&#39;</span>, </span>
<span id="cb32-3"><a href="step-by-step-scrna-seq-pipeline.html#cb32-3" tabindex="-1"></a>                               <span class="at">species =</span> Org,</span>
<span id="cb32-4"><a href="step-by-step-scrna-seq-pipeline.html#cb32-4" tabindex="-1"></a>                               <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step4.Identify_Cell_Types/&#39;</span>))</span>
<span id="cb32-5"><a href="step-by-step-scrna-seq-pipeline.html#cb32-5" tabindex="-1"></a></span>
<span id="cb32-6"><a href="step-by-step-scrna-seq-pipeline.html#cb32-6" tabindex="-1"></a><span class="cf">if</span>(Org <span class="sc">==</span> <span class="st">&#39;hsa&#39;</span>){</span>
<span id="cb32-7"><a href="step-by-step-scrna-seq-pipeline.html#cb32-7" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">paste0</span>(databasePath,<span class="st">&quot;/HematoMap.reference.rdata&quot;</span>)) <span class="co">#the data can be downloaded via the link https://cloud.tsinghua.edu.cn/d/759fd04333274d3f9946</span></span>
<span id="cb32-8"><a href="step-by-step-scrna-seq-pipeline.html#cb32-8" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">intersect</span>(<span class="fu">rownames</span>(HematoMap.reference), <span class="fu">rownames</span>(sc_object))) <span class="sc">&lt;</span> <span class="dv">1000</span>){</span>
<span id="cb32-9"><a href="step-by-step-scrna-seq-pipeline.html#cb32-9" tabindex="-1"></a>      HematoMap.reference <span class="ot">&lt;-</span> <span class="fu">RenameGenesSeurat</span>(<span class="at">obj =</span> HematoMap.reference,</span>
<span id="cb32-10"><a href="step-by-step-scrna-seq-pipeline.html#cb32-10" tabindex="-1"></a>                                               <span class="at">newnames =</span> <span class="fu">toupper</span>(<span class="fu">rownames</span>(HematoMap.reference)),</span>
<span id="cb32-11"><a href="step-by-step-scrna-seq-pipeline.html#cb32-11" tabindex="-1"></a>                                               <span class="at">gene.use =</span> <span class="fu">rownames</span>(HematoMap.reference), </span>
<span id="cb32-12"><a href="step-by-step-scrna-seq-pipeline.html#cb32-12" tabindex="-1"></a>                                               <span class="at">de.assay =</span> <span class="st">&quot;RNA&quot;</span>,</span>
<span id="cb32-13"><a href="step-by-step-scrna-seq-pipeline.html#cb32-13" tabindex="-1"></a>                                               <span class="at">lassays =</span> <span class="st">&quot;RNA&quot;</span>)</span>
<span id="cb32-14"><a href="step-by-step-scrna-seq-pipeline.html#cb32-14" tabindex="-1"></a>  }</span>
<span id="cb32-15"><a href="step-by-step-scrna-seq-pipeline.html#cb32-15" tabindex="-1"></a>  </span>
<span id="cb32-16"><a href="step-by-step-scrna-seq-pipeline.html#cb32-16" tabindex="-1"></a><span class="cf">if</span>(sc_object<span class="sc">@</span>active.assay <span class="sc">==</span> <span class="st">&#39;integrated&#39;</span>){</span>
<span id="cb32-17"><a href="step-by-step-scrna-seq-pipeline.html#cb32-17" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object) <span class="ot">&lt;-</span> <span class="st">&#39;RNA&#39;</span></span>
<span id="cb32-18"><a href="step-by-step-scrna-seq-pipeline.html#cb32-18" tabindex="-1"></a>    sc_object <span class="ot">&lt;-</span> <span class="fu">mapDataToRef</span>(<span class="at">ref_object =</span> HematoMap.reference,</span>
<span id="cb32-19"><a href="step-by-step-scrna-seq-pipeline.html#cb32-19" tabindex="-1"></a>                              <span class="at">ref_labels =</span> HematoMap.reference<span class="sc">@</span>meta.data<span class="sc">$</span>CellType,</span>
<span id="cb32-20"><a href="step-by-step-scrna-seq-pipeline.html#cb32-20" tabindex="-1"></a>                              <span class="at">query_object =</span> sc_object,</span>
<span id="cb32-21"><a href="step-by-step-scrna-seq-pipeline.html#cb32-21" tabindex="-1"></a>                              <span class="at">PCs =</span> PCs,</span>
<span id="cb32-22"><a href="step-by-step-scrna-seq-pipeline.html#cb32-22" tabindex="-1"></a>                              <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step4.Identify_Cell_Types/&#39;</span>))</span>
<span id="cb32-23"><a href="step-by-step-scrna-seq-pipeline.html#cb32-23" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object) <span class="ot">&lt;-</span> <span class="st">&#39;integrated&#39;</span></span>
<span id="cb32-24"><a href="step-by-step-scrna-seq-pipeline.html#cb32-24" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb32-25"><a href="step-by-step-scrna-seq-pipeline.html#cb32-25" tabindex="-1"></a>    sc_object <span class="ot">&lt;-</span> <span class="fu">mapDataToRef</span>(<span class="at">ref_object =</span> HematoMap.reference,</span>
<span id="cb32-26"><a href="step-by-step-scrna-seq-pipeline.html#cb32-26" tabindex="-1"></a>                              <span class="at">ref_labels =</span> HematoMap.reference<span class="sc">@</span>meta.data<span class="sc">$</span>CellType,</span>
<span id="cb32-27"><a href="step-by-step-scrna-seq-pipeline.html#cb32-27" tabindex="-1"></a>                              <span class="at">query_object =</span> sc_object,</span>
<span id="cb32-28"><a href="step-by-step-scrna-seq-pipeline.html#cb32-28" tabindex="-1"></a>                              <span class="at">PCs =</span> PCs,</span>
<span id="cb32-29"><a href="step-by-step-scrna-seq-pipeline.html#cb32-29" tabindex="-1"></a>                              <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step4.Identify_Cell_Types/&#39;</span>))</span>
<span id="cb32-30"><a href="step-by-step-scrna-seq-pipeline.html#cb32-30" tabindex="-1"></a>}</span>
<span id="cb32-31"><a href="step-by-step-scrna-seq-pipeline.html#cb32-31" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Set the cell labels.</li>
</ul>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="step-by-step-scrna-seq-pipeline.html#cb33-1" tabindex="-1"></a><span class="co"># set the cell labels</span></span>
<span id="cb33-2"><a href="step-by-step-scrna-seq-pipeline.html#cb33-2" tabindex="-1"></a><span class="cf">if</span>(Step4_Use_Which_Labels <span class="sc">==</span> <span class="st">&#39;clustering&#39;</span>){</span>
<span id="cb33-3"><a href="step-by-step-scrna-seq-pipeline.html#cb33-3" tabindex="-1"></a>  sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>seurat_clusters</span>
<span id="cb33-4"><a href="step-by-step-scrna-seq-pipeline.html#cb33-4" tabindex="-1"></a>  <span class="fu">Idents</span>(sc_object) <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels</span>
<span id="cb33-5"><a href="step-by-step-scrna-seq-pipeline.html#cb33-5" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step4_Use_Which_Labels <span class="sc">==</span> <span class="st">&#39;abcCellmap.1&#39;</span>){</span>
<span id="cb33-6"><a href="step-by-step-scrna-seq-pipeline.html#cb33-6" tabindex="-1"></a>  sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>Seurat.RNACluster</span>
<span id="cb33-7"><a href="step-by-step-scrna-seq-pipeline.html#cb33-7" tabindex="-1"></a>  <span class="fu">Idents</span>(sc_object) <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels</span>
<span id="cb33-8"><a href="step-by-step-scrna-seq-pipeline.html#cb33-8" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step4_Use_Which_Labels <span class="sc">==</span> <span class="st">&#39;abcCellmap.2&#39;</span>){</span>
<span id="cb33-9"><a href="step-by-step-scrna-seq-pipeline.html#cb33-9" tabindex="-1"></a>  sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>scmap.RNACluster</span>
<span id="cb33-10"><a href="step-by-step-scrna-seq-pipeline.html#cb33-10" tabindex="-1"></a>  <span class="fu">Idents</span>(sc_object) <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels</span>
<span id="cb33-11"><a href="step-by-step-scrna-seq-pipeline.html#cb33-11" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step4_Use_Which_Labels <span class="sc">==</span> <span class="st">&#39;abcCellmap.3&#39;</span>){</span>
<span id="cb33-12"><a href="step-by-step-scrna-seq-pipeline.html#cb33-12" tabindex="-1"></a>  sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>Seurat.Immunophenotype</span>
<span id="cb33-13"><a href="step-by-step-scrna-seq-pipeline.html#cb33-13" tabindex="-1"></a>  <span class="fu">Idents</span>(sc_object) <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels          </span>
<span id="cb33-14"><a href="step-by-step-scrna-seq-pipeline.html#cb33-14" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step4_Use_Which_Labels <span class="sc">==</span> <span class="st">&#39;abcCellmap.4&#39;</span>){</span>
<span id="cb33-15"><a href="step-by-step-scrna-seq-pipeline.html#cb33-15" tabindex="-1"></a>  sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>scmap.Immunophenotype</span>
<span id="cb33-16"><a href="step-by-step-scrna-seq-pipeline.html#cb33-16" tabindex="-1"></a>  <span class="fu">Idents</span>(sc_object) <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels          </span>
<span id="cb33-17"><a href="step-by-step-scrna-seq-pipeline.html#cb33-17" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step4_Use_Which_Labels <span class="sc">==</span> <span class="st">&#39;HematoMap&#39;</span>){</span>
<span id="cb33-18"><a href="step-by-step-scrna-seq-pipeline.html#cb33-18" tabindex="-1"></a>  <span class="cf">if</span>(Org <span class="sc">==</span> <span class="st">&#39;hsa&#39;</span>){</span>
<span id="cb33-19"><a href="step-by-step-scrna-seq-pipeline.html#cb33-19" tabindex="-1"></a>    sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>predicted.id</span>
<span id="cb33-20"><a href="step-by-step-scrna-seq-pipeline.html#cb33-20" tabindex="-1"></a>    <span class="fu">Idents</span>(sc_object) <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels</span>
<span id="cb33-21"><a href="step-by-step-scrna-seq-pipeline.html#cb33-21" tabindex="-1"></a>  }<span class="cf">else</span>{<span class="fu">print</span>(<span class="st">&quot;&#39;HematoMap&#39; is only applicable to human data (&#39;Org&#39; = &#39;hsa&#39;).&quot;</span>)}    </span>
<span id="cb33-22"><a href="step-by-step-scrna-seq-pipeline.html#cb33-22" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Step4_Use_Which_Labels <span class="sc">==</span> <span class="st">&#39;changeLabels&#39;</span>){</span>
<span id="cb33-23"><a href="step-by-step-scrna-seq-pipeline.html#cb33-23" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(Step4_Cluster_Labels) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(Step4_Changed_Labels) <span class="sc">&amp;&amp;</span> <span class="fu">length</span>(Step4_Cluster_Labels) <span class="sc">==</span> <span class="fu">length</span>(Step4_Changed_Labels)){</span>
<span id="cb33-24"><a href="step-by-step-scrna-seq-pipeline.html#cb33-24" tabindex="-1"></a>   sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>seurat_clusters,</span>
<span id="cb33-25"><a href="step-by-step-scrna-seq-pipeline.html#cb33-25" tabindex="-1"></a>                                                       <span class="at">from =</span> <span class="fu">as.character</span>(Step4_Cluster_Labels),</span>
<span id="cb33-26"><a href="step-by-step-scrna-seq-pipeline.html#cb33-26" tabindex="-1"></a>                                                       <span class="at">to =</span> <span class="fu">as.character</span>(Step4_Changed_Labels),</span>
<span id="cb33-27"><a href="step-by-step-scrna-seq-pipeline.html#cb33-27" tabindex="-1"></a>                                                       <span class="at">warn_missing =</span> <span class="cn">FALSE</span>)</span>
<span id="cb33-28"><a href="step-by-step-scrna-seq-pipeline.html#cb33-28" tabindex="-1"></a>   <span class="fu">Idents</span>(sc_object) <span class="ot">&lt;-</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels</span>
<span id="cb33-29"><a href="step-by-step-scrna-seq-pipeline.html#cb33-29" tabindex="-1"></a>  }<span class="cf">else</span>{</span>
<span id="cb33-30"><a href="step-by-step-scrna-seq-pipeline.html#cb33-30" tabindex="-1"></a>   <span class="fu">print</span>(<span class="st">&quot;Please input the &#39;Step4_Cluster_Labels&#39; parameter as Seurat clustering labels, and the &#39;Step4_Changed_Labels&#39; parameter as new labels. Please note that these two parameters should be of equal length.&quot;</span>)</span>
<span id="cb33-31"><a href="step-by-step-scrna-seq-pipeline.html#cb33-31" tabindex="-1"></a>  }</span>
<span id="cb33-32"><a href="step-by-step-scrna-seq-pipeline.html#cb33-32" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb33-33"><a href="step-by-step-scrna-seq-pipeline.html#cb33-33" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">&#39;Please set the &quot;Step4_Use_Which_Labels&quot; parameter as &quot;clustering&quot;, &quot;abcCellmap.1&quot;, &quot;abcCellmap.2&quot;, &quot;HematoMap&quot; or &quot;changeLabels&quot;.&#39;</span>)</span>
<span id="cb33-34"><a href="step-by-step-scrna-seq-pipeline.html#cb33-34" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="step-by-step-scrna-seq-pipeline.html#cb34-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb34-2"><a href="step-by-step-scrna-seq-pipeline.html#cb34-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb34-3"><a href="step-by-step-scrna-seq-pipeline.html#cb34-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb34-4"><a href="step-by-step-scrna-seq-pipeline.html#cb34-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb34-5"><a href="step-by-step-scrna-seq-pipeline.html#cb34-5" tabindex="-1"></a>var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb34-6"><a href="step-by-step-scrna-seq-pipeline.html#cb34-6" tabindex="-1"></a><span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb34-7"><a href="step-by-step-scrna-seq-pipeline.html#cb34-7" tabindex="-1"></a>}  </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-32"></span>
<img src="sc_img/step4/predicted.id.png" alt="UMAP plots showing cell type annotation results" width="400px" />
<p class="caption">
Figure 4.7: UMAP plots showing cell type annotation results
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-33"></span>
<img src="sc_img/step4/predicted.scmap.Immunophenotype.png" alt="Immunophenotype and RNACluster label predicted by scmap" width="50%" /><img src="sc_img/step4/predicted.scmap.RNACluster.png" alt="Immunophenotype and RNACluster label predicted by scmap" width="50%" />
<p class="caption">
Figure 4.8: Immunophenotype and RNACluster label predicted by scmap
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-34"></span>
<img src="sc_img/step4/predicted.Seurat.Immunophenotype.png" alt="Immunophenotype and RNACluster label predicted by Seurat" width="50%" /><img src="sc_img/step4/predicted.Seurat.RNACluster.png" alt="Immunophenotype and RNACluster label predicted by Seurat" width="50%" />
<p class="caption">
Figure 4.9: Immunophenotype and RNACluster label predicted by Seurat
</p>
</div>
</div>
<div id="codes-for-running-the-cnv-analysis" class="section level3 hasAnchor" number="4.5.2">
<h3><span class="header-section-number">4.5.2</span> codes for running the CNV analysis<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-the-cnv-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="step-by-step-scrna-seq-pipeline.html#cb35-1" tabindex="-1"></a><span class="fu">sc_CNV</span>(<span class="at">sc_object=</span>sc_object,</span>
<span id="cb35-2"><a href="step-by-step-scrna-seq-pipeline.html#cb35-2" tabindex="-1"></a>       <span class="at">save_path=</span><span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step4.Identify_Cell_Types/&#39;</span>),</span>
<span id="cb35-3"><a href="step-by-step-scrna-seq-pipeline.html#cb35-3" tabindex="-1"></a>       <span class="at">assay =</span> <span class="st">&#39;RNA&#39;</span>,</span>
<span id="cb35-4"><a href="step-by-step-scrna-seq-pipeline.html#cb35-4" tabindex="-1"></a>       <span class="at">LOW.DR =</span> <span class="fl">0.05</span>, <span class="co">#refer to the Copykat documentation for detailed explanations of the parameters</span></span>
<span id="cb35-5"><a href="step-by-step-scrna-seq-pipeline.html#cb35-5" tabindex="-1"></a>       <span class="at">UP.DR =</span> <span class="fl">0.1</span>,</span>
<span id="cb35-6"><a href="step-by-step-scrna-seq-pipeline.html#cb35-6" tabindex="-1"></a>       <span class="at">win.size =</span> <span class="dv">25</span>,</span>
<span id="cb35-7"><a href="step-by-step-scrna-seq-pipeline.html#cb35-7" tabindex="-1"></a>       <span class="at">distance =</span> <span class="st">&quot;euclidean&quot;</span>,</span>
<span id="cb35-8"><a href="step-by-step-scrna-seq-pipeline.html#cb35-8" tabindex="-1"></a>       <span class="at">genome =</span> <span class="cn">NULL</span>,</span>
<span id="cb35-9"><a href="step-by-step-scrna-seq-pipeline.html#cb35-9" tabindex="-1"></a>       <span class="at">n.cores =</span> ncores, <span class="co">#note: this step will take a long time, using more ncores could shorten the running time</span></span>
<span id="cb35-10"><a href="step-by-step-scrna-seq-pipeline.html#cb35-10" tabindex="-1"></a>       <span class="at">species =</span> Org)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-36"></span>
<img src="sc_img/step4/copykat_heatmap.jpeg" alt="copykat heatmap" width="800px" />
<p class="caption">
Figure 4.10: copykat heatmap
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-37"></span>
<img src="sc_img/step4/CNV_state.png" alt="UMAP plot showing CNV state predicted by copykat" width="400px" />
<p class="caption">
Figure 4.11: UMAP plot showing CNV state predicted by copykat
</p>
</div>
</div>
</div>
<div id="step-5.-visualization" class="section level2 hasAnchor" number="4.6">
<h2><span class="header-section-number">4.6</span> Step 5. Visualization<a href="step-by-step-scrna-seq-pipeline.html#step-5.-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this step, users are allowed to gain the statistical results about the numbers and proportions of cell groups, and also use three dimensional reduction methods (TSNE, UMAP, phateR) to visualize the results.</p>
<div id="codes-for-peforming-three-dimensional-reduction-methods" class="section level3 hasAnchor" number="4.6.1">
<h3><span class="header-section-number">4.6.1</span> codes for peforming three dimensional reduction methods<a href="step-by-step-scrna-seq-pipeline.html#codes-for-peforming-three-dimensional-reduction-methods" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the visualization results.</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="step-by-step-scrna-seq-pipeline.html#cb36-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step5. Visualization.&#39;</span>)</span>
<span id="cb36-2"><a href="step-by-step-scrna-seq-pipeline.html#cb36-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step5.Visualization/&#39;</span>))) {</span>
<span id="cb36-3"><a href="step-by-step-scrna-seq-pipeline.html#cb36-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step5.Visualization/&#39;</span>))</span>
<span id="cb36-4"><a href="step-by-step-scrna-seq-pipeline.html#cb36-4" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Perform visualization using UMAP and TSNE.</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="step-by-step-scrna-seq-pipeline.html#cb37-1" tabindex="-1"></a><span class="co"># plot cell types</span></span>
<span id="cb37-2"><a href="step-by-step-scrna-seq-pipeline.html#cb37-2" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step5.Visualization/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;tsne cell types.pdf&#39;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb37-3"><a href="step-by-step-scrna-seq-pipeline.html#cb37-3" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;tsne&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;ident&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb37-4"><a href="step-by-step-scrna-seq-pipeline.html#cb37-4" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb37-5"><a href="step-by-step-scrna-seq-pipeline.html#cb37-5" tabindex="-1"></a></span>
<span id="cb37-6"><a href="step-by-step-scrna-seq-pipeline.html#cb37-6" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step5.Visualization/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;umap cell types.pdf&#39;</span>), <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb37-7"><a href="step-by-step-scrna-seq-pipeline.html#cb37-7" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;ident&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb37-8"><a href="step-by-step-scrna-seq-pipeline.html#cb37-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb37-9"><a href="step-by-step-scrna-seq-pipeline.html#cb37-9" tabindex="-1"></a></span>
<span id="cb37-10"><a href="step-by-step-scrna-seq-pipeline.html#cb37-10" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step5.Visualization/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;tsne cell types.png&#39;</span>), <span class="at">width =</span> <span class="dv">600</span>, <span class="at">height =</span> <span class="dv">600</span>)</span>
<span id="cb37-11"><a href="step-by-step-scrna-seq-pipeline.html#cb37-11" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;tsne&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;ident&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb37-12"><a href="step-by-step-scrna-seq-pipeline.html#cb37-12" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb37-13"><a href="step-by-step-scrna-seq-pipeline.html#cb37-13" tabindex="-1"></a></span>
<span id="cb37-14"><a href="step-by-step-scrna-seq-pipeline.html#cb37-14" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step5.Visualization/&#39;</span>), <span class="st">&#39;/sc_object &#39;</span>,<span class="st">&#39;umap cell types.png&#39;</span>), <span class="at">width =</span> <span class="dv">600</span>, <span class="at">height =</span> <span class="dv">600</span>)</span>
<span id="cb37-15"><a href="step-by-step-scrna-seq-pipeline.html#cb37-15" tabindex="-1"></a> <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;ident&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb37-16"><a href="step-by-step-scrna-seq-pipeline.html#cb37-16" tabindex="-1"></a><span class="fu">dev.off</span>()    </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-40"></span>
<img src="sc_img/step5/sc_object%20tsne%20cell%20types.png" alt="UMAP and TSNE visualization" width="50%" /><img src="sc_img/step5/sc_object%20umap%20cell%20types.png" alt="UMAP and TSNE visualization" width="50%" />
<p class="caption">
Figure 4.12: UMAP and TSNE visualization
</p>
</div>
<ul>
<li>Set the parameters for phateR.</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="step-by-step-scrna-seq-pipeline.html#cb38-1" tabindex="-1"></a>phate.knn <span class="ot">=</span> <span class="dv">50</span> <span class="co">#The number of nearest neighbors to consider in the phateR algorithm. Default 50.</span></span>
<span id="cb38-2"><a href="step-by-step-scrna-seq-pipeline.html#cb38-2" tabindex="-1"></a>phate.npca <span class="ot">=</span> <span class="dv">20</span> <span class="co">#The number of principal components to use in the phateR algorithm. Default 20.</span></span>
<span id="cb38-3"><a href="step-by-step-scrna-seq-pipeline.html#cb38-3" tabindex="-1"></a>phate.t <span class="ot">=</span> <span class="dv">10</span> <span class="co">#The t-value for the phateR algorithm, which controls the level of exploration. Default 10.</span></span>
<span id="cb38-4"><a href="step-by-step-scrna-seq-pipeline.html#cb38-4" tabindex="-1"></a>phate.ndim <span class="ot">=</span> <span class="dv">2</span> <span class="co">#The number of dimensions for the output embedding in the phateR algorithm. Default 2.</span></span></code></pre></div>
<ul>
<li>Run phateR for dimensional reduction and visualization.</li>
</ul>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="step-by-step-scrna-seq-pipeline.html#cb39-1" tabindex="-1"></a><span class="co"># run phateR</span></span>
<span id="cb39-2"><a href="step-by-step-scrna-seq-pipeline.html#cb39-2" tabindex="-1"></a><span class="cf">if</span>( (<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">&amp;</span> Step2_Quality_Control.RemoveBatches ){</span>
<span id="cb39-3"><a href="step-by-step-scrna-seq-pipeline.html#cb39-3" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object) <span class="ot">&lt;-</span> <span class="st">&#39;integrated&#39;</span></span>
<span id="cb39-4"><a href="step-by-step-scrna-seq-pipeline.html#cb39-4" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb39-5"><a href="step-by-step-scrna-seq-pipeline.html#cb39-5" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object) <span class="ot">&lt;-</span> <span class="st">&#39;RNA&#39;</span>}</span>
<span id="cb39-6"><a href="step-by-step-scrna-seq-pipeline.html#cb39-6" tabindex="-1"></a></span>
<span id="cb39-7"><a href="step-by-step-scrna-seq-pipeline.html#cb39-7" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(pythonPath)){</span>
<span id="cb39-8"><a href="step-by-step-scrna-seq-pipeline.html#cb39-8" tabindex="-1"></a>  <span class="fu">run_phateR</span>(<span class="at">sc_object =</span> sc_object,</span>
<span id="cb39-9"><a href="step-by-step-scrna-seq-pipeline.html#cb39-9" tabindex="-1"></a>             <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step5.Visualization/&#39;</span>),</span>
<span id="cb39-10"><a href="step-by-step-scrna-seq-pipeline.html#cb39-10" tabindex="-1"></a>             <span class="at">pythonPath =</span> pythonPath,</span>
<span id="cb39-11"><a href="step-by-step-scrna-seq-pipeline.html#cb39-11" tabindex="-1"></a>             <span class="at">phate.knn =</span> phate.knn,</span>
<span id="cb39-12"><a href="step-by-step-scrna-seq-pipeline.html#cb39-12" tabindex="-1"></a>             <span class="at">phate.npca =</span> phate.npca,</span>
<span id="cb39-13"><a href="step-by-step-scrna-seq-pipeline.html#cb39-13" tabindex="-1"></a>             <span class="at">phate.t =</span> phate.t,</span>
<span id="cb39-14"><a href="step-by-step-scrna-seq-pipeline.html#cb39-14" tabindex="-1"></a>             <span class="at">phate.ndim =</span> phate.ndim)     </span>
<span id="cb39-15"><a href="step-by-step-scrna-seq-pipeline.html#cb39-15" tabindex="-1"></a>}</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-43"></span>
<img src="sc_img/step5/phateR.png" alt="phateR result" width="400px" />
<p class="caption">
Figure 4.13: phateR result
</p>
</div>
<ul>
<li>Run HematoMap to visualize cluster tree plots.</li>
</ul>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="step-by-step-scrna-seq-pipeline.html#cb40-1" tabindex="-1"></a><span class="fu">run_HematoMap</span>(<span class="at">sc_object =</span> sc_object,</span>
<span id="cb40-2"><a href="step-by-step-scrna-seq-pipeline.html#cb40-2" tabindex="-1"></a>              <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir,<span class="st">&#39;/Step5.Visualization/&#39;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-45"></span>
<img src="sc_img/step5/HematoMap.png" alt="HematoMap result" width="400px" />
<p class="caption">
Figure 4.14: HematoMap result
</p>
</div>
</div>
<div id="codes-for-calculating-the-proportions" class="section level3 hasAnchor" number="4.6.2">
<h3><span class="header-section-number">4.6.2</span> codes for calculating the proportions<a href="step-by-step-scrna-seq-pipeline.html#codes-for-calculating-the-proportions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>The statistical results for the numbers and proportions of cell groups.</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="step-by-step-scrna-seq-pipeline.html#cb41-1" tabindex="-1"></a><span class="co"># statistical results</span></span>
<span id="cb41-2"><a href="step-by-step-scrna-seq-pipeline.html#cb41-2" tabindex="-1"></a>cells_labels <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(<span class="fu">rownames</span>(sc_object<span class="sc">@</span>meta.data), <span class="fu">as.character</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels)))</span>
<span id="cb41-3"><a href="step-by-step-scrna-seq-pipeline.html#cb41-3" tabindex="-1"></a><span class="fu">colnames</span>(cells_labels) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;cell_id&#39;</span>, <span class="st">&#39;cluster_id&#39;</span>)</span>
<span id="cb41-4"><a href="step-by-step-scrna-seq-pipeline.html#cb41-4" tabindex="-1"></a>cluster_counts <span class="ot">&lt;-</span> cells_labels <span class="sc">%&gt;%</span></span>
<span id="cb41-5"><a href="step-by-step-scrna-seq-pipeline.html#cb41-5" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster_id) <span class="sc">%&gt;%</span></span>
<span id="cb41-6"><a href="step-by-step-scrna-seq-pipeline.html#cb41-6" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">count =</span> <span class="fu">n</span>())</span>
<span id="cb41-7"><a href="step-by-step-scrna-seq-pipeline.html#cb41-7" tabindex="-1"></a>total_cells <span class="ot">&lt;-</span> <span class="fu">nrow</span>(cells_labels)</span>
<span id="cb41-8"><a href="step-by-step-scrna-seq-pipeline.html#cb41-8" tabindex="-1"></a>cluster_counts <span class="ot">&lt;-</span> cluster_counts <span class="sc">%&gt;%</span></span>
<span id="cb41-9"><a href="step-by-step-scrna-seq-pipeline.html#cb41-9" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">proportion =</span> count <span class="sc">/</span> total_cells)</span>
<span id="cb41-10"><a href="step-by-step-scrna-seq-pipeline.html#cb41-10" tabindex="-1"></a></span>
<span id="cb41-11"><a href="step-by-step-scrna-seq-pipeline.html#cb41-11" tabindex="-1"></a>cluster_counts <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(cluster_counts)</span>
<span id="cb41-12"><a href="step-by-step-scrna-seq-pipeline.html#cb41-12" tabindex="-1"></a></span>
<span id="cb41-13"><a href="step-by-step-scrna-seq-pipeline.html#cb41-13" tabindex="-1"></a>cluster_counts<span class="sc">$</span>percentages <span class="ot">&lt;-</span> scales<span class="sc">::</span><span class="fu">percent</span>(cluster_counts<span class="sc">$</span>proportion, <span class="at">accuracy =</span> <span class="fl">0.1</span>)</span>
<span id="cb41-14"><a href="step-by-step-scrna-seq-pipeline.html#cb41-14" tabindex="-1"></a>cluster_counts <span class="ot">&lt;-</span> cluster_counts[,<span class="sc">-</span><span class="fu">which</span>(<span class="fu">colnames</span>(cluster_counts)<span class="sc">==</span><span class="st">&#39;proportion&#39;</span>)]</span>
<span id="cb41-15"><a href="step-by-step-scrna-seq-pipeline.html#cb41-15" tabindex="-1"></a></span>
<span id="cb41-16"><a href="step-by-step-scrna-seq-pipeline.html#cb41-16" tabindex="-1"></a>cluster_counts<span class="sc">$</span>cluster_id_count_percentages <span class="ot">&lt;-</span> <span class="fu">paste</span>(cluster_counts<span class="sc">$</span>cluster_id, <span class="st">&quot; (&quot;</span>, cluster_counts<span class="sc">$</span>count, <span class="st">&#39; cells; &#39;</span>, cluster_counts<span class="sc">$</span>percentages, <span class="st">&quot;)&quot;</span>, <span class="at">sep=</span><span class="st">&#39;&#39;</span>)</span>
<span id="cb41-17"><a href="step-by-step-scrna-seq-pipeline.html#cb41-17" tabindex="-1"></a></span>
<span id="cb41-18"><a href="step-by-step-scrna-seq-pipeline.html#cb41-18" tabindex="-1"></a>cluster_counts <span class="ot">&lt;-</span> cluster_counts[<span class="fu">order</span>(cluster_counts<span class="sc">$</span>count, <span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb41-19"><a href="step-by-step-scrna-seq-pipeline.html#cb41-19" tabindex="-1"></a></span>
<span id="cb41-20"><a href="step-by-step-scrna-seq-pipeline.html#cb41-20" tabindex="-1"></a>cluster_counts <span class="ot">&lt;-</span> <span class="fu">rbind</span>(cluster_counts, <span class="fu">c</span>(<span class="st">&#39;Total&#39;</span>, <span class="fu">sum</span>(cluster_counts<span class="sc">$</span>count), <span class="st">&#39;100%&#39;</span>, <span class="st">&#39;all cells&#39;</span>))</span>
<span id="cb41-21"><a href="step-by-step-scrna-seq-pipeline.html#cb41-21" tabindex="-1"></a></span>
<span id="cb41-22"><a href="step-by-step-scrna-seq-pipeline.html#cb41-22" tabindex="-1"></a>sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>cluster_id_count_percentages <span class="ot">&lt;-</span> <span class="fu">mapvalues</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels, </span>
<span id="cb41-23"><a href="step-by-step-scrna-seq-pipeline.html#cb41-23" tabindex="-1"></a>                                                              <span class="at">from=</span>cluster_counts<span class="sc">$</span>cluster_id, </span>
<span id="cb41-24"><a href="step-by-step-scrna-seq-pipeline.html#cb41-24" tabindex="-1"></a>                                                              <span class="at">to=</span>cluster_counts<span class="sc">$</span>cluster_id_count_percentages,</span>
<span id="cb41-25"><a href="step-by-step-scrna-seq-pipeline.html#cb41-25" tabindex="-1"></a>                                                              <span class="at">warn_missing=</span><span class="cn">FALSE</span>)</span>
<span id="cb41-26"><a href="step-by-step-scrna-seq-pipeline.html#cb41-26" tabindex="-1"></a></span>
<span id="cb41-27"><a href="step-by-step-scrna-seq-pipeline.html#cb41-27" tabindex="-1"></a><span class="fu">colnames</span>(sc_object<span class="sc">@</span>meta.data)[<span class="fu">which</span>(<span class="fu">colnames</span>(sc_object<span class="sc">@</span>meta.data) <span class="sc">==</span> <span class="st">&#39;cluster_id_count_percentages&#39;</span>)] <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&#39;Total &#39;</span>, <span class="fu">nrow</span>(sc_object<span class="sc">@</span>meta.data), <span class="st">&#39; cells&#39;</span>, <span class="at">sep=</span><span class="st">&#39;&#39;</span>)</span>
<span id="cb41-28"><a href="step-by-step-scrna-seq-pipeline.html#cb41-28" tabindex="-1"></a></span>
<span id="cb41-29"><a href="step-by-step-scrna-seq-pipeline.html#cb41-29" tabindex="-1"></a>cluster_counts <span class="ot">&lt;-</span> cluster_counts[,<span class="sc">-</span><span class="fu">which</span>(<span class="fu">colnames</span>(cluster_counts)<span class="sc">==</span><span class="st">&#39;cluster_id_count_percentages&#39;</span>)]</span>
<span id="cb41-30"><a href="step-by-step-scrna-seq-pipeline.html#cb41-30" tabindex="-1"></a></span>
<span id="cb41-31"><a href="step-by-step-scrna-seq-pipeline.html#cb41-31" tabindex="-1"></a><span class="fu">colnames</span>(cluster_counts) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;Cell types&#39;</span>, <span class="st">&#39;Cell counts&#39;</span>, <span class="st">&#39;Percentages&#39;</span>)</span>
<span id="cb41-32"><a href="step-by-step-scrna-seq-pipeline.html#cb41-32" tabindex="-1"></a></span>
<span id="cb41-33"><a href="step-by-step-scrna-seq-pipeline.html#cb41-33" tabindex="-1"></a><span class="co"># names(colorvector) &lt;- mapvalues(names(colorvector), </span></span>
<span id="cb41-34"><a href="step-by-step-scrna-seq-pipeline.html#cb41-34" tabindex="-1"></a><span class="co">#                                 from=cluster_counts$cluster_id, </span></span>
<span id="cb41-35"><a href="step-by-step-scrna-seq-pipeline.html#cb41-35" tabindex="-1"></a><span class="co">#                                 to=cluster_counts$cluster_id_count_percentages,</span></span>
<span id="cb41-36"><a href="step-by-step-scrna-seq-pipeline.html#cb41-36" tabindex="-1"></a><span class="co">#                                 warn_missing=FALSE)</span></span>
<span id="cb41-37"><a href="step-by-step-scrna-seq-pipeline.html#cb41-37" tabindex="-1"></a></span>
<span id="cb41-38"><a href="step-by-step-scrna-seq-pipeline.html#cb41-38" tabindex="-1"></a><span class="fu">write.csv</span>(cluster_counts, <span class="at">file=</span><span class="fu">paste</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step5.Visualization/&#39;</span>), <span class="st">&#39;/cell types_cell counts_percentages.csv&#39;</span>, <span class="at">sep=</span><span class="st">&#39;&#39;</span>), <span class="at">quote=</span><span class="cn">FALSE</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<ul>
<li>The UMAP visualization.</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="step-by-step-scrna-seq-pipeline.html#cb42-1" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">paste</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step5.Visualization&#39;</span>), <span class="st">&#39;/cell types_cell counts_percentages_umap.pdf&#39;</span>, <span class="at">sep=</span><span class="st">&#39;&#39;</span>), <span class="at">width =</span> <span class="dv">14</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb42-2"><a href="step-by-step-scrna-seq-pipeline.html#cb42-2" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">DimPlot</span>(sc_object, </span>
<span id="cb42-3"><a href="step-by-step-scrna-seq-pipeline.html#cb42-3" tabindex="-1"></a>                <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb42-4"><a href="step-by-step-scrna-seq-pipeline.html#cb42-4" tabindex="-1"></a>                <span class="at">group.by =</span> <span class="fu">paste</span>(<span class="st">&#39;Total &#39;</span>, <span class="fu">nrow</span>(sc_object<span class="sc">@</span>meta.data), <span class="st">&#39; cells&#39;</span>, <span class="at">sep=</span><span class="st">&#39;&#39;</span>), </span>
<span id="cb42-5"><a href="step-by-step-scrna-seq-pipeline.html#cb42-5" tabindex="-1"></a>                <span class="at">label =</span> <span class="cn">FALSE</span>, </span>
<span id="cb42-6"><a href="step-by-step-scrna-seq-pipeline.html#cb42-6" tabindex="-1"></a>                <span class="at">pt.size =</span> <span class="fl">0.1</span>, </span>
<span id="cb42-7"><a href="step-by-step-scrna-seq-pipeline.html#cb42-7" tabindex="-1"></a>                <span class="at">raster =</span> <span class="cn">FALSE</span>))</span>
<span id="cb42-8"><a href="step-by-step-scrna-seq-pipeline.html#cb42-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="step-by-step-scrna-seq-pipeline.html#cb43-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb43-2"><a href="step-by-step-scrna-seq-pipeline.html#cb43-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb43-3"><a href="step-by-step-scrna-seq-pipeline.html#cb43-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb43-4"><a href="step-by-step-scrna-seq-pipeline.html#cb43-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb43-5"><a href="step-by-step-scrna-seq-pipeline.html#cb43-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb43-6"><a href="step-by-step-scrna-seq-pipeline.html#cb43-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb43-7"><a href="step-by-step-scrna-seq-pipeline.html#cb43-7" tabindex="-1"></a>}  </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-49"></span>
<img src="sc_img/step5/cell%20types_cell%20counts_percentages_umap.jpg" alt="UMAP plot showing cell type and corresponding proportion" width="800px" />
<p class="caption">
Figure 4.15: UMAP plot showing cell type and corresponding proportion
</p>
</div>
</div>
</div>
<div id="step-6.-find-degs" class="section level2 hasAnchor" number="4.7">
<h2><span class="header-section-number">4.7</span> Step 6. Find DEGs<a href="step-by-step-scrna-seq-pipeline.html#step-6.-find-degs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this step, users can find DEGs (differentially expressed genes) across different cell type group using FindAllMarkers, use GPTCelltype to predict cell label, perform GO and KEGG enrichment analysis, and perform subnetwork analysis for each cell type group.</p>
<div id="codes-for-finding-degs" class="section level3 hasAnchor" number="4.7.1">
<h3><span class="header-section-number">4.7.1</span> codes for finding DEGs<a href="step-by-step-scrna-seq-pipeline.html#codes-for-finding-degs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Set the parameters for identifying differentially expressed genes.</li>
</ul>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="step-by-step-scrna-seq-pipeline.html#cb44-1" tabindex="-1"></a>min.pct <span class="ot">=</span> <span class="fl">0.25</span></span>
<span id="cb44-2"><a href="step-by-step-scrna-seq-pipeline.html#cb44-2" tabindex="-1"></a>logfc.threshold <span class="ot">=</span> <span class="fl">0.25</span></span></code></pre></div>
<ul>
<li>Create a folder for the DEGs analysis.</li>
</ul>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="step-by-step-scrna-seq-pipeline.html#cb45-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step6. Find DEGs.&#39;</span>)</span>
<span id="cb45-2"><a href="step-by-step-scrna-seq-pipeline.html#cb45-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/&#39;</span>))) {</span>
<span id="cb45-3"><a href="step-by-step-scrna-seq-pipeline.html#cb45-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/&#39;</span>))</span>
<span id="cb45-4"><a href="step-by-step-scrna-seq-pipeline.html#cb45-4" tabindex="-1"></a>}   </span></code></pre></div>
<ul>
<li>Identify DEGs using Wilcoxon Rank-Sum Test.</li>
</ul>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="step-by-step-scrna-seq-pipeline.html#cb46-1" tabindex="-1"></a>sc_object.markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(sc_object, <span class="at">only.pos =</span> <span class="cn">TRUE</span>, <span class="at">min.pct =</span> min.pct, <span class="at">logfc.threshold =</span> logfc.threshold)</span>
<span id="cb46-2"><a href="step-by-step-scrna-seq-pipeline.html#cb46-2" tabindex="-1"></a><span class="fu">write.csv</span>(sc_object.markers,</span>
<span id="cb46-3"><a href="step-by-step-scrna-seq-pipeline.html#cb46-3" tabindex="-1"></a>          <span class="at">file =</span> <span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/&#39;</span>),<span class="st">&#39;sc_object.markerGenes.csv&#39;</span>),</span>
<span id="cb46-4"><a href="step-by-step-scrna-seq-pipeline.html#cb46-4" tabindex="-1"></a>          <span class="at">quote=</span><span class="cn">FALSE</span>)</span>
<span id="cb46-5"><a href="step-by-step-scrna-seq-pipeline.html#cb46-5" tabindex="-1"></a></span>
<span id="cb46-6"><a href="step-by-step-scrna-seq-pipeline.html#cb46-6" tabindex="-1"></a><span class="co"># visualization</span></span>
<span id="cb46-7"><a href="step-by-step-scrna-seq-pipeline.html#cb46-7" tabindex="-1"></a>sc_object.markers.top5 <span class="ot">&lt;-</span> sc_object.markers <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">5</span>, <span class="at">wt =</span> avg_log2FC)</span>
<span id="cb46-8"><a href="step-by-step-scrna-seq-pipeline.html#cb46-8" tabindex="-1"></a></span>
<span id="cb46-9"><a href="step-by-step-scrna-seq-pipeline.html#cb46-9" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/&#39;</span>), <span class="st">&#39;sc_object_markerGenesTop5.pdf&#39;</span>), </span>
<span id="cb46-10"><a href="step-by-step-scrna-seq-pipeline.html#cb46-10" tabindex="-1"></a>    <span class="at">width =</span> <span class="fl">0.5</span><span class="sc">*</span><span class="fu">length</span>(<span class="fu">unique</span>(sc_object.markers.top5<span class="sc">$</span>gene)), </span>
<span id="cb46-11"><a href="step-by-step-scrna-seq-pipeline.html#cb46-11" tabindex="-1"></a>    <span class="at">height =</span> <span class="fl">0.5</span><span class="sc">*</span><span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">Idents</span>(sc_object))))</span>
<span id="cb46-12"><a href="step-by-step-scrna-seq-pipeline.html#cb46-12" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">DotPlot</span>(sc_object,</span>
<span id="cb46-13"><a href="step-by-step-scrna-seq-pipeline.html#cb46-13" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">unique</span>(sc_object.markers.top5<span class="sc">$</span>gene),</span>
<span id="cb46-14"><a href="step-by-step-scrna-seq-pipeline.html#cb46-14" tabindex="-1"></a>            <span class="at">cols=</span><span class="fu">c</span>(<span class="st">&quot;lightgrey&quot;</span>,<span class="st">&#39;red&#39;</span>))<span class="sc">+</span><span class="fu">theme</span>(<span class="at">axis.text.x =</span><span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)))</span>
<span id="cb46-15"><a href="step-by-step-scrna-seq-pipeline.html#cb46-15" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb46-16"><a href="step-by-step-scrna-seq-pipeline.html#cb46-16" tabindex="-1"></a></span>
<span id="cb46-17"><a href="step-by-step-scrna-seq-pipeline.html#cb46-17" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/&#39;</span>), <span class="st">&#39;sc_object_markerGenesTop5.png&#39;</span>), </span>
<span id="cb46-18"><a href="step-by-step-scrna-seq-pipeline.html#cb46-18" tabindex="-1"></a>    <span class="at">width =</span> <span class="dv">20</span><span class="sc">*</span><span class="fu">length</span>(<span class="fu">unique</span>(sc_object.markers.top5<span class="sc">$</span>gene)), </span>
<span id="cb46-19"><a href="step-by-step-scrna-seq-pipeline.html#cb46-19" tabindex="-1"></a>    <span class="at">height =</span> <span class="dv">30</span><span class="sc">*</span><span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">Idents</span>(sc_object))))</span>
<span id="cb46-20"><a href="step-by-step-scrna-seq-pipeline.html#cb46-20" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">DotPlot</span>(sc_object,</span>
<span id="cb46-21"><a href="step-by-step-scrna-seq-pipeline.html#cb46-21" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">unique</span>(sc_object.markers.top5<span class="sc">$</span>gene),</span>
<span id="cb46-22"><a href="step-by-step-scrna-seq-pipeline.html#cb46-22" tabindex="-1"></a>            <span class="at">cols=</span><span class="fu">c</span>(<span class="st">&quot;lightgrey&quot;</span>,<span class="st">&#39;red&#39;</span>))<span class="sc">+</span><span class="fu">theme</span>(<span class="at">axis.text.x =</span><span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)))</span>
<span id="cb46-23"><a href="step-by-step-scrna-seq-pipeline.html#cb46-23" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-53"></span>
<img src="sc_img/step6/sc_object_markerGenesTop5.png" alt="Dotplot showing marker genes of each cell type group" width="600px" />
<p class="caption">
Figure 4.16: Dotplot showing marker genes of each cell type group
</p>
</div>
</div>
<div id="codes-for-using-gptcelltype" class="section level3 hasAnchor" number="4.7.2">
<h3><span class="header-section-number">4.7.2</span> codes for using GPTCelltype<a href="step-by-step-scrna-seq-pipeline.html#codes-for-using-gptcelltype" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Set the parameters for GPTCelltype.</li>
</ul>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="step-by-step-scrna-seq-pipeline.html#cb47-1" tabindex="-1"></a>your_openai_API_key <span class="ot">=</span> <span class="st">&#39;&#39;</span></span>
<span id="cb47-2"><a href="step-by-step-scrna-seq-pipeline.html#cb47-2" tabindex="-1"></a>tissuename <span class="ot">=</span> <span class="st">&#39;human bone marrow&#39;</span></span>
<span id="cb47-3"><a href="step-by-step-scrna-seq-pipeline.html#cb47-3" tabindex="-1"></a>gptmodel <span class="ot">=</span> <span class="st">&#39;gpt-3.5&#39;</span></span></code></pre></div>
<ul>
<li>Use GPTCelltype to assist cell type annotation.</li>
</ul>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="step-by-step-scrna-seq-pipeline.html#cb48-1" tabindex="-1"></a><span class="fu">GPT_annotation</span>( <span class="at">marker.genes =</span> sc_object.markers,</span>
<span id="cb48-2"><a href="step-by-step-scrna-seq-pipeline.html#cb48-2" tabindex="-1"></a>                <span class="at">your_openai_API_key =</span> your_openai_API_key,</span>
<span id="cb48-3"><a href="step-by-step-scrna-seq-pipeline.html#cb48-3" tabindex="-1"></a>                <span class="at">tissuename =</span> tissuename,</span>
<span id="cb48-4"><a href="step-by-step-scrna-seq-pipeline.html#cb48-4" tabindex="-1"></a>                <span class="at">gptmodel =</span> gptmodel,</span>
<span id="cb48-5"><a href="step-by-step-scrna-seq-pipeline.html#cb48-5" tabindex="-1"></a>                <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/&#39;</span>))</span></code></pre></div>
</div>
<div id="perform-go-and-kegg-enrichment." class="section level3 hasAnchor" number="4.7.3">
<h3><span class="header-section-number">4.7.3</span> Perform GO and KEGG enrichment.<a href="step-by-step-scrna-seq-pipeline.html#perform-go-and-kegg-enrichment." class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="step-by-step-scrna-seq-pipeline.html#cb49-1" tabindex="-1"></a><span class="co"># GO enrichment</span></span>
<span id="cb49-2"><a href="step-by-step-scrna-seq-pipeline.html#cb49-2" tabindex="-1"></a><span class="cf">if</span>(Org<span class="sc">==</span><span class="st">&#39;mmu&#39;</span>){</span>
<span id="cb49-3"><a href="step-by-step-scrna-seq-pipeline.html#cb49-3" tabindex="-1"></a>  OrgDb <span class="ot">&lt;-</span> <span class="st">&#39;org.Mm.eg.db&#39;</span></span>
<span id="cb49-4"><a href="step-by-step-scrna-seq-pipeline.html#cb49-4" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Org<span class="sc">==</span><span class="st">&#39;hsa&#39;</span>){</span>
<span id="cb49-5"><a href="step-by-step-scrna-seq-pipeline.html#cb49-5" tabindex="-1"></a>  OrgDb <span class="ot">&lt;-</span> <span class="st">&#39;org.Hs.eg.db&#39;</span></span>
<span id="cb49-6"><a href="step-by-step-scrna-seq-pipeline.html#cb49-6" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb49-7"><a href="step-by-step-scrna-seq-pipeline.html#cb49-7" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">&quot;Org should be &#39;mmu&#39; or &#39;hsa&#39;.&quot;</span>)</span>
<span id="cb49-8"><a href="step-by-step-scrna-seq-pipeline.html#cb49-8" tabindex="-1"></a>}</span>
<span id="cb49-9"><a href="step-by-step-scrna-seq-pipeline.html#cb49-9" tabindex="-1"></a></span>
<span id="cb49-10"><a href="step-by-step-scrna-seq-pipeline.html#cb49-10" tabindex="-1"></a><span class="fu">HemaScopeREnrichment</span>(<span class="at">DEGs=</span>sc_object.markers,</span>
<span id="cb49-11"><a href="step-by-step-scrna-seq-pipeline.html#cb49-11" tabindex="-1"></a>                     <span class="at">OrgDb=</span>OrgDb,</span>
<span id="cb49-12"><a href="step-by-step-scrna-seq-pipeline.html#cb49-12" tabindex="-1"></a>                     <span class="at">output.dir=</span><span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/&#39;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-57"></span>
<img src="sc_img/step6/X0_BP_GOenrich.jpg" alt="Barplot showing GO（BP）and KEGG enrichment results of each cell type group" width="50%" /><img src="sc_img/step6/X0_KEGGenrich.jpg" alt="Barplot showing GO（BP）and KEGG enrichment results of each cell type group" width="50%" />
<p class="caption">
Figure 4.17: Barplot showing GO（BP）and KEGG enrichment results of each cell type group
</p>
</div>
</div>
<div id="perform-subnetwork-analysis" class="section level3 hasAnchor" number="4.7.4">
<h3><span class="header-section-number">4.7.4</span> Perform subnetwork analysis<a href="step-by-step-scrna-seq-pipeline.html#perform-subnetwork-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the results of gene network analysis.</li>
</ul>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="step-by-step-scrna-seq-pipeline.html#cb50-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/OpenXGR/&#39;</span>))) {</span>
<span id="cb50-2"><a href="step-by-step-scrna-seq-pipeline.html#cb50-2" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/OpenXGR/&#39;</span>))</span>
<span id="cb50-3"><a href="step-by-step-scrna-seq-pipeline.html#cb50-3" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Perform gene network analysis.</li>
</ul>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="step-by-step-scrna-seq-pipeline.html#cb51-1" tabindex="-1"></a><span class="fu">OpenXGR_SAG</span>(<span class="at">sc_object.markers =</span> sc_object.markers,</span>
<span id="cb51-2"><a href="step-by-step-scrna-seq-pipeline.html#cb51-2" tabindex="-1"></a>            <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step6.Find_DEGs/OpenXGR/&#39;</span>),</span>
<span id="cb51-3"><a href="step-by-step-scrna-seq-pipeline.html#cb51-3" tabindex="-1"></a>            <span class="at">subnet.size =</span> <span class="dv">10</span>)</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="step-by-step-scrna-seq-pipeline.html#cb52-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb52-2"><a href="step-by-step-scrna-seq-pipeline.html#cb52-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb52-3"><a href="step-by-step-scrna-seq-pipeline.html#cb52-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb52-4"><a href="step-by-step-scrna-seq-pipeline.html#cb52-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb52-5"><a href="step-by-step-scrna-seq-pipeline.html#cb52-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb52-6"><a href="step-by-step-scrna-seq-pipeline.html#cb52-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb52-7"><a href="step-by-step-scrna-seq-pipeline.html#cb52-7" tabindex="-1"></a>}  </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-61"></span>
<img src="sc_img/step6/SAG-network_cluster_1.jpg" alt="Figure showing subnetwork of each cell type group identified by OpenXGR" width="400px" />
<p class="caption">
Figure 4.18: Figure showing subnetwork of each cell type group identified by OpenXGR
</p>
</div>
</div>
</div>
<div id="step-7.-assign-cell-cycles" class="section level2 hasAnchor" number="4.8">
<h2><span class="header-section-number">4.8</span> Step 7. Assign Cell Cycles<a href="step-by-step-scrna-seq-pipeline.html#step-7.-assign-cell-cycles" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This step assigns cell cycle phases by analyzing cell cycle-related genes and generates plots of the cell cycle analysis results.</p>
<div id="function-arguments-1" class="section level3 hasAnchor" number="4.8.1">
<h3><span class="header-section-number">4.8.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>sc_object</code>: A Seurat object containing single-cell RNA sequencing data.</li>
<li><code>counts_matrix</code>: The ‘counts’ slot in the Seurat object.</li>
<li><code>data_matrix</code>: The ‘data’ slot in the Seurat object.</li>
<li><code>cellcycleCutoff</code>: The cutoff value for distinguishing between cycling and quiescent cells. Cells with a G1G2Score below this cutoff are considered quiescent.</li>
<li><code>cellTypeOrders</code>: The order of cell types for visualization. If not provided, the function will use the unique cell types in the input Seurat object.</li>
<li><code>databasePath</code>: The path to the database required for the analysis.</li>
<li><code>Org</code>: A character vector specifying the species of cell cycle genes, can be ‘mmu’ (mouse) or ‘hsa’ (human).</li>
</ul>
</div>
<div id="codes-for-step7" class="section level3 hasAnchor" number="4.8.2">
<h3><span class="header-section-number">4.8.2</span> codes for step7<a href="step-by-step-scrna-seq-pipeline.html#codes-for-step7" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the results of cell cycle analysis.</li>
</ul>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="step-by-step-scrna-seq-pipeline.html#cb53-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step7. Assign cell cycles.&#39;</span>)</span>
<span id="cb53-2"><a href="step-by-step-scrna-seq-pipeline.html#cb53-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step7.Assign_cell_cycles/&#39;</span>))) {</span>
<span id="cb53-3"><a href="step-by-step-scrna-seq-pipeline.html#cb53-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step7.Assign_cell_cycles/&#39;</span>))</span>
<span id="cb53-4"><a href="step-by-step-scrna-seq-pipeline.html#cb53-4" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Set the parameters for the cell cycle analysis.</li>
</ul>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="step-by-step-scrna-seq-pipeline.html#cb54-1" tabindex="-1"></a>cellcycleCutoff <span class="ot">=</span> <span class="cn">NULL</span></span></code></pre></div>
<ul>
<li>Run the cell cycle analysis.</li>
</ul>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="step-by-step-scrna-seq-pipeline.html#cb55-1" tabindex="-1"></a>datasets.before.batch.removal <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>),<span class="st">&#39;datasets.before.batch.removal.rds&#39;</span>))  </span>
<span id="cb55-2"><a href="step-by-step-scrna-seq-pipeline.html#cb55-2" tabindex="-1"></a>sc_object <span class="ot">&lt;-</span> <span class="fu">cellCycle</span>(<span class="at">sc_object=</span>sc_object,</span>
<span id="cb55-3"><a href="step-by-step-scrna-seq-pipeline.html#cb55-3" tabindex="-1"></a>                       <span class="at">counts_matrix =</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> datasets.before.batch.removal, <span class="at">slot =</span> <span class="st">&quot;counts&quot;</span>)<span class="sc">%&gt;%</span><span class="fu">as.matrix</span>(),</span>
<span id="cb55-4"><a href="step-by-step-scrna-seq-pipeline.html#cb55-4" tabindex="-1"></a>                       <span class="at">data_matrix =</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> datasets.before.batch.removal, <span class="at">slot =</span> <span class="st">&quot;data&quot;</span>)<span class="sc">%&gt;%</span><span class="fu">as.matrix</span>(),</span>
<span id="cb55-5"><a href="step-by-step-scrna-seq-pipeline.html#cb55-5" tabindex="-1"></a>                       <span class="at">cellcycleCutoff =</span> cellcycleCutoff,</span>
<span id="cb55-6"><a href="step-by-step-scrna-seq-pipeline.html#cb55-6" tabindex="-1"></a>                       <span class="at">cellTypeOrders =</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels),</span>
<span id="cb55-7"><a href="step-by-step-scrna-seq-pipeline.html#cb55-7" tabindex="-1"></a>                       <span class="at">output.dir=</span><span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step7.Assign_cell_cycles/&#39;</span>),</span>
<span id="cb55-8"><a href="step-by-step-scrna-seq-pipeline.html#cb55-8" tabindex="-1"></a>                       <span class="at">databasePath =</span> databasePath,</span>
<span id="cb55-9"><a href="step-by-step-scrna-seq-pipeline.html#cb55-9" tabindex="-1"></a>                       <span class="at">Org =</span> Org)</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="step-by-step-scrna-seq-pipeline.html#cb56-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb56-2"><a href="step-by-step-scrna-seq-pipeline.html#cb56-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb56-3"><a href="step-by-step-scrna-seq-pipeline.html#cb56-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb56-4"><a href="step-by-step-scrna-seq-pipeline.html#cb56-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb56-5"><a href="step-by-step-scrna-seq-pipeline.html#cb56-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb56-6"><a href="step-by-step-scrna-seq-pipeline.html#cb56-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb56-7"><a href="step-by-step-scrna-seq-pipeline.html#cb56-7" tabindex="-1"></a>}  </span></code></pre></div>
</div>
<div id="outputs-1" class="section level3 hasAnchor" number="4.8.3">
<h3><span class="header-section-number">4.8.3</span> Outputs<a href="step-by-step-scrna-seq-pipeline.html#outputs-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-66"></span>
<img src="sc_img/step7/CellCycle.jpg" alt="Barplot showing the proportion of different cell cycle within each cell type group" width="400px" />
<p class="caption">
Figure 4.19: Barplot showing the proportion of different cell cycle within each cell type group
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-67"></span>
<img src="sc_img/step7/G1G2Score.jpg" alt="Density plot showing the distribution of cell cycle scores" width="400px" />
<p class="caption">
Figure 4.20: Density plot showing the distribution of cell cycle scores
</p>
</div>
</div>
</div>
<div id="step-8.-calculate-heterogeneity" class="section level2 hasAnchor" number="4.9">
<h2><span class="header-section-number">4.9</span> Step 8. Calculate Heterogeneity<a href="step-by-step-scrna-seq-pipeline.html#step-8.-calculate-heterogeneity" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This step quantifies cell heterogeneity by computing Spearman correlation coefficients between cells within the same cell type groups.</p>
<div id="function-arguments-2" class="section level3 hasAnchor" number="4.9.1">
<h3><span class="header-section-number">4.9.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>expression_matrix</code>: A numeric matrix representing the expression data, where rows are genes and columns are cells. The matrix should be appropriately preprocessed and filtered before using this function.</li>
<li><code>cell_types_groups</code>: A data frame specifying cell type annotations for each cell, including cell type labels and group information.</li>
<li><code>cellTypeOrders</code>: The order of cell types for visualization. If not provided, the function will use the unique cell types in the input cell_types_groups.</li>
</ul>
</div>
<div id="codes-for-step8" class="section level3 hasAnchor" number="4.9.2">
<h3><span class="header-section-number">4.9.2</span> codes for step8<a href="step-by-step-scrna-seq-pipeline.html#codes-for-step8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the results of heterogeneity calculation.</li>
</ul>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="step-by-step-scrna-seq-pipeline.html#cb57-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step8. Calculate heterogeneity.&#39;</span>)</span>
<span id="cb57-2"><a href="step-by-step-scrna-seq-pipeline.html#cb57-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step8.Calculate_heterogeneity/&#39;</span>))) {</span>
<span id="cb57-3"><a href="step-by-step-scrna-seq-pipeline.html#cb57-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step8.Calculate_heterogeneity/&#39;</span>))</span>
<span id="cb57-4"><a href="step-by-step-scrna-seq-pipeline.html#cb57-4" tabindex="-1"></a>}  </span></code></pre></div>
<ul>
<li>Run heterogeneity calculation process.</li>
</ul>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="step-by-step-scrna-seq-pipeline.html#cb58-1" tabindex="-1"></a>expression_matrix <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> datasets.before.batch.removal, <span class="at">slot =</span> <span class="st">&quot;data&quot;</span>)<span class="sc">%&gt;%</span><span class="fu">as.matrix</span>()</span>
<span id="cb58-2"><a href="step-by-step-scrna-seq-pipeline.html#cb58-2" tabindex="-1"></a>expression_matrix <span class="ot">&lt;-</span> expression_matrix[,<span class="fu">rownames</span>(sc_object<span class="sc">@</span>meta.data)]</span>
<span id="cb58-3"><a href="step-by-step-scrna-seq-pipeline.html#cb58-3" tabindex="-1"></a>cell_types_groups <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels,</span>
<span id="cb58-4"><a href="step-by-step-scrna-seq-pipeline.html#cb58-4" tabindex="-1"></a>                                         sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>datasetID))</span>
<span id="cb58-5"><a href="step-by-step-scrna-seq-pipeline.html#cb58-5" tabindex="-1"></a><span class="fu">colnames</span>(cell_types_groups) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;clusters&#39;</span>, <span class="st">&#39;datasetID&#39;</span>)</span>
<span id="cb58-6"><a href="step-by-step-scrna-seq-pipeline.html#cb58-6" tabindex="-1"></a></span>
<span id="cb58-7"><a href="step-by-step-scrna-seq-pipeline.html#cb58-7" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(ViolinPlot.cellTypeOrders)){</span>
<span id="cb58-8"><a href="step-by-step-scrna-seq-pipeline.html#cb58-8" tabindex="-1"></a>  cellTypes_orders <span class="ot">&lt;-</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels)</span>
<span id="cb58-9"><a href="step-by-step-scrna-seq-pipeline.html#cb58-9" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb58-10"><a href="step-by-step-scrna-seq-pipeline.html#cb58-10" tabindex="-1"></a>  cellTypes_orders <span class="ot">&lt;-</span> ViolinPlot.cellTypeOrders  </span>
<span id="cb58-11"><a href="step-by-step-scrna-seq-pipeline.html#cb58-11" tabindex="-1"></a>} </span>
<span id="cb58-12"><a href="step-by-step-scrna-seq-pipeline.html#cb58-12" tabindex="-1"></a></span>
<span id="cb58-13"><a href="step-by-step-scrna-seq-pipeline.html#cb58-13" tabindex="-1"></a><span class="fu">heterogeneity</span>(<span class="at">expression_matrix =</span> expression_matrix,</span>
<span id="cb58-14"><a href="step-by-step-scrna-seq-pipeline.html#cb58-14" tabindex="-1"></a>              <span class="at">cell_types_groups =</span> cell_types_groups,</span>
<span id="cb58-15"><a href="step-by-step-scrna-seq-pipeline.html#cb58-15" tabindex="-1"></a>              <span class="at">cellTypeOrders =</span> cellTypes_orders,</span>
<span id="cb58-16"><a href="step-by-step-scrna-seq-pipeline.html#cb58-16" tabindex="-1"></a>              <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step8.Calculate_heterogeneity/&#39;</span>))</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="step-by-step-scrna-seq-pipeline.html#cb59-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb59-2"><a href="step-by-step-scrna-seq-pipeline.html#cb59-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb59-3"><a href="step-by-step-scrna-seq-pipeline.html#cb59-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb59-4"><a href="step-by-step-scrna-seq-pipeline.html#cb59-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb59-5"><a href="step-by-step-scrna-seq-pipeline.html#cb59-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb59-6"><a href="step-by-step-scrna-seq-pipeline.html#cb59-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb59-7"><a href="step-by-step-scrna-seq-pipeline.html#cb59-7" tabindex="-1"></a>} </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-71"></span>
<img src="sc_img/step8/Heterogeneity_spearmanCorrelation.png" alt="Box plot showing the Spearman correlation coefficients between cells within the same cell type groups(here we take data including more samples as an example)" width="400px" />
<p class="caption">
Figure 4.21: Box plot showing the Spearman correlation coefficients between cells within the same cell type groups(here we take data including more samples as an example)
</p>
</div>
</div>
</div>
<div id="step-9.-violin-plot-for-marker-genes" class="section level2 hasAnchor" number="4.10">
<h2><span class="header-section-number">4.10</span> Step 9. Violin Plot for Marker Genes<a href="step-by-step-scrna-seq-pipeline.html#step-9.-violin-plot-for-marker-genes" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This step generates violin plots for marker genes across different cell types.</p>
<div id="function-arguments-3" class="section level3 hasAnchor" number="4.10.1">
<h3><span class="header-section-number">4.10.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>dataMatrix</code>: A data frame or matrix representing the expression data, where rows are cells and columns are genes.</li>
<li><code>features</code>: A character vector specifying the marker genes to plot in the violin plots.</li>
<li><code>CellTypes</code>: A factor vector containing cell type annotations for each cell.</li>
<li><code>cellTypeOrders</code>: A character vector specifying the order of cell types for plotting. Defaults to unique values in <code>CellTypes</code>.</li>
<li><code>cellTypeColors</code>: A character vector specifying the colors to use for cell type groups. Defaults to a color palette.</li>
</ul>
</div>
<div id="codes-for-step9" class="section level3 hasAnchor" number="4.10.2">
<h3><span class="header-section-number">4.10.2</span> codes for step9<a href="step-by-step-scrna-seq-pipeline.html#codes-for-step9" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the violin plots of marker genes.</li>
</ul>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="step-by-step-scrna-seq-pipeline.html#cb60-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step9. Violin plot for marker genes.&#39;</span>)</span>
<span id="cb60-2"><a href="step-by-step-scrna-seq-pipeline.html#cb60-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step9.Violin_plot_for_marker_genes/&#39;</span>))) {</span>
<span id="cb60-3"><a href="step-by-step-scrna-seq-pipeline.html#cb60-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step9.Violin_plot_for_marker_genes/&#39;</span>))</span>
<span id="cb60-4"><a href="step-by-step-scrna-seq-pipeline.html#cb60-4" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Run violin plot visualization.</li>
</ul>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="step-by-step-scrna-seq-pipeline.html#cb61-1" tabindex="-1"></a><span class="cf">if</span>( (<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">&amp;</span> Step2_Quality_Control.RemoveBatches ){</span>
<span id="cb61-2"><a href="step-by-step-scrna-seq-pipeline.html#cb61-2" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object) <span class="ot">&lt;-</span> <span class="st">&#39;integrated&#39;</span></span>
<span id="cb61-3"><a href="step-by-step-scrna-seq-pipeline.html#cb61-3" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb61-4"><a href="step-by-step-scrna-seq-pipeline.html#cb61-4" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object) <span class="ot">&lt;-</span> <span class="st">&#39;RNA&#39;</span>}</span>
<span id="cb61-5"><a href="step-by-step-scrna-seq-pipeline.html#cb61-5" tabindex="-1"></a></span>
<span id="cb61-6"><a href="step-by-step-scrna-seq-pipeline.html#cb61-6" tabindex="-1"></a>dataMatrix <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> sc_object, <span class="at">slot =</span> <span class="st">&quot;scale.data&quot;</span>)</span>
<span id="cb61-7"><a href="step-by-step-scrna-seq-pipeline.html#cb61-7" tabindex="-1"></a></span>
<span id="cb61-8"><a href="step-by-step-scrna-seq-pipeline.html#cb61-8" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(marker.genes)<span class="sc">&amp;</span>(Org <span class="sc">==</span> <span class="st">&#39;mmu&#39;</span>)){</span>
<span id="cb61-9"><a href="step-by-step-scrna-seq-pipeline.html#cb61-9" tabindex="-1"></a>    <span class="co"># mpp genes are from &#39;The bone marrow microenvironment at single cell resolution&#39;</span></span>
<span id="cb61-10"><a href="step-by-step-scrna-seq-pipeline.html#cb61-10" tabindex="-1"></a>    <span class="co"># the other genes are from &#39;single cell characterization of haematopoietic progenitors and their trajectories in homeostasis and perturbed haematopoiesis&#39;</span></span>
<span id="cb61-11"><a href="step-by-step-scrna-seq-pipeline.html#cb61-11" tabindex="-1"></a>    <span class="co"># the aliases of these genes were changed in gecodeM16：Gpr64 -&gt; Adgrg2, Sdpr -&gt; Cavin2, Hbb-b1 -&gt; Hbb-bs, Sfpi1 -&gt; Spi1</span></span>
<span id="cb61-12"><a href="step-by-step-scrna-seq-pipeline.html#cb61-12" tabindex="-1"></a>    HSC_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;Slamf1&#39;</span>, <span class="st">&#39;Itga2b&#39;</span>, <span class="st">&#39;Kit&#39;</span>, <span class="st">&#39;Ly6a&#39;</span>, <span class="st">&#39;Bmi1&#39;</span>, <span class="st">&#39;Gata2&#39;</span>, <span class="st">&#39;Hlf&#39;</span>, <span class="st">&#39;Meis1&#39;</span>, <span class="st">&#39;Mpl&#39;</span>, <span class="st">&#39;Mcl1&#39;</span>, <span class="st">&#39;Gfi1&#39;</span>, <span class="st">&#39;Gfi1b&#39;</span>, <span class="st">&#39;Hoxb5&#39;</span>)</span>
<span id="cb61-13"><a href="step-by-step-scrna-seq-pipeline.html#cb61-13" tabindex="-1"></a>    Mpp_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;Mki67&#39;</span>, <span class="st">&#39;Mpo&#39;</span>, <span class="st">&#39;Elane&#39;</span>, <span class="st">&#39;Ctsg&#39;</span>, <span class="st">&#39;Calr&#39;</span>)</span>
<span id="cb61-14"><a href="step-by-step-scrna-seq-pipeline.html#cb61-14" tabindex="-1"></a>    Erythroid_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;Klf1&#39;</span>, <span class="st">&#39;Gata1&#39;</span>, <span class="st">&#39;Mpl&#39;</span>, <span class="st">&#39;Epor&#39;</span>, <span class="st">&#39;Vwf&#39;</span>, <span class="st">&#39;Zfpm1&#39;</span>, <span class="st">&#39;Fhl1&#39;</span>, <span class="st">&#39;Adgrg2&#39;</span>, <span class="st">&#39;Cavin2&#39;</span>,<span class="st">&#39;Gypa&#39;</span>, <span class="st">&#39;Tfrc&#39;</span>, <span class="st">&#39;Hbb-bs&#39;</span>, <span class="st">&#39;Hbb-y&#39;</span>)</span>
<span id="cb61-15"><a href="step-by-step-scrna-seq-pipeline.html#cb61-15" tabindex="-1"></a>    Lymphoid_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;Tcf3&#39;</span>, <span class="st">&#39;Ikzf1&#39;</span>, <span class="st">&#39;Notch1&#39;</span>, <span class="st">&#39;Flt3&#39;</span>, <span class="st">&#39;Dntt&#39;</span>, <span class="st">&#39;Btg2&#39;</span>, <span class="st">&#39;Tcf7&#39;</span>, <span class="st">&#39;Rag1&#39;</span>, <span class="st">&#39;Ptprc&#39;</span>, <span class="st">&#39;Ly6a&#39;</span>, <span class="st">&#39;Blnk&#39;</span>)</span>
<span id="cb61-16"><a href="step-by-step-scrna-seq-pipeline.html#cb61-16" tabindex="-1"></a>    Myeloid_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;Gfi1&#39;</span>, <span class="st">&#39;Spi1&#39;</span>, <span class="st">&#39;Mpo&#39;</span>, <span class="st">&#39;Csf2rb&#39;</span>, <span class="st">&#39;Csf1r&#39;</span>, <span class="st">&#39;Gfi1b&#39;</span>, <span class="st">&#39;Hk3&#39;</span>, <span class="st">&#39;Csf2ra&#39;</span>, <span class="st">&#39;Csf3r&#39;</span>, <span class="st">&#39;Sp1&#39;</span>, <span class="st">&#39;Fcgr3&#39;</span>)</span>
<span id="cb61-17"><a href="step-by-step-scrna-seq-pipeline.html#cb61-17" tabindex="-1"></a>    marker.genes <span class="ot">&lt;-</span> <span class="fu">c</span>(HSC_lineage_signatures, Mpp_genes, Erythroid_lineage_signatures, Lymphoid_lineage_signatures, Myeloid_lineage_signatures)</span>
<span id="cb61-18"><a href="step-by-step-scrna-seq-pipeline.html#cb61-18" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(<span class="fu">is.null</span>(marker.genes)<span class="sc">&amp;</span>(Org <span class="sc">==</span> <span class="st">&#39;hsa&#39;</span>)){</span>
<span id="cb61-19"><a href="step-by-step-scrna-seq-pipeline.html#cb61-19" tabindex="-1"></a>    HSPCs_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;CD34&#39;</span>,<span class="st">&#39;KIT&#39;</span>,<span class="st">&#39;AVP&#39;</span>,<span class="st">&#39;FLT3&#39;</span>,<span class="st">&#39;MME&#39;</span>,<span class="st">&#39;CD7&#39;</span>,<span class="st">&#39;CD38&#39;</span>,<span class="st">&#39;CSF1R&#39;</span>,<span class="st">&#39;FCGR1A&#39;</span>,<span class="st">&#39;MPO&#39;</span>,<span class="st">&#39;ELANE&#39;</span>,<span class="st">&#39;IL3RA&#39;</span>)</span>
<span id="cb61-20"><a href="step-by-step-scrna-seq-pipeline.html#cb61-20" tabindex="-1"></a>    Myeloids_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;LYZ&#39;</span>,<span class="st">&#39;CD36&#39;</span>,<span class="st">&#39;MPO&#39;</span>,<span class="st">&#39;FCGR1A&#39;</span>,<span class="st">&#39;CD4&#39;</span>,<span class="st">&#39;CD14&#39;</span>,<span class="st">&#39;CD300E&#39;</span>,<span class="st">&#39;ITGAX&#39;</span>,<span class="st">&#39;FCGR3A&#39;</span>,<span class="st">&#39;FLT3&#39;</span>,<span class="st">&#39;AXL&#39;</span>,</span>
<span id="cb61-21"><a href="step-by-step-scrna-seq-pipeline.html#cb61-21" tabindex="-1"></a>                  <span class="st">&#39;SIGLEC6&#39;</span>,<span class="st">&#39;CLEC4C&#39;</span>,<span class="st">&#39;IRF4&#39;</span>,<span class="st">&#39;LILRA4&#39;</span>,<span class="st">&#39;IL3RA&#39;</span>,<span class="st">&#39;IRF8&#39;</span>,<span class="st">&#39;IRF7&#39;</span>,<span class="st">&#39;XCR1&#39;</span>,<span class="st">&#39;CD1C&#39;</span>,<span class="st">&#39;THBD&#39;</span>,</span>
<span id="cb61-22"><a href="step-by-step-scrna-seq-pipeline.html#cb61-22" tabindex="-1"></a>                  <span class="st">&#39;MRC1&#39;</span>,<span class="st">&#39;CD34&#39;</span>,<span class="st">&#39;KIT&#39;</span>,<span class="st">&#39;ITGA2B&#39;</span>,<span class="st">&#39;PF4&#39;</span>,<span class="st">&#39;CD9&#39;</span>,<span class="st">&#39;ENG&#39;</span>,<span class="st">&#39;KLF&#39;</span>,<span class="st">&#39;TFRC&#39;</span>)</span>
<span id="cb61-23"><a href="step-by-step-scrna-seq-pipeline.html#cb61-23" tabindex="-1"></a>    B_cells_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;CD79A&#39;</span>,<span class="st">&#39;IGLL1&#39;</span>,<span class="st">&#39;RAG1&#39;</span>,<span class="st">&#39;RAG2&#39;</span>,<span class="st">&#39;VPREB1&#39;</span>,<span class="st">&#39;MME&#39;</span>,<span class="st">&#39;IL7R&#39;</span>,<span class="st">&#39;DNTT&#39;</span>,<span class="st">&#39;MKI67&#39;</span>,<span class="st">&#39;PCNA&#39;</span>,<span class="st">&#39;TCL1A&#39;</span>,<span class="st">&#39;MS4A1&#39;</span>,<span class="st">&#39;IGHD&#39;</span>,<span class="st">&#39;CD27&#39;</span>,<span class="st">&#39;IGHG3&#39;</span>)</span>
<span id="cb61-24"><a href="step-by-step-scrna-seq-pipeline.html#cb61-24" tabindex="-1"></a>    T_NK_cells_lineage_signatures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;CD3D&#39;</span>,<span class="st">&#39;CD3E&#39;</span>,<span class="st">&#39;CD8A&#39;</span>,<span class="st">&#39;CCR7&#39;</span>,<span class="st">&#39;IL7R&#39;</span>,<span class="st">&#39;SELL&#39;</span>,<span class="st">&#39;KLRG1&#39;</span>,<span class="st">&#39;CD27&#39;</span>,<span class="st">&#39;GNLY&#39;</span>,</span>
<span id="cb61-25"><a href="step-by-step-scrna-seq-pipeline.html#cb61-25" tabindex="-1"></a>                    <span class="st">&#39;NKG7&#39;</span>,<span class="st">&#39;PDCD1&#39;</span>,<span class="st">&#39;TNFRSF9&#39;</span>,<span class="st">&#39;LAG3&#39;</span>,<span class="st">&#39;CD160&#39;</span>,<span class="st">&#39;CD4&#39;</span>,<span class="st">&#39;CD40LG&#39;</span>,<span class="st">&#39;IL2RA&#39;</span>,</span>
<span id="cb61-26"><a href="step-by-step-scrna-seq-pipeline.html#cb61-26" tabindex="-1"></a>                    <span class="st">&#39;FOXP3&#39;</span>,<span class="st">&#39;DUSP4&#39;</span>,<span class="st">&#39;IL2RB&#39;</span>,<span class="st">&#39;KLRF1&#39;</span>,<span class="st">&#39;FCGR3A&#39;</span>,<span class="st">&#39;NCAM1&#39;</span>,<span class="st">&#39;XCL1&#39;</span>,<span class="st">&#39;MKI67&#39;</span>,<span class="st">&#39;PCNA&#39;</span>,<span class="st">&#39;KLRF&#39;</span>)</span>
<span id="cb61-27"><a href="step-by-step-scrna-seq-pipeline.html#cb61-27" tabindex="-1"></a>    marker.genes <span class="ot">&lt;-</span> <span class="fu">c</span>(HSPCs_lineage_signatures, Myeloids_lineage_signatures, B_cells_lineage_signatures, T_NK_cells_lineage_signatures)</span>
<span id="cb61-28"><a href="step-by-step-scrna-seq-pipeline.html#cb61-28" tabindex="-1"></a>}</span>
<span id="cb61-29"><a href="step-by-step-scrna-seq-pipeline.html#cb61-29" tabindex="-1"></a></span>
<span id="cb61-30"><a href="step-by-step-scrna-seq-pipeline.html#cb61-30" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(ViolinPlot.cellTypeOrders)){</span>
<span id="cb61-31"><a href="step-by-step-scrna-seq-pipeline.html#cb61-31" tabindex="-1"></a>  ViolinPlot.cellTypeOrders <span class="ot">&lt;-</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels)</span>
<span id="cb61-32"><a href="step-by-step-scrna-seq-pipeline.html#cb61-32" tabindex="-1"></a>}</span>
<span id="cb61-33"><a href="step-by-step-scrna-seq-pipeline.html#cb61-33" tabindex="-1"></a></span>
<span id="cb61-34"><a href="step-by-step-scrna-seq-pipeline.html#cb61-34" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(ViolinPlot.cellTypeColors)){</span>
<span id="cb61-35"><a href="step-by-step-scrna-seq-pipeline.html#cb61-35" tabindex="-1"></a>  ViolinPlot.cellTypeColors <span class="ot">&lt;-</span> viridis<span class="sc">::</span><span class="fu">viridis</span>(<span class="fu">length</span>(<span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels)))</span>
<span id="cb61-36"><a href="step-by-step-scrna-seq-pipeline.html#cb61-36" tabindex="-1"></a>}</span>
<span id="cb61-37"><a href="step-by-step-scrna-seq-pipeline.html#cb61-37" tabindex="-1"></a></span>
<span id="cb61-38"><a href="step-by-step-scrna-seq-pipeline.html#cb61-38" tabindex="-1"></a><span class="fu">combinedViolinPlot</span>(<span class="at">dataMatrix =</span> dataMatrix,</span>
<span id="cb61-39"><a href="step-by-step-scrna-seq-pipeline.html#cb61-39" tabindex="-1"></a>                   <span class="at">features =</span> marker.genes,</span>
<span id="cb61-40"><a href="step-by-step-scrna-seq-pipeline.html#cb61-40" tabindex="-1"></a>                   <span class="at">CellTypes =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels,</span>
<span id="cb61-41"><a href="step-by-step-scrna-seq-pipeline.html#cb61-41" tabindex="-1"></a>                   <span class="at">cellTypeOrders =</span> ViolinPlot.cellTypeOrders,</span>
<span id="cb61-42"><a href="step-by-step-scrna-seq-pipeline.html#cb61-42" tabindex="-1"></a>                   <span class="at">cellTypeColors =</span> ViolinPlot.cellTypeColors,</span>
<span id="cb61-43"><a href="step-by-step-scrna-seq-pipeline.html#cb61-43" tabindex="-1"></a>                   <span class="at">Org =</span> Org,</span>
<span id="cb61-44"><a href="step-by-step-scrna-seq-pipeline.html#cb61-44" tabindex="-1"></a>                   <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step9.Violin_plot_for_marker_genes/&#39;</span>),</span>
<span id="cb61-45"><a href="step-by-step-scrna-seq-pipeline.html#cb61-45" tabindex="-1"></a>                   <span class="at">databasePath =</span> databasePath)</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="step-by-step-scrna-seq-pipeline.html#cb62-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb62-2"><a href="step-by-step-scrna-seq-pipeline.html#cb62-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb62-3"><a href="step-by-step-scrna-seq-pipeline.html#cb62-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb62-4"><a href="step-by-step-scrna-seq-pipeline.html#cb62-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb62-5"><a href="step-by-step-scrna-seq-pipeline.html#cb62-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb62-6"><a href="step-by-step-scrna-seq-pipeline.html#cb62-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb62-7"><a href="step-by-step-scrna-seq-pipeline.html#cb62-7" tabindex="-1"></a>}  </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-75"></span>
<img src="sc_img/step9/combinedViolin.png" alt="Violin plot showing the expression of marker genes between cell type groups" width="400px" />
<p class="caption">
Figure 4.22: Violin plot showing the expression of marker genes between cell type groups
</p>
</div>
</div>
</div>
<div id="step-10.-calculate-lineage-scores" class="section level2 hasAnchor" number="4.11">
<h2><span class="header-section-number">4.11</span> Step 10. Calculate Lineage Scores<a href="step-by-step-scrna-seq-pipeline.html#step-10.-calculate-lineage-scores" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This step calculates lineage scores for specified gene sets based on the provided expression data. It then generates a heatmap of lineage scores and a heatmap of gene expression patterns.</p>
<div id="function-arguments-4" class="section level3 hasAnchor" number="4.11.1">
<h3><span class="header-section-number">4.11.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>expression_matrix</code>: A data frame or matrix representing the expression data, where rows are cells and columns are genes.</li>
<li><code>cellTypes</code>: A character vector specifying cell type annotations for each cell. e.g. c(“HSC”,“HSC”,“HSC”,“MPP1”,“MPP2”,“MPP2”,“MPP2” …)</li>
<li><code>cellTypes_orders</code>: A character vector specifying the order of cell types for plotting. e.g. c(“HSC”,“MPP1”,“MPP2”)</li>
<li><code>cellTypes_colors</code>: A character vector specifying the colors to use for cell type groups. e.g. c(“HSC” = ‘#006d2c’,“MPP1” = ‘#4292c6’,“MPP2”= ‘#810f7c’).</li>
<li><code>groups</code>: A character vector specifying groups or clusters within each cell type.</li>
<li><code>groups_orders</code>: A character vector specifying the order of groups or clusters for plotting.</li>
<li><code>groups_colors</code>: A character vector specifying the colors to use for group or cluster annotations. e.g. c(‘group1’=‘#d73027’,‘group2’=‘#2171b5’)</li>
<li><code>lineage.genelist</code>: A list of gene sets representing lineage markers.</li>
<li><code>lineage.names</code>: A character vector specifying the names of the lineages.</li>
</ul>
</div>
<div id="codes-for-step10" class="section level3 hasAnchor" number="4.11.2">
<h3><span class="header-section-number">4.11.2</span> codes for step10<a href="step-by-step-scrna-seq-pipeline.html#codes-for-step10" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the results of lineage score calculation.</li>
</ul>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="step-by-step-scrna-seq-pipeline.html#cb63-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step10. Calculate lineage scores.&#39;</span>)</span>
<span id="cb63-2"><a href="step-by-step-scrna-seq-pipeline.html#cb63-2" tabindex="-1"></a><span class="co"># we use normalized data here</span></span>
<span id="cb63-3"><a href="step-by-step-scrna-seq-pipeline.html#cb63-3" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step10.Calculate_lineage_scores/&#39;</span>))) {</span>
<span id="cb63-4"><a href="step-by-step-scrna-seq-pipeline.html#cb63-4" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step10.Calculate_lineage_scores/&#39;</span>))</span>
<span id="cb63-5"><a href="step-by-step-scrna-seq-pipeline.html#cb63-5" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Run lineage score calculation.</li>
</ul>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="step-by-step-scrna-seq-pipeline.html#cb64-1" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(lineage.genelist)<span class="sc">&amp;</span><span class="fu">is.null</span>(lineage.names)<span class="sc">&amp;</span>(Org <span class="sc">==</span> <span class="st">&#39;mmu&#39;</span>)){</span>
<span id="cb64-2"><a href="step-by-step-scrna-seq-pipeline.html#cb64-2" tabindex="-1"></a>    lineage.genelist <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">list</span>(HSC_lineage_signatures),</span>
<span id="cb64-3"><a href="step-by-step-scrna-seq-pipeline.html#cb64-3" tabindex="-1"></a>                          <span class="fu">list</span>(Mpp_genes),</span>
<span id="cb64-4"><a href="step-by-step-scrna-seq-pipeline.html#cb64-4" tabindex="-1"></a>                          <span class="fu">list</span>(Erythroid_lineage_signatures),</span>
<span id="cb64-5"><a href="step-by-step-scrna-seq-pipeline.html#cb64-5" tabindex="-1"></a>                          <span class="fu">list</span>(Lymphoid_lineage_signatures),</span>
<span id="cb64-6"><a href="step-by-step-scrna-seq-pipeline.html#cb64-6" tabindex="-1"></a>                          <span class="fu">list</span>(Myeloid_lineage_signatures))</span>
<span id="cb64-7"><a href="step-by-step-scrna-seq-pipeline.html#cb64-7" tabindex="-1"></a>    lineage.names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;HSC_lineage_signatures&#39;</span>,</span>
<span id="cb64-8"><a href="step-by-step-scrna-seq-pipeline.html#cb64-8" tabindex="-1"></a>                       <span class="st">&#39;Mpp_genes&#39;</span>,</span>
<span id="cb64-9"><a href="step-by-step-scrna-seq-pipeline.html#cb64-9" tabindex="-1"></a>                       <span class="st">&#39;Erythroid_lineage_signatures&#39;</span>,</span>
<span id="cb64-10"><a href="step-by-step-scrna-seq-pipeline.html#cb64-10" tabindex="-1"></a>                       <span class="st">&#39;Lymphoid_lineage_signatures&#39;</span>,</span>
<span id="cb64-11"><a href="step-by-step-scrna-seq-pipeline.html#cb64-11" tabindex="-1"></a>                       <span class="st">&#39;Myeloid_lineage_signatures&#39;</span>)</span>
<span id="cb64-12"><a href="step-by-step-scrna-seq-pipeline.html#cb64-12" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(<span class="fu">is.null</span>(lineage.genelist)<span class="sc">&amp;</span><span class="fu">is.null</span>(lineage.names)<span class="sc">&amp;</span>(Org <span class="sc">==</span> <span class="st">&#39;hsa&#39;</span>)){</span>
<span id="cb64-13"><a href="step-by-step-scrna-seq-pipeline.html#cb64-13" tabindex="-1"></a>    lineage.genelist <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">list</span>(HSPCs_lineage_signatures),</span>
<span id="cb64-14"><a href="step-by-step-scrna-seq-pipeline.html#cb64-14" tabindex="-1"></a>                          <span class="fu">list</span>(Myeloids_lineage_signatures),</span>
<span id="cb64-15"><a href="step-by-step-scrna-seq-pipeline.html#cb64-15" tabindex="-1"></a>                          <span class="fu">list</span>(B_cells_lineage_signatures),</span>
<span id="cb64-16"><a href="step-by-step-scrna-seq-pipeline.html#cb64-16" tabindex="-1"></a>                          <span class="fu">list</span>(T_NK_cells_lineage_signatures))</span>
<span id="cb64-17"><a href="step-by-step-scrna-seq-pipeline.html#cb64-17" tabindex="-1"></a>    lineage.names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;HSPCs_lineage_signatures&#39;</span>,</span>
<span id="cb64-18"><a href="step-by-step-scrna-seq-pipeline.html#cb64-18" tabindex="-1"></a>                       <span class="st">&#39;Myeloids_lineage_signatures&#39;</span>,</span>
<span id="cb64-19"><a href="step-by-step-scrna-seq-pipeline.html#cb64-19" tabindex="-1"></a>                       <span class="st">&#39;B_cells_lineage_signatures&#39;</span>,</span>
<span id="cb64-20"><a href="step-by-step-scrna-seq-pipeline.html#cb64-20" tabindex="-1"></a>                       <span class="st">&#39;T_NK_cells_lineage_signatures&#39;</span>) </span>
<span id="cb64-21"><a href="step-by-step-scrna-seq-pipeline.html#cb64-21" tabindex="-1"></a>}</span>
<span id="cb64-22"><a href="step-by-step-scrna-seq-pipeline.html#cb64-22" tabindex="-1"></a></span>
<span id="cb64-23"><a href="step-by-step-scrna-seq-pipeline.html#cb64-23" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(ViolinPlot.cellTypeOrders)){</span>
<span id="cb64-24"><a href="step-by-step-scrna-seq-pipeline.html#cb64-24" tabindex="-1"></a>  cellTypes_orders <span class="ot">&lt;-</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels)</span>
<span id="cb64-25"><a href="step-by-step-scrna-seq-pipeline.html#cb64-25" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb64-26"><a href="step-by-step-scrna-seq-pipeline.html#cb64-26" tabindex="-1"></a>  cellTypes_orders <span class="ot">&lt;-</span> ViolinPlot.cellTypeOrders  </span>
<span id="cb64-27"><a href="step-by-step-scrna-seq-pipeline.html#cb64-27" tabindex="-1"></a>}</span>
<span id="cb64-28"><a href="step-by-step-scrna-seq-pipeline.html#cb64-28" tabindex="-1"></a>    </span>
<span id="cb64-29"><a href="step-by-step-scrna-seq-pipeline.html#cb64-29" tabindex="-1"></a><span class="fu">lineageScores</span>(<span class="at">expression_matrix =</span> expression_matrix,</span>
<span id="cb64-30"><a href="step-by-step-scrna-seq-pipeline.html#cb64-30" tabindex="-1"></a>              <span class="at">cellTypes =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels,</span>
<span id="cb64-31"><a href="step-by-step-scrna-seq-pipeline.html#cb64-31" tabindex="-1"></a>              <span class="at">cellTypes_orders =</span> cellTypes_orders,</span>
<span id="cb64-32"><a href="step-by-step-scrna-seq-pipeline.html#cb64-32" tabindex="-1"></a>              <span class="at">cellTypes_colors =</span> ViolinPlot.cellTypeColors,</span>
<span id="cb64-33"><a href="step-by-step-scrna-seq-pipeline.html#cb64-33" tabindex="-1"></a>              <span class="at">groups =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>datasetID,</span>
<span id="cb64-34"><a href="step-by-step-scrna-seq-pipeline.html#cb64-34" tabindex="-1"></a>              <span class="at">groups_orders =</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>datasetID),</span>
<span id="cb64-35"><a href="step-by-step-scrna-seq-pipeline.html#cb64-35" tabindex="-1"></a>              <span class="at">groups_colors =</span> groups_colors,</span>
<span id="cb64-36"><a href="step-by-step-scrna-seq-pipeline.html#cb64-36" tabindex="-1"></a>              <span class="at">lineage.genelist =</span> lineage.genelist,</span>
<span id="cb64-37"><a href="step-by-step-scrna-seq-pipeline.html#cb64-37" tabindex="-1"></a>              <span class="at">lineage.names =</span> lineage.names,</span>
<span id="cb64-38"><a href="step-by-step-scrna-seq-pipeline.html#cb64-38" tabindex="-1"></a>              <span class="at">Org =</span> Org,</span>
<span id="cb64-39"><a href="step-by-step-scrna-seq-pipeline.html#cb64-39" tabindex="-1"></a>              <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step10.Calculate_lineage_scores/&#39;</span>),</span>
<span id="cb64-40"><a href="step-by-step-scrna-seq-pipeline.html#cb64-40" tabindex="-1"></a>              <span class="at">databasePath =</span> databasePath)</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="step-by-step-scrna-seq-pipeline.html#cb65-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb65-2"><a href="step-by-step-scrna-seq-pipeline.html#cb65-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb65-3"><a href="step-by-step-scrna-seq-pipeline.html#cb65-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb65-4"><a href="step-by-step-scrna-seq-pipeline.html#cb65-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb65-5"><a href="step-by-step-scrna-seq-pipeline.html#cb65-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb65-6"><a href="step-by-step-scrna-seq-pipeline.html#cb65-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb65-7"><a href="step-by-step-scrna-seq-pipeline.html#cb65-7" tabindex="-1"></a>} </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-79"></span>
<img src="sc_img/step10/lineageGenesHeatmap.png" alt="Heatmap showing the expression of lineage genes for each cell" width="400px" />
<p class="caption">
Figure 4.23: Heatmap showing the expression of lineage genes for each cell
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-80"></span>
<img src="sc_img/step10/lineageScoresHeatmap.png" alt="Heatmap showing the score of lineage signatures for  each cell" width="800px" />
<p class="caption">
Figure 4.24: Heatmap showing the score of lineage signatures for each cell
</p>
</div>
</div>
</div>
<div id="step-11.-gsva" class="section level2 hasAnchor" number="4.12">
<h2><span class="header-section-number">4.12</span> Step 11. GSVA<a href="step-by-step-scrna-seq-pipeline.html#step-11.-gsva" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This step runs GSVA analysis, which calculates enrichment scores for gene sets in each cell using the provided gene list. It also performs differential GSVA analysis between specified cell groups and generates heatmaps of the results.</p>
<div id="function-arguments-5" class="section level3 hasAnchor" number="4.12.1">
<h3><span class="header-section-number">4.12.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>sc_object</code>: A Seurat object containing the single-cell RNA-seq data.</li>
<li><code>GSVA.genelist</code>: A list of gene sets for GSVA analysis.</li>
<li><code>GSVA.cellTypes</code>: A character vector specifying the cell types or labels for each cell.</li>
<li><code>GSVA.cellTypes.orders</code>: A character vector specifying the order of cell types for visualization.</li>
<li><code>GSVA.cellGroups</code>: A character vector specifying the cell groups or conditions for each cell.</li>
<li><code>GSVA.identify.cellType.features</code>: Logical. If TRUE, identify cell type-specific features.</li>
<li><code>GSVA.identify.diff.features</code>: Logical. If TRUE, identify differentially expressed features between cell groups.</li>
<li><code>GSVA.comparison.design</code>: A list specifying the experimental design for differential GSVA analysis.</li>
<li><code>OrgDB</code>: An organism-specific annotation database (OrgDb) for gene symbol conversion. e.g. org.Mm.eg.db or org.Hs.eg.db.</li>
</ul>
</div>
<div id="codes-for-running-step11" class="section level3 hasAnchor" number="4.12.2">
<h3><span class="header-section-number">4.12.2</span> codes for running step11<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step11" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create a folder for saving the results of GSVA.</li>
</ul>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="step-by-step-scrna-seq-pipeline.html#cb66-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step11. GSVA.&#39;</span>)</span>
<span id="cb66-2"><a href="step-by-step-scrna-seq-pipeline.html#cb66-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step11.GSVA/&#39;</span>))) {</span>
<span id="cb66-3"><a href="step-by-step-scrna-seq-pipeline.html#cb66-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step11.GSVA/&#39;</span>))</span>
<span id="cb66-4"><a href="step-by-step-scrna-seq-pipeline.html#cb66-4" tabindex="-1"></a>} </span></code></pre></div>
<ul>
<li>Run GSVA.</li>
</ul>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="step-by-step-scrna-seq-pipeline.html#cb67-1" tabindex="-1"></a><span class="fu">setwd</span>(wdir)</span>
<span id="cb67-2"><a href="step-by-step-scrna-seq-pipeline.html#cb67-2" tabindex="-1"></a></span>
<span id="cb67-3"><a href="step-by-step-scrna-seq-pipeline.html#cb67-3" tabindex="-1"></a><span class="cf">if</span>(Org<span class="sc">==</span><span class="st">&#39;mmu&#39;</span>){</span>
<span id="cb67-4"><a href="step-by-step-scrna-seq-pipeline.html#cb67-4" tabindex="-1"></a>  <span class="fu">load</span>(<span class="fu">paste0</span>(databasePath,<span class="st">&quot;/mouse_c2_v5p2.rdata&quot;</span>))</span>
<span id="cb67-5"><a href="step-by-step-scrna-seq-pipeline.html#cb67-5" tabindex="-1"></a>  GSVA.genelist <span class="ot">&lt;-</span> Mm.c2</span>
<span id="cb67-6"><a href="step-by-step-scrna-seq-pipeline.html#cb67-6" tabindex="-1"></a>  <span class="fu">assign</span>(<span class="st">&#39;OrgDB&#39;</span>, org.Mm.eg.db)</span>
<span id="cb67-7"><a href="step-by-step-scrna-seq-pipeline.html#cb67-7" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Org<span class="sc">==</span><span class="st">&#39;hsa&#39;</span>){</span>
<span id="cb67-8"><a href="step-by-step-scrna-seq-pipeline.html#cb67-8" tabindex="-1"></a>  <span class="fu">load</span>(<span class="fu">paste0</span>(databasePath,<span class="st">&quot;/human_c2_v5p2.rdata&quot;</span>))</span>
<span id="cb67-9"><a href="step-by-step-scrna-seq-pipeline.html#cb67-9" tabindex="-1"></a>  GSVA.genelist <span class="ot">&lt;-</span> Hs.c2</span>
<span id="cb67-10"><a href="step-by-step-scrna-seq-pipeline.html#cb67-10" tabindex="-1"></a>  <span class="fu">assign</span>(<span class="st">&#39;OrgDB&#39;</span>, org.Hs.eg.db)</span>
<span id="cb67-11"><a href="step-by-step-scrna-seq-pipeline.html#cb67-11" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb67-12"><a href="step-by-step-scrna-seq-pipeline.html#cb67-12" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">&quot;Org should be &#39;mmu&#39; or &#39;hsa&#39;.&quot;</span>)</span>
<span id="cb67-13"><a href="step-by-step-scrna-seq-pipeline.html#cb67-13" tabindex="-1"></a>}</span>
<span id="cb67-14"><a href="step-by-step-scrna-seq-pipeline.html#cb67-14" tabindex="-1"></a></span>
<span id="cb67-15"><a href="step-by-step-scrna-seq-pipeline.html#cb67-15" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(ViolinPlot.cellTypeOrders)){</span>
<span id="cb67-16"><a href="step-by-step-scrna-seq-pipeline.html#cb67-16" tabindex="-1"></a>  cellTypes_orders <span class="ot">&lt;-</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels)</span>
<span id="cb67-17"><a href="step-by-step-scrna-seq-pipeline.html#cb67-17" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb67-18"><a href="step-by-step-scrna-seq-pipeline.html#cb67-18" tabindex="-1"></a>  cellTypes_orders <span class="ot">&lt;-</span> ViolinPlot.cellTypeOrders  </span>
<span id="cb67-19"><a href="step-by-step-scrna-seq-pipeline.html#cb67-19" tabindex="-1"></a>}    </span>
<span id="cb67-20"><a href="step-by-step-scrna-seq-pipeline.html#cb67-20" tabindex="-1"></a><span class="fu">run_GSVA</span>(<span class="at">sc_object =</span> sc_object,</span>
<span id="cb67-21"><a href="step-by-step-scrna-seq-pipeline.html#cb67-21" tabindex="-1"></a>         <span class="at">GSVA.genelist =</span> GSVA.genelist,</span>
<span id="cb67-22"><a href="step-by-step-scrna-seq-pipeline.html#cb67-22" tabindex="-1"></a>         <span class="at">GSVA.cellTypes =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels,</span>
<span id="cb67-23"><a href="step-by-step-scrna-seq-pipeline.html#cb67-23" tabindex="-1"></a>         <span class="at">GSVA.cellTypes.orders =</span> cellTypes_orders,</span>
<span id="cb67-24"><a href="step-by-step-scrna-seq-pipeline.html#cb67-24" tabindex="-1"></a>         <span class="at">GSVA.cellGroups =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>datasetID,</span>
<span id="cb67-25"><a href="step-by-step-scrna-seq-pipeline.html#cb67-25" tabindex="-1"></a>         <span class="at">GSVA.identify.cellType.features =</span> Step11_GSVA.identify.cellType.features,</span>
<span id="cb67-26"><a href="step-by-step-scrna-seq-pipeline.html#cb67-26" tabindex="-1"></a>         <span class="at">GSVA.identify.diff.features =</span> Step11_GSVA.identify.diff.features,</span>
<span id="cb67-27"><a href="step-by-step-scrna-seq-pipeline.html#cb67-27" tabindex="-1"></a>         <span class="at">GSVA.comparison.design =</span> Step11_GSVA.comparison.design,</span>
<span id="cb67-28"><a href="step-by-step-scrna-seq-pipeline.html#cb67-28" tabindex="-1"></a>         <span class="at">OrgDB =</span> OrgDB,</span>
<span id="cb67-29"><a href="step-by-step-scrna-seq-pipeline.html#cb67-29" tabindex="-1"></a>         <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step11.GSVA/&#39;</span>))</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="step-by-step-scrna-seq-pipeline.html#cb68-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb68-2"><a href="step-by-step-scrna-seq-pipeline.html#cb68-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb68-3"><a href="step-by-step-scrna-seq-pipeline.html#cb68-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb68-4"><a href="step-by-step-scrna-seq-pipeline.html#cb68-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb68-5"><a href="step-by-step-scrna-seq-pipeline.html#cb68-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb68-6"><a href="step-by-step-scrna-seq-pipeline.html#cb68-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb68-7"><a href="step-by-step-scrna-seq-pipeline.html#cb68-7" tabindex="-1"></a>}  </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-84"></span>
<img src="sc_img/step11/GSVA_Heatmap.png" alt="GSVA Heatmap showing the enriched pathways of each cell type group" width="800px" />
<p class="caption">
Figure 4.25: GSVA Heatmap showing the enriched pathways of each cell type group
</p>
</div>
</div>
</div>
<div id="step-12.-construct-trajectories" class="section level2 hasAnchor" number="4.13">
<h2><span class="header-section-number">4.13</span> Step 12. Construct Trajectories<a href="step-by-step-scrna-seq-pipeline.html#step-12.-construct-trajectories" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this step, users are allowed to construct trajectories using three methods including Monocle2, slingshot and scVelo.</p>
<div id="data-preparation" class="section level3 hasAnchor" number="4.13.1">
<h3><span class="header-section-number">4.13.1</span> data preparation<a href="step-by-step-scrna-seq-pipeline.html#data-preparation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Load gene symbols and ensemble IDs.</li>
</ul>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="step-by-step-scrna-seq-pipeline.html#cb69-1" tabindex="-1"></a><span class="fu">DefaultAssay</span>(sc_object) <span class="ot">&lt;-</span> <span class="st">&#39;RNA&#39;</span></span>
<span id="cb69-2"><a href="step-by-step-scrna-seq-pipeline.html#cb69-2" tabindex="-1"></a>countsSlot <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> sc_object, <span class="at">slot =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb69-3"><a href="step-by-step-scrna-seq-pipeline.html#cb69-3" tabindex="-1"></a>gene_metadata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">rownames</span>(countsSlot))</span>
<span id="cb69-4"><a href="step-by-step-scrna-seq-pipeline.html#cb69-4" tabindex="-1"></a><span class="fu">rownames</span>(gene_metadata) <span class="ot">&lt;-</span> gene_metadata[,<span class="dv">1</span>]        </span>
<span id="cb69-5"><a href="step-by-step-scrna-seq-pipeline.html#cb69-5" tabindex="-1"></a><span class="cf">if</span>(Org <span class="sc">==</span> <span class="st">&#39;mmu&#39;</span>){</span>
<span id="cb69-6"><a href="step-by-step-scrna-seq-pipeline.html#cb69-6" tabindex="-1"></a>   <span class="fu">load</span>(<span class="fu">paste0</span>(databasePath,<span class="st">&quot;/mouseGeneSymbolandEnsembleID.rdata&quot;</span>))</span>
<span id="cb69-7"><a href="step-by-step-scrna-seq-pipeline.html#cb69-7" tabindex="-1"></a>   gene_metadata <span class="sc">$</span> ensembleID <span class="ot">&lt;-</span> <span class="fu">mapvalues</span>(<span class="at">x =</span> gene_metadata[,<span class="dv">1</span>],</span>
<span id="cb69-8"><a href="step-by-step-scrna-seq-pipeline.html#cb69-8" tabindex="-1"></a>                                           <span class="at">from =</span> mouseGeneSymbolandEnsembleID<span class="sc">$</span>geneName,</span>
<span id="cb69-9"><a href="step-by-step-scrna-seq-pipeline.html#cb69-9" tabindex="-1"></a>                                           <span class="at">to =</span> mouseGeneSymbolandEnsembleID<span class="sc">$</span>ensemblIDNoDot,</span>
<span id="cb69-10"><a href="step-by-step-scrna-seq-pipeline.html#cb69-10" tabindex="-1"></a>                                           <span class="at">warn_missing =</span> <span class="cn">FALSE</span>)</span>
<span id="cb69-11"><a href="step-by-step-scrna-seq-pipeline.html#cb69-11" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(Org <span class="sc">==</span> <span class="st">&#39;hsa&#39;</span>){</span>
<span id="cb69-12"><a href="step-by-step-scrna-seq-pipeline.html#cb69-12" tabindex="-1"></a>   <span class="fu">load</span>(<span class="fu">paste0</span>(databasePath,<span class="st">&quot;/humanGeneSymbolandEnsembleID.rdata&quot;</span>))</span>
<span id="cb69-13"><a href="step-by-step-scrna-seq-pipeline.html#cb69-13" tabindex="-1"></a>   gene_metadata <span class="sc">$</span> ensembleID <span class="ot">&lt;-</span> <span class="fu">mapvalues</span>(<span class="at">x =</span> gene_metadata[,<span class="dv">1</span>],</span>
<span id="cb69-14"><a href="step-by-step-scrna-seq-pipeline.html#cb69-14" tabindex="-1"></a>                                           <span class="at">from =</span> humanGeneSymbolandEnsembleID<span class="sc">$</span>geneName,</span>
<span id="cb69-15"><a href="step-by-step-scrna-seq-pipeline.html#cb69-15" tabindex="-1"></a>                                           <span class="at">to =</span> humanGeneSymbolandEnsembleID<span class="sc">$</span>ensemblIDNoDot,</span>
<span id="cb69-16"><a href="step-by-step-scrna-seq-pipeline.html#cb69-16" tabindex="-1"></a>                                           <span class="at">warn_missing =</span> <span class="cn">FALSE</span>) </span>
<span id="cb69-17"><a href="step-by-step-scrna-seq-pipeline.html#cb69-17" tabindex="-1"></a>}</span>
<span id="cb69-18"><a href="step-by-step-scrna-seq-pipeline.html#cb69-18" tabindex="-1"></a><span class="fu">colnames</span>(gene_metadata) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;gene_short_name&#39;</span>,<span class="st">&#39;ensembleID&#39;</span>)</span></code></pre></div>
<ul>
<li>Create folders for saving the results of trajectory construction.</li>
</ul>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="step-by-step-scrna-seq-pipeline.html#cb70-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step12. Construct trajectories.&#39;</span>)</span>
<span id="cb70-2"><a href="step-by-step-scrna-seq-pipeline.html#cb70-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/&#39;</span>))) {</span>
<span id="cb70-3"><a href="step-by-step-scrna-seq-pipeline.html#cb70-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/&#39;</span>))</span>
<span id="cb70-4"><a href="step-by-step-scrna-seq-pipeline.html#cb70-4" tabindex="-1"></a>}</span>
<span id="cb70-5"><a href="step-by-step-scrna-seq-pipeline.html#cb70-5" tabindex="-1"></a></span>
<span id="cb70-6"><a href="step-by-step-scrna-seq-pipeline.html#cb70-6" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/monocle2/&#39;</span>))) {</span>
<span id="cb70-7"><a href="step-by-step-scrna-seq-pipeline.html#cb70-7" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/monocle2/&#39;</span>))</span>
<span id="cb70-8"><a href="step-by-step-scrna-seq-pipeline.html#cb70-8" tabindex="-1"></a>} </span>
<span id="cb70-9"><a href="step-by-step-scrna-seq-pipeline.html#cb70-9" tabindex="-1"></a></span>
<span id="cb70-10"><a href="step-by-step-scrna-seq-pipeline.html#cb70-10" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/slingshot/&#39;</span>))) {</span>
<span id="cb70-11"><a href="step-by-step-scrna-seq-pipeline.html#cb70-11" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/slingshot/&#39;</span>))</span>
<span id="cb70-12"><a href="step-by-step-scrna-seq-pipeline.html#cb70-12" tabindex="-1"></a>} </span>
<span id="cb70-13"><a href="step-by-step-scrna-seq-pipeline.html#cb70-13" tabindex="-1"></a></span>
<span id="cb70-14"><a href="step-by-step-scrna-seq-pipeline.html#cb70-14" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/scVelo/&#39;</span>))) {</span>
<span id="cb70-15"><a href="step-by-step-scrna-seq-pipeline.html#cb70-15" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/scVelo/&#39;</span>))</span>
<span id="cb70-16"><a href="step-by-step-scrna-seq-pipeline.html#cb70-16" tabindex="-1"></a>} </span></code></pre></div>
<ul>
<li>Prepare the input data.</li>
</ul>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="step-by-step-scrna-seq-pipeline.html#cb71-1" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">is.null</span>(Step12_Construct_Trajectories.clusters)){</span>
<span id="cb71-2"><a href="step-by-step-scrna-seq-pipeline.html#cb71-2" tabindex="-1"></a>    sc_object.subset <span class="ot">&lt;-</span> sc_object</span>
<span id="cb71-3"><a href="step-by-step-scrna-seq-pipeline.html#cb71-3" tabindex="-1"></a>    countsSlot.subset <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> sc_object.subset, <span class="at">slot =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb71-4"><a href="step-by-step-scrna-seq-pipeline.html#cb71-4" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb71-5"><a href="step-by-step-scrna-seq-pipeline.html#cb71-5" tabindex="-1"></a>    sc_object.subset <span class="ot">&lt;-</span> <span class="fu">subset</span>(sc_object, <span class="at">subset =</span> selectLabels <span class="sc">%in%</span> Step12_Construct_Trajectories.clusters)</span>
<span id="cb71-6"><a href="step-by-step-scrna-seq-pipeline.html#cb71-6" tabindex="-1"></a>    countsSlot.subset <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> sc_object.subset, <span class="at">slot =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb71-7"><a href="step-by-step-scrna-seq-pipeline.html#cb71-7" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="monocle2" class="section level3 hasAnchor" number="4.13.2">
<h3><span class="header-section-number">4.13.2</span> monocle2<a href="step-by-step-scrna-seq-pipeline.html#monocle2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Running monocle2 involves several steps:</p>
<ol style="list-style-type: decimal">
<li><p>Creating a Monocle cellDataSet using the provided cellData, phenoData, and featureData.</p></li>
<li><p>Estimating size factors, dispersions, and detecting highly variable genes.</p></li>
<li><p>Performing differential gene expression analysis to identify genes associated with cell state changes.</p></li>
<li><p>Ordering cells along the inferred trajectories and reducing dimensionality.</p></li>
<li><p>Generating and saving trajectory plots, including cell trajectory by “State” and by “Cell Types.”</p></li>
</ol>
<div id="function-arguments-6" class="section level4 hasAnchor" number="4.13.2.1">
<h4><span class="header-section-number">4.13.2.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-6" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><code>cellData</code>: A matrix of gene expression values, where columns represent cells and rows represent genes.</li>
<li><code>phenoData</code>: A data frame containing cell metadata, such as cell labels or other relevant information.</li>
<li><code>featureData</code>: A data frame containing information about features (genes) in the dataset.</li>
<li><code>lowerDetectionLimit</code>: The lower detection limit for gene expression. Genes with expression values below this limit will be treated as non-detected.</li>
<li><code>expressionFamily</code>: The family of the expression distribution used in Monocle analysis.</li>
<li><code>cellTypes</code>: A character vector specifying cell types or labels used for coloring in trajectory plots.</li>
<li><code>monocle.orders</code>: A character vector specifying the order of cell types in the Monocle analysis.</li>
<li><code>monocle.colors</code>: A character vector specifying colors for cell types in trajectory plots.</li>
</ul>
</div>
<div id="codes-for-running-monocle2" class="section level4 hasAnchor" number="4.13.2.2">
<h4><span class="header-section-number">4.13.2.2</span> codes for running monocle2<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-monocle2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="step-by-step-scrna-seq-pipeline.html#cb72-1" tabindex="-1"></a>phenoData <span class="ot">&lt;-</span> sc_object.subset<span class="sc">@</span>meta.data</span>
<span id="cb72-2"><a href="step-by-step-scrna-seq-pipeline.html#cb72-2" tabindex="-1"></a>featureData <span class="ot">&lt;-</span> gene_metadata</span>
<span id="cb72-3"><a href="step-by-step-scrna-seq-pipeline.html#cb72-3" tabindex="-1"></a><span class="fu">run_monocle</span>(<span class="at">cellData =</span> countsSlot.subset,</span>
<span id="cb72-4"><a href="step-by-step-scrna-seq-pipeline.html#cb72-4" tabindex="-1"></a>            <span class="at">phenoData =</span> phenoData,</span>
<span id="cb72-5"><a href="step-by-step-scrna-seq-pipeline.html#cb72-5" tabindex="-1"></a>            <span class="at">featureData =</span> featureData,</span>
<span id="cb72-6"><a href="step-by-step-scrna-seq-pipeline.html#cb72-6" tabindex="-1"></a>            <span class="at">lowerDetectionLimit =</span> <span class="fl">0.5</span>,</span>
<span id="cb72-7"><a href="step-by-step-scrna-seq-pipeline.html#cb72-7" tabindex="-1"></a>            <span class="at">expressionFamily =</span> VGAM<span class="sc">::</span><span class="fu">negbinomial.size</span>(),</span>
<span id="cb72-8"><a href="step-by-step-scrna-seq-pipeline.html#cb72-8" tabindex="-1"></a>            <span class="at">cellTypes=</span><span class="st">&#39;selectLabels&#39;</span>,</span>
<span id="cb72-9"><a href="step-by-step-scrna-seq-pipeline.html#cb72-9" tabindex="-1"></a>            <span class="at">monocle.orders=</span>Step12_Construct_Trajectories.clusters,</span>
<span id="cb72-10"><a href="step-by-step-scrna-seq-pipeline.html#cb72-10" tabindex="-1"></a>            <span class="at">monocle.colors =</span> ViolinPlot.cellTypeColors,</span>
<span id="cb72-11"><a href="step-by-step-scrna-seq-pipeline.html#cb72-11" tabindex="-1"></a>            <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/monocle2/&#39;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-89"></span>
<img src="sc_img/step12/monocle_cell_trajectory_cellTypes.png" alt="Figures showing cells in different trajectory states (left) and corresponding cell type groups (right)" width="50%" /><img src="sc_img/step12/monocle_cell_trajectory_state.png" alt="Figures showing cells in different trajectory states (left) and corresponding cell type groups (right)" width="50%" />
<p class="caption">
Figure 4.26: Figures showing cells in different trajectory states (left) and corresponding cell type groups (right)
</p>
</div>
</div>
</div>
<div id="slingshot" class="section level3 hasAnchor" number="4.13.3">
<h3><span class="header-section-number">4.13.3</span> Slingshot<a href="step-by-step-scrna-seq-pipeline.html#slingshot" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Running Slingshot to infer cell trajectories and lineage relationships involves several steps:</p>
<ol style="list-style-type: decimal">
<li><p>Constructs a Slingshot object using PCA embeddings, cell types, start clusters, and end clusters.</p></li>
<li><p>Computes and plots the trajectory curves.</p></li>
<li><p>Computes and plots pseudotime values along the trajectory.</p></li>
</ol>
<div id="function-arguments-7" class="section level4 hasAnchor" number="4.13.3.1">
<h4><span class="header-section-number">4.13.3.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-7" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><code>slingshot.PCAembeddings</code>: A matrix containing the PCA embeddings of the single-cell data, typically obtained from dimensionality reduction techniques like PCA.</li>
<li><code>slingshot.cellTypes</code>: A character vector specifying cell types or labels for each cell.</li>
<li><code>slingshot.start.clus</code>: A character vector specifying the initial cluster(s) from which cell trajectories should start.</li>
<li><code>slingshot.end.clus</code>: A character vector specifying the target cluster(s) where cell trajectories should end.</li>
<li><code>slingshot.colors</code>: A vector of colors corresponding to cell types for plotting. If not provided, default colors will be used.</li>
</ul>
</div>
<div id="codes-for-running-slingshot" class="section level4 hasAnchor" number="4.13.3.2">
<h4><span class="header-section-number">4.13.3.2</span> codes for running Slingshot<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-slingshot" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="step-by-step-scrna-seq-pipeline.html#cb73-1" tabindex="-1"></a><span class="cf">if</span>( (<span class="fu">length</span>(input.data.dirs) <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">&amp;</span> Step2_Quality_Control.RemoveBatches ){</span>
<span id="cb73-2"><a href="step-by-step-scrna-seq-pipeline.html#cb73-2" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object.subset) <span class="ot">&lt;-</span> <span class="st">&#39;integrated&#39;</span></span>
<span id="cb73-3"><a href="step-by-step-scrna-seq-pipeline.html#cb73-3" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb73-4"><a href="step-by-step-scrna-seq-pipeline.html#cb73-4" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(sc_object.subset) <span class="ot">&lt;-</span> <span class="st">&#39;RNA&#39;</span>}</span>
<span id="cb73-5"><a href="step-by-step-scrna-seq-pipeline.html#cb73-5" tabindex="-1"></a><span class="fu">run_slingshot</span>(<span class="at">slingshot.PCAembeddings =</span> <span class="fu">Embeddings</span>(sc_object.subset, <span class="at">reduction =</span> <span class="st">&quot;pca&quot;</span>)[, PCs],</span>
<span id="cb73-6"><a href="step-by-step-scrna-seq-pipeline.html#cb73-6" tabindex="-1"></a>              <span class="at">slingshot.cellTypes =</span> sc_object.subset<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels,</span>
<span id="cb73-7"><a href="step-by-step-scrna-seq-pipeline.html#cb73-7" tabindex="-1"></a>              <span class="at">slingshot.start.clus =</span> slingshot.start.clus,</span>
<span id="cb73-8"><a href="step-by-step-scrna-seq-pipeline.html#cb73-8" tabindex="-1"></a>              <span class="at">slingshot.end.clus =</span> slingshot.end.clus,</span>
<span id="cb73-9"><a href="step-by-step-scrna-seq-pipeline.html#cb73-9" tabindex="-1"></a>              <span class="at">slingshot.colors =</span> slingshot.colors,</span>
<span id="cb73-10"><a href="step-by-step-scrna-seq-pipeline.html#cb73-10" tabindex="-1"></a>              <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/slingshot/&#39;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-91"></span>
<img src="sc_img/step12/slingshot_curve.png" alt="Figures showing slingshot curve and infered pseudotime value" width="50%" /><img src="sc_img/step12/slingshot_pseudotime.png" alt="Figures showing slingshot curve and infered pseudotime value" width="50%" />
<p class="caption">
Figure 4.27: Figures showing slingshot curve and infered pseudotime value
</p>
</div>
</div>
</div>
<div id="scvelo" class="section level3 hasAnchor" number="4.13.4">
<h3><span class="header-section-number">4.13.4</span> scVelo<a href="step-by-step-scrna-seq-pipeline.html#scvelo" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>scVelo is implemented in Python, and it takes a Seurat object, cell embeddings, and cell type information as input. The process of data preparation includes the following steps:</p>
<ol style="list-style-type: decimal">
<li><p>Format the Seurat object metadata, including cell types and sample names.</p></li>
<li><p>Extract the spliced, unspliced, and ambiguous count matrices from the Seurat object.</p></li>
<li><p>Combine the metadata and cell embeddings.</p></li>
<li><p>Write the necessary input files for scVelo analysis, including cell embeddings, count matrices, and metadata.</p></li>
</ol>
<div id="function-arguments-8" class="section level4 hasAnchor" number="4.13.4.1">
<h4><span class="header-section-number">4.13.4.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-8" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><code>sc_object</code>: A Seurat object containing the single-cell RNA-seq data.</li>
<li><code>loom.files.path</code>: A character vector specifying the path(s) to the loom files for scVelo analysis.</li>
<li><code>scvelo.reduction</code>: A character specifying the reduction method used for scVelo analysis (default is ‘pca’).</li>
<li><code>scvelo.column</code>: A character specifying the column in the Seurat object metadata containing cell types.</li>
</ul>
</div>
<div id="codes-for-running-scvelo" class="section level4 hasAnchor" number="4.13.4.2">
<h4><span class="header-section-number">4.13.4.2</span> codes for running Scvelo<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-scvelo" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="step-by-step-scrna-seq-pipeline.html#cb74-1" tabindex="-1"></a><span class="cf">if</span>((<span class="sc">!</span><span class="fu">is.null</span>(loom.files.path))<span class="sc">&amp;</span>(<span class="sc">!</span><span class="fu">is.null</span>(pythonPath))){</span>
<span id="cb74-2"><a href="step-by-step-scrna-seq-pipeline.html#cb74-2" tabindex="-1"></a>    <span class="fu">prepareDataForScvelo</span>(<span class="at">sc_object =</span> sc_object.subset,</span>
<span id="cb74-3"><a href="step-by-step-scrna-seq-pipeline.html#cb74-3" tabindex="-1"></a>                         <span class="at">loom.files.path =</span> loom.files.path,</span>
<span id="cb74-4"><a href="step-by-step-scrna-seq-pipeline.html#cb74-4" tabindex="-1"></a>                         <span class="at">scvelo.reduction =</span> <span class="st">&#39;pca&#39;</span>,</span>
<span id="cb74-5"><a href="step-by-step-scrna-seq-pipeline.html#cb74-5" tabindex="-1"></a>                         <span class="at">scvelo.column =</span> <span class="st">&#39;selectLabels&#39;</span>,</span>
<span id="cb74-6"><a href="step-by-step-scrna-seq-pipeline.html#cb74-6" tabindex="-1"></a>                         <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step12.Construct_trajectories/scVelo/&#39;</span>))</span>
<span id="cb74-7"><a href="step-by-step-scrna-seq-pipeline.html#cb74-7" tabindex="-1"></a></span>
<span id="cb74-8"><a href="step-by-step-scrna-seq-pipeline.html#cb74-8" tabindex="-1"></a>    reticulate<span class="sc">::</span><span class="fu">py_run_string</span>(<span class="fu">paste0</span>(<span class="st">&quot;import os</span><span class="sc">\n</span><span class="st">outputDir = &#39;&quot;</span>, output.dir, <span class="st">&quot;&#39;&quot;</span>))</span>
<span id="cb74-9"><a href="step-by-step-scrna-seq-pipeline.html#cb74-9" tabindex="-1"></a>    reticulate<span class="sc">::</span><span class="fu">py_run_file</span>(<span class="fu">file.path</span>(<span class="fu">system.file</span>(<span class="at">package =</span> <span class="st">&quot;HemaScopeR&quot;</span>), <span class="st">&quot;python/sc_run_scvelo.py&quot;</span>), <span class="at">convert =</span> <span class="cn">FALSE</span>)</span>
<span id="cb74-10"><a href="step-by-step-scrna-seq-pipeline.html#cb74-10" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="step-by-step-scrna-seq-pipeline.html#cb75-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb75-2"><a href="step-by-step-scrna-seq-pipeline.html#cb75-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb75-3"><a href="step-by-step-scrna-seq-pipeline.html#cb75-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb75-4"><a href="step-by-step-scrna-seq-pipeline.html#cb75-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb75-5"><a href="step-by-step-scrna-seq-pipeline.html#cb75-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb75-6"><a href="step-by-step-scrna-seq-pipeline.html#cb75-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb75-7"><a href="step-by-step-scrna-seq-pipeline.html#cb75-7" tabindex="-1"></a>}</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-94"></span>
<img src="sc_img/step12/velocity_embedding_stream.png" alt="Figure showing trajectory predicted by scvelo" width="400px" />
<p class="caption">
Figure 4.28: Figure showing trajectory predicted by scvelo
</p>
</div>
</div>
</div>
</div>
<div id="step-13.-tf-analysis" class="section level2 hasAnchor" number="4.14">
<h2><span class="header-section-number">4.14</span> Step 13. TF Analysis<a href="step-by-step-scrna-seq-pipeline.html#step-13.-tf-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This step runs SCENIC (Single-Cell Regulatory Network Inference and Clustering) analysis, including the construction of a co-expression network, gene filtering, correlation, and the GENIE3 algorithm to infer regulatory networks.</p>
<div id="function-arguments-9" class="section level3 hasAnchor" number="4.14.1">
<h3><span class="header-section-number">4.14.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-9" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>countMatrix</code>: A matrix containing the raw counts of the single-cell RNA-seq data.</li>
<li><code>cellTypes</code>: A character vector specifying the cell types or labels for each cell.</li>
<li><code>datasetID</code>: A character vector specifying the dataset IDs for each cell.</li>
<li><code>cellTypes_colors</code>: A named vector of colors for cell type visualization.</li>
<li><code>cellTypes_orders</code>: A character vector specifying the desired order of cell types.</li>
<li><code>groups_colors</code>: A named vector of colors for grouping visualization.</li>
<li><code>groups_orders</code>: A character vector specifying the desired order of groups.</li>
<li><code>Org</code>: A character vector specifying the organism (‘mmu’ for mouse or ‘hsa’ for human).</li>
</ul>
</div>
<div id="codes-for-running-step13" class="section level3 hasAnchor" number="4.14.2">
<h3><span class="header-section-number">4.14.2</span> codes for running step13<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step13" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create folders for saving the results of TF analysis.</li>
</ul>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="step-by-step-scrna-seq-pipeline.html#cb76-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step13. TF analysis.&#39;</span>)</span>
<span id="cb76-2"><a href="step-by-step-scrna-seq-pipeline.html#cb76-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step13.TF_analysis/&#39;</span>))) {</span>
<span id="cb76-3"><a href="step-by-step-scrna-seq-pipeline.html#cb76-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step13.TF_analysis/&#39;</span>))</span>
<span id="cb76-4"><a href="step-by-step-scrna-seq-pipeline.html#cb76-4" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Run SCENIC to perform TF analysis.</li>
</ul>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="step-by-step-scrna-seq-pipeline.html#cb77-1" tabindex="-1"></a><span class="fu">run_SCENIC</span>(<span class="at">countMatrix =</span> countsSlot,</span>
<span id="cb77-2"><a href="step-by-step-scrna-seq-pipeline.html#cb77-2" tabindex="-1"></a>           <span class="at">cellTypes =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels,</span>
<span id="cb77-3"><a href="step-by-step-scrna-seq-pipeline.html#cb77-3" tabindex="-1"></a>           <span class="at">datasetID =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>datasetID,</span>
<span id="cb77-4"><a href="step-by-step-scrna-seq-pipeline.html#cb77-4" tabindex="-1"></a>           <span class="at">cellTypes_colors =</span> Step13_TF_Analysis.cellTypes_colors,</span>
<span id="cb77-5"><a href="step-by-step-scrna-seq-pipeline.html#cb77-5" tabindex="-1"></a>           <span class="at">cellTypes_orders =</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels),</span>
<span id="cb77-6"><a href="step-by-step-scrna-seq-pipeline.html#cb77-6" tabindex="-1"></a>           <span class="at">groups_colors =</span> Step13_TF_Analysis.groups_colors,</span>
<span id="cb77-7"><a href="step-by-step-scrna-seq-pipeline.html#cb77-7" tabindex="-1"></a>           <span class="at">groups_orders =</span> <span class="fu">unique</span>(sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>datasetID),</span>
<span id="cb77-8"><a href="step-by-step-scrna-seq-pipeline.html#cb77-8" tabindex="-1"></a>           <span class="at">Org =</span> Org,</span>
<span id="cb77-9"><a href="step-by-step-scrna-seq-pipeline.html#cb77-9" tabindex="-1"></a>           <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step13.TF_analysis/&#39;</span>),</span>
<span id="cb77-10"><a href="step-by-step-scrna-seq-pipeline.html#cb77-10" tabindex="-1"></a>           <span class="at">pythonPath =</span> pythonPath,</span>
<span id="cb77-11"><a href="step-by-step-scrna-seq-pipeline.html#cb77-11" tabindex="-1"></a>           <span class="at">databasePath =</span> databasePath)</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="step-by-step-scrna-seq-pipeline.html#cb78-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb78-2"><a href="step-by-step-scrna-seq-pipeline.html#cb78-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb78-3"><a href="step-by-step-scrna-seq-pipeline.html#cb78-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb78-4"><a href="step-by-step-scrna-seq-pipeline.html#cb78-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb78-5"><a href="step-by-step-scrna-seq-pipeline.html#cb78-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb78-6"><a href="step-by-step-scrna-seq-pipeline.html#cb78-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb78-7"><a href="step-by-step-scrna-seq-pipeline.html#cb78-7" tabindex="-1"></a>}</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-98"></span>
<img src="sc_img/step13/TFs_Heatmap.png" alt="Heatmap showing predicted regulon activity for each cell" width="800px" />
<p class="caption">
Figure 4.29: Heatmap showing predicted regulon activity for each cell
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-99"></span>
<img src="sc_img/step13/Step4_BoxplotActiveCellsRegulon.jpg" alt="Heatmap showing statistics of regulons" width="400px" />
<p class="caption">
Figure 4.30: Heatmap showing statistics of regulons
</p>
</div>
</div>
</div>
<div id="step-14.-cell-cell-interaction" class="section level2 hasAnchor" number="4.15">
<h2><span class="header-section-number">4.15</span> Step 14. Cell-Cell Interaction<a href="step-by-step-scrna-seq-pipeline.html#step-14.-cell-cell-interaction" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The step takes expression data, cluster labels, and other parameters to perform cell-cell communication analysis using the CellChat package. It includes the following steps:</p>
<ol style="list-style-type: decimal">
<li><p>Data input and preprocessing.</p></li>
<li><p>Initialization of a CellChat object.</p></li>
<li><p>Set the ligand-receptor interaction database based on the specified organism.</p></li>
<li><p>Preprocess the expression data for cell-cell communication analysis.</p></li>
<li><p>Identify overexpressed genes and interactions.</p></li>
<li><p>Project data based on protein-protein interaction networks.</p></li>
<li><p>Inference of cell-cell communication network.</p></li>
<li><p>Visualization of the communication network.</p></li>
<li><p>Systems analysis of cell-cell communication network.</p></li>
</ol>
<div id="function-arguments-10" class="section level3 hasAnchor" number="4.15.1">
<h3><span class="header-section-number">4.15.1</span> Function arguments:<a href="step-by-step-scrna-seq-pipeline.html#function-arguments-10" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><code>data.input</code>: A matrix of expression data, where rows represent genes and columns represent cells. Row names should be in the format of gene symbols.</li>
<li><code>labels</code>: A vector of cluster labels for each cell, corresponding to the columns of data.input.</li>
<li><code>cell.orders</code>: A character vector specifying the order of cell types or clusters in the analysis.</li>
<li><code>cell.colors</code>: A character vector specifying colors for cell types or clusters in the analysis.</li>
<li><code>sample.names</code>: A vector of sample or cell names, corresponding to the columns of data.input.</li>
<li><code>Org</code>: A string indicating the organism used in the analysis. It should be either “mmu” (mouse) or “hsa” (human).</li>
<li><code>sorting</code>: A logical value indicating whether to consider cell population size in communication analysis.</li>
</ul>
</div>
<div id="codes-for-running-step14" class="section level3 hasAnchor" number="4.15.2">
<h3><span class="header-section-number">4.15.2</span> codes for running step14<a href="step-by-step-scrna-seq-pipeline.html#codes-for-running-step14" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Create folders for saving the results of cell-cell interaction analysis.</li>
</ul>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="step-by-step-scrna-seq-pipeline.html#cb79-1" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&#39;Step14. Cell-cell interaction.&#39;</span>)</span>
<span id="cb79-2"><a href="step-by-step-scrna-seq-pipeline.html#cb79-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step14.Cell_cell_interection/&#39;</span>))) {</span>
<span id="cb79-3"><a href="step-by-step-scrna-seq-pipeline.html#cb79-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step14.Cell_cell_interection/&#39;</span>))</span>
<span id="cb79-4"><a href="step-by-step-scrna-seq-pipeline.html#cb79-4" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>Run CellChat to perform cell-cell interaction analysis.</li>
</ul>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="step-by-step-scrna-seq-pipeline.html#cb80-1" tabindex="-1"></a>tempwd <span class="ot">&lt;-</span> <span class="fu">getwd</span>()</span>
<span id="cb80-2"><a href="step-by-step-scrna-seq-pipeline.html#cb80-2" tabindex="-1"></a><span class="fu">run_CellChat</span>(<span class="at">data.input=</span>countsSlot,</span>
<span id="cb80-3"><a href="step-by-step-scrna-seq-pipeline.html#cb80-3" tabindex="-1"></a>             <span class="at">labels =</span> sc_object<span class="sc">@</span>meta.data<span class="sc">$</span>selectLabels,</span>
<span id="cb80-4"><a href="step-by-step-scrna-seq-pipeline.html#cb80-4" tabindex="-1"></a>             <span class="at">cell.orders =</span> ViolinPlot.cellTypeOrders,</span>
<span id="cb80-5"><a href="step-by-step-scrna-seq-pipeline.html#cb80-5" tabindex="-1"></a>             <span class="at">cell.colors =</span> ViolinPlot.cellTypeColors,</span>
<span id="cb80-6"><a href="step-by-step-scrna-seq-pipeline.html#cb80-6" tabindex="-1"></a>             <span class="at">sample.names =</span> <span class="fu">rownames</span>(sc_object<span class="sc">@</span>meta.data),</span>
<span id="cb80-7"><a href="step-by-step-scrna-seq-pipeline.html#cb80-7" tabindex="-1"></a>             <span class="at">Org =</span> Org,</span>
<span id="cb80-8"><a href="step-by-step-scrna-seq-pipeline.html#cb80-8" tabindex="-1"></a>             <span class="at">sorting =</span> sorting,</span>
<span id="cb80-9"><a href="step-by-step-scrna-seq-pipeline.html#cb80-9" tabindex="-1"></a>             <span class="at">output.dir =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/Step14.Cell_cell_interection/&#39;</span>))</span>
<span id="cb80-10"><a href="step-by-step-scrna-seq-pipeline.html#cb80-10" tabindex="-1"></a><span class="fu">setwd</span>(tempwd)</span></code></pre></div>
<ul>
<li>Save the variables.</li>
</ul>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="step-by-step-scrna-seq-pipeline.html#cb81-1" tabindex="-1"></a><span class="co"># Get the names of all variables in the current environment</span></span>
<span id="cb81-2"><a href="step-by-step-scrna-seq-pipeline.html#cb81-2" tabindex="-1"></a>variable_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb81-3"><a href="step-by-step-scrna-seq-pipeline.html#cb81-3" tabindex="-1"></a><span class="co"># Loop through the variable names and save them as RDS files</span></span>
<span id="cb81-4"><a href="step-by-step-scrna-seq-pipeline.html#cb81-4" tabindex="-1"></a><span class="cf">for</span> (var_name <span class="cf">in</span> variable_names) {</span>
<span id="cb81-5"><a href="step-by-step-scrna-seq-pipeline.html#cb81-5" tabindex="-1"></a>  var <span class="ot">&lt;-</span> <span class="fu">get</span>(var_name)  <span class="co"># Get the variable by its name</span></span>
<span id="cb81-6"><a href="step-by-step-scrna-seq-pipeline.html#cb81-6" tabindex="-1"></a>  <span class="fu">saveRDS</span>(var, <span class="at">file =</span> <span class="fu">paste0</span>(output.dir, <span class="st">&#39;/RDSfiles/&#39;</span>, var_name, <span class="st">&quot;.rds&quot;</span>))  <span class="co"># Save as RDS with the variable&#39;s name</span></span>
<span id="cb81-7"><a href="step-by-step-scrna-seq-pipeline.html#cb81-7" tabindex="-1"></a>}  </span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-103"></span>
<img src="sc_img/step14/circle%20plot%20of%20the%20number%20of%20interactions.jpg" alt="Figures showing the interaction number and strength between each cell group" width="50%" /><img src="sc_img/step14/circle%20plot%20of%20the%20interaction%20weights%20or%20strength.jpg" alt="Figures showing the interaction number and strength between each cell group" width="50%" />
<p class="caption">
Figure 4.31: Figures showing the interaction number and strength between each cell group
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-104"></span>
<img src="sc_img/step14/incoming%20signaling.jpg" alt="Heatmap showing the strength of incoming and outgoing signals for each cell type group across various pathways." width="50%" /><img src="sc_img/step14/outgoing%20signaling.jpg" alt="Heatmap showing the strength of incoming and outgoing signals for each cell type group across various pathways." width="50%" />
<p class="caption">
Figure 4.32: Heatmap showing the strength of incoming and outgoing signals for each cell type group across various pathways.
</p>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-105"></span>
<img src="sc_img/step14/cell-cell%20communication%20mediated%20by%20multiple%20ligand-receptors.jpg" alt="Figure showing LRs interaction between each cell type group" width="400px" />
<p class="caption">
Figure 4.33: Figure showing LRs interaction between each cell type group
</p>
</div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="integrated-scrna-seq-pipeline.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="integrated-st-pipeline.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/03-step_by_step_scRNA_seq_pipeline.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
